// @pipecat-ai/small-webrtc-transport/dist/index.js
var $7ytRB$pipecataiclientjs = require("@pipecat-ai/client-js");
var $7ytRB$dailycodailyjs = require("@daily-co/daily-js");
var $7ytRB$lodashcloneDeep = require("lodash/cloneDeep");


function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || Object.prototype.hasOwnProperty.call(dest, key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

$parcel$export(module.exports, "WavMediaManager", () => $23859655abfc5f5c$export$45c5b9bfba2f6304);
$parcel$export(module.exports, "DailyMediaManager", () => $1c088932741d88e6$export$c95c65abc5f47125);
/**
 * Raw wav audio file contents
 * @typedef {Object} WavPackerAudioType
 * @property {Blob} blob
 * @property {string} url
 * @property {number} channelCount
 * @property {number} sampleRate
 * @property {number} duration
 */ /**
 * Utility class for assembling PCM16 "audio/wav" data
 * @class
 */ class $a61750b8fbee4dae$export$13afda237b1c9846 {
    /**
   * Converts Float32Array of amplitude data to ArrayBuffer in Int16Array format
   * @param {Float32Array} float32Array
   * @returns {ArrayBuffer}
   */ static floatTo16BitPCM(float32Array) {
        const buffer = new ArrayBuffer(float32Array.length * 2);
        const view = new DataView(buffer);
        let offset = 0;
        for(let i = 0; i < float32Array.length; i++, offset += 2){
            let s = Math.max(-1, Math.min(1, float32Array[i]));
            view.setInt16(offset, s < 0 ? s * 0x8000 : s * 0x7fff, true);
        }
        return buffer;
    }
    /**
   * Concatenates two ArrayBuffers
   * @param {ArrayBuffer} leftBuffer
   * @param {ArrayBuffer} rightBuffer
   * @returns {ArrayBuffer}
   */ static mergeBuffers(leftBuffer, rightBuffer) {
        const tmpArray = new Uint8Array(leftBuffer.byteLength + rightBuffer.byteLength);
        tmpArray.set(new Uint8Array(leftBuffer), 0);
        tmpArray.set(new Uint8Array(rightBuffer), leftBuffer.byteLength);
        return tmpArray.buffer;
    }
    /**
   * Packs data into an Int16 format
   * @private
   * @param {number} size 0 = 1x Int16, 1 = 2x Int16
   * @param {number} arg value to pack
   * @returns
   */ _packData(size, arg) {
        return [
            new Uint8Array([
                arg,
                arg >> 8
            ]),
            new Uint8Array([
                arg,
                arg >> 8,
                arg >> 16,
                arg >> 24
            ])
        ][size];
    }
    /**
   * Packs audio into "audio/wav" Blob
   * @param {number} sampleRate
   * @param {{bitsPerSample: number, channels: Array<Float32Array>, data: Int16Array}} audio
   * @returns {WavPackerAudioType}
   */ pack(sampleRate, audio) {
        if (!audio?.bitsPerSample) throw new Error(`Missing "bitsPerSample"`);
        else if (!audio?.channels) throw new Error(`Missing "channels"`);
        else if (!audio?.data) throw new Error(`Missing "data"`);
        const { bitsPerSample: bitsPerSample, channels: channels, data: data } = audio;
        const output = [
            // Header
            "RIFF",
            this._packData(1, 52),
            "WAVE",
            // chunk 1
            "fmt ",
            this._packData(1, 16),
            this._packData(0, 1),
            this._packData(0, channels.length),
            this._packData(1, sampleRate),
            this._packData(1, sampleRate * channels.length * bitsPerSample / 8),
            this._packData(0, channels.length * bitsPerSample / 8),
            this._packData(0, bitsPerSample),
            // chunk 2
            "data",
            this._packData(1, channels[0].length * channels.length * bitsPerSample / 8),
            data
        ];
        const blob = new Blob(output, {
            type: "audio/mpeg"
        });
        const url = URL.createObjectURL(blob);
        return {
            blob: blob,
            url: url,
            channelCount: channels.length,
            sampleRate: sampleRate,
            duration: data.byteLength / (channels.length * sampleRate * 2)
        };
    }
}
globalThis.WavPacker = $a61750b8fbee4dae$export$13afda237b1c9846;


/**
 * Constants for help with visualization
 * Helps map frequency ranges from Fast Fourier Transform
 * to human-interpretable ranges, notably music ranges and
 * human vocal ranges.
 */ // Eighth octave frequencies
const $e10a9de47f58e137$var$octave8Frequencies = [
    4186.01,
    4434.92,
    4698.63,
    4978.03,
    5274.04,
    5587.65,
    5919.91,
    6271.93,
    6644.88,
    7040.0,
    7458.62,
    7902.13
];
// Labels for each of the above frequencies
const $e10a9de47f58e137$var$octave8FrequencyLabels = [
    "C",
    "C#",
    "D",
    "D#",
    "E",
    "F",
    "F#",
    "G",
    "G#",
    "A",
    "A#",
    "B"
];
const $e10a9de47f58e137$export$776c63898ae5b636 = [];
const $e10a9de47f58e137$export$facd167cc27ea9b0 = [];
for(let i = 1; i <= 8; i++)for(let f = 0; f < $e10a9de47f58e137$var$octave8Frequencies.length; f++){
    const freq = $e10a9de47f58e137$var$octave8Frequencies[f];
    $e10a9de47f58e137$export$776c63898ae5b636.push(freq / Math.pow(2, 8 - i));
    $e10a9de47f58e137$export$facd167cc27ea9b0.push($e10a9de47f58e137$var$octave8FrequencyLabels[f] + i);
}
/**
 * Subset of the note frequencies between 32 and 2000 Hz
 * 6 octave range: C1 to B6
 */ const $e10a9de47f58e137$var$voiceFrequencyRange = [
    32.0,
    2000.0
];
const $e10a9de47f58e137$export$dbc1581ed2cfa183 = $e10a9de47f58e137$export$776c63898ae5b636.filter((_, i)=>{
    return $e10a9de47f58e137$export$776c63898ae5b636[i] > $e10a9de47f58e137$var$voiceFrequencyRange[0] && $e10a9de47f58e137$export$776c63898ae5b636[i] < $e10a9de47f58e137$var$voiceFrequencyRange[1];
});
const $e10a9de47f58e137$export$30a6f2881311088f = $e10a9de47f58e137$export$facd167cc27ea9b0.filter((_, i)=>{
    return $e10a9de47f58e137$export$776c63898ae5b636[i] > $e10a9de47f58e137$var$voiceFrequencyRange[0] && $e10a9de47f58e137$export$776c63898ae5b636[i] < $e10a9de47f58e137$var$voiceFrequencyRange[1];
});


class $5853841ab58516d4$export$2c3136da0bf130f9 {
    /**
   * Retrieves frequency domain data from an AnalyserNode adjusted to a decibel range
   * returns human-readable formatting and labels
   * @param {AnalyserNode} analyser
   * @param {number} sampleRate
   * @param {Float32Array} [fftResult]
   * @param {"frequency"|"music"|"voice"} [analysisType]
   * @param {number} [minDecibels] default -100
   * @param {number} [maxDecibels] default -30
   * @returns {AudioAnalysisOutputType}
   */ static getFrequencies(analyser, sampleRate, fftResult, analysisType = "frequency", minDecibels = -100, maxDecibels = -30) {
        if (!fftResult) {
            fftResult = new Float32Array(analyser.frequencyBinCount);
            analyser.getFloatFrequencyData(fftResult);
        }
        const nyquistFrequency = sampleRate / 2;
        const frequencyStep = 1 / fftResult.length * nyquistFrequency;
        let outputValues;
        let frequencies;
        let labels;
        if (analysisType === "music" || analysisType === "voice") {
            const useFrequencies = analysisType === "voice" ? (0, $e10a9de47f58e137$export$dbc1581ed2cfa183) : (0, $e10a9de47f58e137$export$776c63898ae5b636);
            const aggregateOutput = Array(useFrequencies.length).fill(minDecibels);
            for(let i = 0; i < fftResult.length; i++){
                const frequency = i * frequencyStep;
                const amplitude = fftResult[i];
                for(let n = useFrequencies.length - 1; n >= 0; n--)if (frequency > useFrequencies[n]) {
                    aggregateOutput[n] = Math.max(aggregateOutput[n], amplitude);
                    break;
                }
            }
            outputValues = aggregateOutput;
            frequencies = analysisType === "voice" ? (0, $e10a9de47f58e137$export$dbc1581ed2cfa183) : (0, $e10a9de47f58e137$export$776c63898ae5b636);
            labels = analysisType === "voice" ? (0, $e10a9de47f58e137$export$30a6f2881311088f) : (0, $e10a9de47f58e137$export$facd167cc27ea9b0);
        } else {
            outputValues = Array.from(fftResult);
            frequencies = outputValues.map((_, i)=>frequencyStep * i);
            labels = frequencies.map((f)=>`${f.toFixed(2)} Hz`);
        }
        // We normalize to {0, 1}
        const normalizedOutput = outputValues.map((v)=>{
            return Math.max(0, Math.min((v - minDecibels) / (maxDecibels - minDecibels), 1));
        });
        const values = new Float32Array(normalizedOutput);
        return {
            values: values,
            frequencies: frequencies,
            labels: labels
        };
    }
    /**
   * Creates a new AudioAnalysis instance for an HTMLAudioElement
   * @param {HTMLAudioElement} audioElement
   * @param {AudioBuffer|null} [audioBuffer] If provided, will cache all frequency domain data from the buffer
   * @returns {AudioAnalysis}
   */ constructor(audioElement, audioBuffer = null){
        this.fftResults = [];
        if (audioBuffer) {
            /**
       * Modified from
       * https://stackoverflow.com/questions/75063715/using-the-web-audio-api-to-analyze-a-song-without-playing
       *
       * We do this to populate FFT values for the audio if provided an `audioBuffer`
       * The reason to do this is that Safari fails when using `createMediaElementSource`
       * This has a non-zero RAM cost so we only opt-in to run it on Safari, Chrome is better
       */ const { length: length, sampleRate: sampleRate } = audioBuffer;
            const offlineAudioContext = new OfflineAudioContext({
                length: length,
                sampleRate: sampleRate
            });
            const source = offlineAudioContext.createBufferSource();
            source.buffer = audioBuffer;
            const analyser = offlineAudioContext.createAnalyser();
            analyser.fftSize = 8192;
            analyser.smoothingTimeConstant = 0.1;
            source.connect(analyser);
            // limit is :: 128 / sampleRate;
            // but we just want 60fps - cuts ~1s from 6MB to 1MB of RAM
            const renderQuantumInSeconds = 1 / 60;
            const durationInSeconds = length / sampleRate;
            const analyze = (index)=>{
                const suspendTime = renderQuantumInSeconds * index;
                if (suspendTime < durationInSeconds) offlineAudioContext.suspend(suspendTime).then(()=>{
                    const fftResult = new Float32Array(analyser.frequencyBinCount);
                    analyser.getFloatFrequencyData(fftResult);
                    this.fftResults.push(fftResult);
                    analyze(index + 1);
                });
                if (index === 1) offlineAudioContext.startRendering();
                else offlineAudioContext.resume();
            };
            source.start(0);
            analyze(1);
            this.audio = audioElement;
            this.context = offlineAudioContext;
            this.analyser = analyser;
            this.sampleRate = sampleRate;
            this.audioBuffer = audioBuffer;
        } else {
            const audioContext = new AudioContext();
            const track = audioContext.createMediaElementSource(audioElement);
            const analyser = audioContext.createAnalyser();
            analyser.fftSize = 8192;
            analyser.smoothingTimeConstant = 0.1;
            track.connect(analyser);
            analyser.connect(audioContext.destination);
            this.audio = audioElement;
            this.context = audioContext;
            this.analyser = analyser;
            this.sampleRate = this.context.sampleRate;
            this.audioBuffer = null;
        }
    }
    /**
   * Gets the current frequency domain data from the playing audio track
   * @param {"frequency"|"music"|"voice"} [analysisType]
   * @param {number} [minDecibels] default -100
   * @param {number} [maxDecibels] default -30
   * @returns {AudioAnalysisOutputType}
   */ getFrequencies(analysisType = "frequency", minDecibels = -100, maxDecibels = -30) {
        let fftResult = null;
        if (this.audioBuffer && this.fftResults.length) {
            const pct = this.audio.currentTime / this.audio.duration;
            const index = Math.min(pct * this.fftResults.length | 0, this.fftResults.length - 1);
            fftResult = this.fftResults[index];
        }
        return $5853841ab58516d4$export$2c3136da0bf130f9.getFrequencies(this.analyser, this.sampleRate, fftResult, analysisType, minDecibels, maxDecibels);
    }
    /**
   * Resume the internal AudioContext if it was suspended due to the lack of
   * user interaction when the AudioAnalysis was instantiated.
   * @returns {Promise<true>}
   */ async resumeIfSuspended() {
        if (this.context.state === "suspended") await this.context.resume();
        return true;
    }
}
globalThis.AudioAnalysis = $5853841ab58516d4$export$2c3136da0bf130f9;


const $d6d3743ef65c7519$export$50b76700e2b15e9 = `
class StreamProcessor extends AudioWorkletProcessor {
  constructor() {
    super();
    this.hasStarted = false;
    this.hasInterrupted = false;
    this.outputBuffers = [];
    this.bufferLength = 128;
    this.write = { buffer: new Float32Array(this.bufferLength), trackId: null };
    this.writeOffset = 0;
    this.trackSampleOffsets = {};
    this.port.onmessage = (event) => {
      if (event.data) {
        const payload = event.data;
        if (payload.event === 'write') {
          const int16Array = payload.buffer;
          const float32Array = new Float32Array(int16Array.length);
          for (let i = 0; i < int16Array.length; i++) {
            float32Array[i] = int16Array[i] / 0x8000; // Convert Int16 to Float32
          }
          this.writeData(float32Array, payload.trackId);
        } else if (
          payload.event === 'offset' ||
          payload.event === 'interrupt'
        ) {
          const requestId = payload.requestId;
          const trackId = this.write.trackId;
          const offset = this.trackSampleOffsets[trackId] || 0;
          this.port.postMessage({
            event: 'offset',
            requestId,
            trackId,
            offset,
          });
          if (payload.event === 'interrupt') {
            this.hasInterrupted = true;
          }
        } else {
          throw new Error(\`Unhandled event "\${payload.event}"\`);
        }
      }
    };
  }

  writeData(float32Array, trackId = null) {
    let { buffer } = this.write;
    let offset = this.writeOffset;
    for (let i = 0; i < float32Array.length; i++) {
      buffer[offset++] = float32Array[i];
      if (offset >= buffer.length) {
        this.outputBuffers.push(this.write);
        this.write = { buffer: new Float32Array(this.bufferLength), trackId };
        buffer = this.write.buffer;
        offset = 0;
      }
    }
    this.writeOffset = offset;
    return true;
  }

  process(inputs, outputs, parameters) {
    const output = outputs[0];
    const outputChannelData = output[0];
    const outputBuffers = this.outputBuffers;
    if (this.hasInterrupted) {
      this.port.postMessage({ event: 'stop' });
      return false;
    } else if (outputBuffers.length) {
      this.hasStarted = true;
      const { buffer, trackId } = outputBuffers.shift();
      for (let i = 0; i < outputChannelData.length; i++) {
        outputChannelData[i] = buffer[i] || 0;
      }
      if (trackId) {
        this.trackSampleOffsets[trackId] =
          this.trackSampleOffsets[trackId] || 0;
        this.trackSampleOffsets[trackId] += buffer.length;
      }
      return true;
    } else if (this.hasStarted) {
      this.port.postMessage({ event: 'stop' });
      return false;
    } else {
      return true;
    }
  }
}

registerProcessor('stream_processor', StreamProcessor);
`;
const $d6d3743ef65c7519$var$script = new Blob([
    $d6d3743ef65c7519$export$50b76700e2b15e9
], {
    type: "application/javascript"
});
const $d6d3743ef65c7519$var$src = URL.createObjectURL($d6d3743ef65c7519$var$script);
const $d6d3743ef65c7519$export$bfa8c596114d74df = $d6d3743ef65c7519$var$src;



class $1e7ce9484a3c4077$export$9698d62c78b8f366 {
    /**
   * Creates a new WavStreamPlayer instance
   * @param {{sampleRate?: number}} options
   * @returns {WavStreamPlayer}
   */ constructor({ sampleRate: sampleRate = 44100 } = {}){
        this.scriptSrc = (0, $d6d3743ef65c7519$export$bfa8c596114d74df);
        this.sampleRate = sampleRate;
        this.context = null;
        this.stream = null;
        this.analyser = null;
        this.trackSampleOffsets = {};
        this.interruptedTrackIds = {};
    }
    /**
   * Connects the audio context and enables output to speakers
   * @returns {Promise<true>}
   */ async connect() {
        this.context = new AudioContext({
            sampleRate: this.sampleRate
        });
        if (this._speakerID) this.context.setSinkId(this._speakerID);
        if (this.context.state === "suspended") await this.context.resume();
        try {
            await this.context.audioWorklet.addModule(this.scriptSrc);
        } catch (e) {
            console.error(e);
            throw new Error(`Could not add audioWorklet module: ${this.scriptSrc}`);
        }
        const analyser = this.context.createAnalyser();
        analyser.fftSize = 8192;
        analyser.smoothingTimeConstant = 0.1;
        this.analyser = analyser;
        return true;
    }
    /**
   * Gets the current frequency domain data from the playing track
   * @param {"frequency"|"music"|"voice"} [analysisType]
   * @param {number} [minDecibels] default -100
   * @param {number} [maxDecibels] default -30
   * @returns {import('./analysis/audio_analysis.js').AudioAnalysisOutputType}
   */ getFrequencies(analysisType = "frequency", minDecibels = -100, maxDecibels = -30) {
        if (!this.analyser) throw new Error("Not connected, please call .connect() first");
        return (0, $5853841ab58516d4$export$2c3136da0bf130f9).getFrequencies(this.analyser, this.sampleRate, null, analysisType, minDecibels, maxDecibels);
    }
    /**
   * @param {string} speaker deviceId
   */ async updateSpeaker(speaker) {
        const _prevSpeaker = this._speakerID;
        this._speakerID = speaker;
        if (this.context) try {
            if (speaker === "default") await this.context.setSinkId();
            else await this.context.setSinkId(speaker);
        } catch (e) {
            console.error(`Could not set sinkId to ${speaker}: ${e}`);
            this._speakerID = _prevSpeaker;
        }
    }
    /**
   * Starts audio streaming
   * @private
   * @returns {Promise<true>}
   */ _start() {
        const streamNode = new AudioWorkletNode(this.context, "stream_processor");
        streamNode.connect(this.context.destination);
        streamNode.port.onmessage = (e)=>{
            const { event: event } = e.data;
            if (event === "stop") {
                streamNode.disconnect();
                this.stream = null;
            } else if (event === "offset") {
                const { requestId: requestId, trackId: trackId, offset: offset } = e.data;
                const currentTime = offset / this.sampleRate;
                this.trackSampleOffsets[requestId] = {
                    trackId: trackId,
                    offset: offset,
                    currentTime: currentTime
                };
            }
        };
        this.analyser.disconnect();
        streamNode.connect(this.analyser);
        this.stream = streamNode;
        return true;
    }
    /**
   * Adds 16BitPCM data to the currently playing audio stream
   * You can add chunks beyond the current play point and they will be queued for play
   * @param {ArrayBuffer|Int16Array} arrayBuffer
   * @param {string} [trackId]
   * @returns {Int16Array}
   */ add16BitPCM(arrayBuffer, trackId = "default") {
        if (typeof trackId !== "string") throw new Error(`trackId must be a string`);
        else if (this.interruptedTrackIds[trackId]) return;
        if (!this.stream) this._start();
        let buffer;
        if (arrayBuffer instanceof Int16Array) buffer = arrayBuffer;
        else if (arrayBuffer instanceof ArrayBuffer) buffer = new Int16Array(arrayBuffer);
        else throw new Error(`argument must be Int16Array or ArrayBuffer`);
        this.stream.port.postMessage({
            event: "write",
            buffer: buffer,
            trackId: trackId
        });
        return buffer;
    }
    /**
   * Gets the offset (sample count) of the currently playing stream
   * @param {boolean} [interrupt]
   * @returns {{trackId: string|null, offset: number, currentTime: number}}
   */ async getTrackSampleOffset(interrupt = false) {
        if (!this.stream) return null;
        const requestId = crypto.randomUUID();
        this.stream.port.postMessage({
            event: interrupt ? "interrupt" : "offset",
            requestId: requestId
        });
        let trackSampleOffset;
        while(!trackSampleOffset){
            trackSampleOffset = this.trackSampleOffsets[requestId];
            await new Promise((r)=>setTimeout(()=>r(), 1));
        }
        const { trackId: trackId } = trackSampleOffset;
        if (interrupt && trackId) this.interruptedTrackIds[trackId] = true;
        return trackSampleOffset;
    }
    /**
   * Strips the current stream and returns the sample offset of the audio
   * @param {boolean} [interrupt]
   * @returns {{trackId: string|null, offset: number, currentTime: number}}
   */ async interrupt() {
        return this.getTrackSampleOffset(true);
    }
}
globalThis.WavStreamPlayer = $1e7ce9484a3c4077$export$9698d62c78b8f366;


const $2cf6a2d8a6d031bc$var$AudioProcessorWorklet = `
class AudioProcessor extends AudioWorkletProcessor {

  constructor() {
    super();
    this.port.onmessage = this.receive.bind(this);
    this.initialize();
  }

  initialize() {
    this.foundAudio = false;
    this.recording = false;
    this.chunks = [];
  }

  /**
   * Concatenates sampled chunks into channels
   * Format is chunk[Left[], Right[]]
   */
  readChannelData(chunks, channel = -1, maxChannels = 9) {
    let channelLimit;
    if (channel !== -1) {
      if (chunks[0] && chunks[0].length - 1 < channel) {
        throw new Error(
          \`Channel \${channel} out of range: max \${chunks[0].length}\`
        );
      }
      channelLimit = channel + 1;
    } else {
      channel = 0;
      channelLimit = Math.min(chunks[0] ? chunks[0].length : 1, maxChannels);
    }
    const channels = [];
    for (let n = channel; n < channelLimit; n++) {
      const length = chunks.reduce((sum, chunk) => {
        return sum + chunk[n].length;
      }, 0);
      const buffers = chunks.map((chunk) => chunk[n]);
      const result = new Float32Array(length);
      let offset = 0;
      for (let i = 0; i < buffers.length; i++) {
        result.set(buffers[i], offset);
        offset += buffers[i].length;
      }
      channels[n] = result;
    }
    return channels;
  }

  /**
   * Combines parallel audio data into correct format,
   * channels[Left[], Right[]] to float32Array[LRLRLRLR...]
   */
  formatAudioData(channels) {
    if (channels.length === 1) {
      // Simple case is only one channel
      const float32Array = channels[0].slice();
      const meanValues = channels[0].slice();
      return { float32Array, meanValues };
    } else {
      const float32Array = new Float32Array(
        channels[0].length * channels.length
      );
      const meanValues = new Float32Array(channels[0].length);
      for (let i = 0; i < channels[0].length; i++) {
        const offset = i * channels.length;
        let meanValue = 0;
        for (let n = 0; n < channels.length; n++) {
          float32Array[offset + n] = channels[n][i];
          meanValue += channels[n][i];
        }
        meanValues[i] = meanValue / channels.length;
      }
      return { float32Array, meanValues };
    }
  }

  /**
   * Converts 32-bit float data to 16-bit integers
   */
  floatTo16BitPCM(float32Array) {
    const buffer = new ArrayBuffer(float32Array.length * 2);
    const view = new DataView(buffer);
    let offset = 0;
    for (let i = 0; i < float32Array.length; i++, offset += 2) {
      let s = Math.max(-1, Math.min(1, float32Array[i]));
      view.setInt16(offset, s < 0 ? s * 0x8000 : s * 0x7fff, true);
    }
    return buffer;
  }

  /**
   * Retrieves the most recent amplitude values from the audio stream
   * @param {number} channel
   */
  getValues(channel = -1) {
    const channels = this.readChannelData(this.chunks, channel);
    const { meanValues } = this.formatAudioData(channels);
    return { meanValues, channels };
  }

  /**
   * Exports chunks as an audio/wav file
   */
  export() {
    const channels = this.readChannelData(this.chunks);
    const { float32Array, meanValues } = this.formatAudioData(channels);
    const audioData = this.floatTo16BitPCM(float32Array);
    return {
      meanValues: meanValues,
      audio: {
        bitsPerSample: 16,
        channels: channels,
        data: audioData,
      },
    };
  }

  receive(e) {
    const { event, id } = e.data;
    let receiptData = {};
    switch (event) {
      case 'start':
        this.recording = true;
        break;
      case 'stop':
        this.recording = false;
        break;
      case 'clear':
        this.initialize();
        break;
      case 'export':
        receiptData = this.export();
        break;
      case 'read':
        receiptData = this.getValues();
        break;
      default:
        break;
    }
    // Always send back receipt
    this.port.postMessage({ event: 'receipt', id, data: receiptData });
  }

  sendChunk(chunk) {
    const channels = this.readChannelData([chunk]);
    const { float32Array, meanValues } = this.formatAudioData(channels);
    const rawAudioData = this.floatTo16BitPCM(float32Array);
    const monoAudioData = this.floatTo16BitPCM(meanValues);
    this.port.postMessage({
      event: 'chunk',
      data: {
        mono: monoAudioData,
        raw: rawAudioData,
      },
    });
  }

  process(inputList, outputList, parameters) {
    // Copy input to output (e.g. speakers)
    // Note that this creates choppy sounds with Mac products
    const sourceLimit = Math.min(inputList.length, outputList.length);
    for (let inputNum = 0; inputNum < sourceLimit; inputNum++) {
      const input = inputList[inputNum];
      const output = outputList[inputNum];
      const channelCount = Math.min(input.length, output.length);
      for (let channelNum = 0; channelNum < channelCount; channelNum++) {
        input[channelNum].forEach((sample, i) => {
          output[channelNum][i] = sample;
        });
      }
    }
    const inputs = inputList[0];
    // There's latency at the beginning of a stream before recording starts
    // Make sure we actually receive audio data before we start storing chunks
    let sliceIndex = 0;
    if (!this.foundAudio) {
      for (const channel of inputs) {
        sliceIndex = 0; // reset for each channel
        if (this.foundAudio) {
          break;
        }
        if (channel) {
          for (const value of channel) {
            if (value !== 0) {
              // find only one non-zero entry in any channel
              this.foundAudio = true;
              break;
            } else {
              sliceIndex++;
            }
          }
        }
      }
    }
    if (inputs && inputs[0] && this.foundAudio && this.recording) {
      // We need to copy the TypedArray, because the \`process\`
      // internals will reuse the same buffer to hold each input
      const chunk = inputs.map((input) => input.slice(sliceIndex));
      this.chunks.push(chunk);
      this.sendChunk(chunk);
    }
    return true;
  }
}

registerProcessor('audio_processor', AudioProcessor);
`;
const $2cf6a2d8a6d031bc$var$script = new Blob([
    $2cf6a2d8a6d031bc$var$AudioProcessorWorklet
], {
    type: "application/javascript"
});
const $2cf6a2d8a6d031bc$var$src = URL.createObjectURL($2cf6a2d8a6d031bc$var$script);
const $2cf6a2d8a6d031bc$export$1f65f50a8cbff43c = $2cf6a2d8a6d031bc$var$src;




class $0d471c38435bdab6$export$439b217ca659a877 {
    /**
   * Create a new WavRecorder instance
   * @param {{sampleRate?: number, outputToSpeakers?: boolean, debug?: boolean}} [options]
   * @returns {WavRecorder}
   */ constructor({ sampleRate: sampleRate = 44100, outputToSpeakers: outputToSpeakers = false, debug: debug = false } = {}){
        // Script source
        this.scriptSrc = (0, $2cf6a2d8a6d031bc$export$1f65f50a8cbff43c);
        // Config
        this.sampleRate = sampleRate;
        this.outputToSpeakers = outputToSpeakers;
        this.debug = !!debug;
        this._deviceChangeCallback = null;
        this._deviceErrorCallback = null;
        this._devices = [];
        this.deviceSelection = null;
        // State variables
        this.stream = null;
        this.processor = null;
        this.source = null;
        this.node = null;
        this.recording = false;
        // Event handling with AudioWorklet
        this._lastEventId = 0;
        this.eventReceipts = {};
        this.eventTimeout = 5000;
        // Process chunks of audio
        this._chunkProcessor = ()=>{};
        this._chunkProcessorSize = void 0;
        this._chunkProcessorBuffer = {
            raw: new ArrayBuffer(0),
            mono: new ArrayBuffer(0)
        };
    }
    /**
   * Decodes audio data from multiple formats to a Blob, url, Float32Array and AudioBuffer
   * @param {Blob|Float32Array|Int16Array|ArrayBuffer|number[]} audioData
   * @param {number} sampleRate
   * @param {number} fromSampleRate
   * @returns {Promise<DecodedAudioType>}
   */ static async decode(audioData, sampleRate = 44100, fromSampleRate = -1) {
        const context = new AudioContext({
            sampleRate: sampleRate
        });
        let arrayBuffer;
        let blob;
        if (audioData instanceof Blob) {
            if (fromSampleRate !== -1) throw new Error(`Can not specify "fromSampleRate" when reading from Blob`);
            blob = audioData;
            arrayBuffer = await blob.arrayBuffer();
        } else if (audioData instanceof ArrayBuffer) {
            if (fromSampleRate !== -1) throw new Error(`Can not specify "fromSampleRate" when reading from ArrayBuffer`);
            arrayBuffer = audioData;
            blob = new Blob([
                arrayBuffer
            ], {
                type: "audio/wav"
            });
        } else {
            let float32Array;
            let data;
            if (audioData instanceof Int16Array) {
                data = audioData;
                float32Array = new Float32Array(audioData.length);
                for(let i = 0; i < audioData.length; i++)float32Array[i] = audioData[i] / 0x8000;
            } else if (audioData instanceof Float32Array) float32Array = audioData;
            else if (audioData instanceof Array) float32Array = new Float32Array(audioData);
            else throw new Error(`"audioData" must be one of: Blob, Float32Arrray, Int16Array, ArrayBuffer, Array<number>`);
            if (fromSampleRate === -1) throw new Error(`Must specify "fromSampleRate" when reading from Float32Array, In16Array or Array`);
            else if (fromSampleRate < 3000) throw new Error(`Minimum "fromSampleRate" is 3000 (3kHz)`);
            if (!data) data = (0, $a61750b8fbee4dae$export$13afda237b1c9846).floatTo16BitPCM(float32Array);
            const audio = {
                bitsPerSample: 16,
                channels: [
                    float32Array
                ],
                data: data
            };
            const packer = new (0, $a61750b8fbee4dae$export$13afda237b1c9846)();
            const result = packer.pack(fromSampleRate, audio);
            blob = result.blob;
            arrayBuffer = await blob.arrayBuffer();
        }
        const audioBuffer = await context.decodeAudioData(arrayBuffer);
        const values = audioBuffer.getChannelData(0);
        const url = URL.createObjectURL(blob);
        return {
            blob: blob,
            url: url,
            values: values,
            audioBuffer: audioBuffer
        };
    }
    /**
   * Logs data in debug mode
   * @param {...any} arguments
   * @returns {true}
   */ log() {
        if (this.debug) this.log(...arguments);
        return true;
    }
    /**
   * Retrieves the current sampleRate for the recorder
   * @returns {number}
   */ getSampleRate() {
        return this.sampleRate;
    }
    /**
   * Retrieves the current status of the recording
   * @returns {"ended"|"paused"|"recording"}
   */ getStatus() {
        if (!this.processor) return "ended";
        else if (!this.recording) return "paused";
        else return "recording";
    }
    /**
   * Sends an event to the AudioWorklet
   * @private
   * @param {string} name
   * @param {{[key: string]: any}} data
   * @param {AudioWorkletNode} [_processor]
   * @returns {Promise<{[key: string]: any}>}
   */ async _event(name, data = {}, _processor = null) {
        _processor = _processor || this.processor;
        if (!_processor) throw new Error("Can not send events without recording first");
        const message = {
            event: name,
            id: this._lastEventId++,
            data: data
        };
        _processor.port.postMessage(message);
        const t0 = new Date().valueOf();
        while(!this.eventReceipts[message.id]){
            if (new Date().valueOf() - t0 > this.eventTimeout) throw new Error(`Timeout waiting for "${name}" event`);
            await new Promise((res)=>setTimeout(()=>res(true), 1));
        }
        const payload = this.eventReceipts[message.id];
        delete this.eventReceipts[message.id];
        return payload;
    }
    /**
   * Sets device change callback, remove if callback provided is `null`
   * @param {(Array<MediaDeviceInfo & {default: boolean}>): void|null} callback
   * @returns {true}
   */ listenForDeviceChange(callback) {
        if (callback === null && this._deviceChangeCallback) {
            navigator.mediaDevices.removeEventListener("devicechange", this._deviceChangeCallback);
            this._deviceChangeCallback = null;
        } else if (callback !== null) {
            // Basically a debounce; we only want this called once when devices change
            // And we only want the most recent callback() to be executed
            // if a few are operating at the same time
            let lastId = 0;
            let lastDevices = [];
            const serializeDevices = (devices)=>devices.map((d)=>d.deviceId).sort().join(",");
            const cb = async ()=>{
                let id = ++lastId;
                const devices = await this.listDevices();
                if (id === lastId) {
                    if (serializeDevices(lastDevices) !== serializeDevices(devices)) {
                        lastDevices = devices;
                        callback(devices.slice());
                    }
                }
            };
            navigator.mediaDevices.addEventListener("devicechange", cb);
            cb();
            this._deviceChangeCallback = cb;
        }
        return true;
    }
    /**
   * Provide a callback for if/when device errors occur
   * @param {(({devices: Array<"cam" | "mic">, type: string, error?: Error}) => void) | null} callback
   * @returns {true}
   */ listenForDeviceErrors(callback) {
        this._deviceErrorCallback = callback;
    }
    /**
   * Manually request permission to use the microphone
   * @returns {Promise<true>}
   */ async requestPermission() {
        const permissionStatus = await navigator.permissions.query({
            name: "microphone"
        });
        if (permissionStatus.state === "denied") {
            if (this._deviceErrorCallback) this._deviceErrorCallback({
                devices: [
                    "mic"
                ],
                type: "unknown",
                error: new Error("Microphone access denied")
            });
        } else if (permissionStatus.state === "prompt") try {
            const stream = await navigator.mediaDevices.getUserMedia({
                audio: true
            });
            const tracks = stream.getTracks();
            tracks.forEach((track)=>track.stop());
        } catch (e) {
            console.error("Error accessing microphone.");
            if (this._deviceErrorCallback) this._deviceErrorCallback({
                devices: [
                    "mic"
                ],
                type: "unknown",
                error: e
            });
        }
        return true;
    }
    /**
   * List all eligible devices for recording, will request permission to use microphone
   * @returns {Promise<Array<MediaDeviceInfo & {default: boolean}>>}
   */ async listDevices() {
        if (!navigator.mediaDevices || !("enumerateDevices" in navigator.mediaDevices)) throw new Error("Could not request user devices");
        await this.requestPermission();
        const devices = await navigator.mediaDevices.enumerateDevices();
        const audioDevices = devices.filter((device)=>device.kind === "audioinput");
        return audioDevices;
    // const defaultDeviceIndex = audioDevices.findIndex(
    //   (device) => device.deviceId === 'default'
    // );
    // const deviceList = [];
    // if (defaultDeviceIndex !== -1) {
    //   let defaultDevice = audioDevices.splice(defaultDeviceIndex, 1)[0];
    //   let existingIndex = audioDevices.findIndex(
    //     (device) => device.groupId === defaultDevice.groupId
    //   );
    //   if (existingIndex !== -1) {
    //     defaultDevice = audioDevices.splice(existingIndex, 1)[0];
    //   }
    //   defaultDevice.default = true;
    //   deviceList.push(defaultDevice);
    // }
    // return deviceList.concat(audioDevices);
    }
    /**
   * Begins a recording session and requests microphone permissions if not already granted
   * Microphone recording indicator will appear on browser tab but status will be "paused"
   * @param {string} [deviceId] if no device provided, default device will be used
   * @returns {Promise<true>}
   */ async begin(deviceId) {
        if (this.processor) throw new Error(`Already connected: please call .end() to start a new session`);
        if (!navigator.mediaDevices || !("getUserMedia" in navigator.mediaDevices)) {
            if (this._deviceErrorCallback) this._deviceErrorCallback({
                devices: [
                    "mic",
                    "cam"
                ],
                type: "undefined-mediadevices"
            });
            throw new Error("Could not request user media");
        }
        deviceId = deviceId ?? this.deviceSelection?.deviceId;
        try {
            const config = {
                audio: true
            };
            if (deviceId) config.audio = {
                deviceId: {
                    exact: deviceId
                }
            };
            this.stream = await navigator.mediaDevices.getUserMedia(config);
        } catch (err) {
            if (this._deviceErrorCallback) this._deviceErrorCallback({
                devices: [
                    "mic"
                ],
                type: "unknown",
                error: err
            });
            throw new Error("Could not start media stream");
        }
        this.listDevices().then((devices)=>{
            deviceId = this.stream.getAudioTracks()[0].getSettings().deviceId;
            console.log("find current device", devices, deviceId, this.stream.getAudioTracks()[0].getSettings());
            this.deviceSelection = devices.find((d)=>d.deviceId === deviceId);
            console.log("current device", this.deviceSelection);
        });
        const context = new AudioContext({
            sampleRate: this.sampleRate
        });
        const source = context.createMediaStreamSource(this.stream);
        // Load and execute the module script.
        try {
            await context.audioWorklet.addModule(this.scriptSrc);
        } catch (e) {
            console.error(e);
            throw new Error(`Could not add audioWorklet module: ${this.scriptSrc}`);
        }
        const processor = new AudioWorkletNode(context, "audio_processor");
        processor.port.onmessage = (e)=>{
            const { event: event, id: id, data: data } = e.data;
            if (event === "receipt") this.eventReceipts[id] = data;
            else if (event === "chunk") {
                if (this._chunkProcessorSize) {
                    const buffer = this._chunkProcessorBuffer;
                    this._chunkProcessorBuffer = {
                        raw: (0, $a61750b8fbee4dae$export$13afda237b1c9846).mergeBuffers(buffer.raw, data.raw),
                        mono: (0, $a61750b8fbee4dae$export$13afda237b1c9846).mergeBuffers(buffer.mono, data.mono)
                    };
                    if (this._chunkProcessorBuffer.mono.byteLength >= this._chunkProcessorSize) {
                        this._chunkProcessor(this._chunkProcessorBuffer);
                        this._chunkProcessorBuffer = {
                            raw: new ArrayBuffer(0),
                            mono: new ArrayBuffer(0)
                        };
                    }
                } else this._chunkProcessor(data);
            }
        };
        const node = source.connect(processor);
        const analyser = context.createAnalyser();
        analyser.fftSize = 8192;
        analyser.smoothingTimeConstant = 0.1;
        node.connect(analyser);
        if (this.outputToSpeakers) {
            // eslint-disable-next-line no-console
            console.warn("Warning: Output to speakers may affect sound quality,\nespecially due to system audio feedback preventative measures.\nuse only for debugging");
            analyser.connect(context.destination);
        }
        this.source = source;
        this.node = node;
        this.analyser = analyser;
        this.processor = processor;
        console.log("begin completed");
        return true;
    }
    /**
   * Gets the current frequency domain data from the recording track
   * @param {"frequency"|"music"|"voice"} [analysisType]
   * @param {number} [minDecibels] default -100
   * @param {number} [maxDecibels] default -30
   * @returns {import('./analysis/audio_analysis.js').AudioAnalysisOutputType}
   */ getFrequencies(analysisType = "frequency", minDecibels = -100, maxDecibels = -30) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        return (0, $5853841ab58516d4$export$2c3136da0bf130f9).getFrequencies(this.analyser, this.sampleRate, null, analysisType, minDecibels, maxDecibels);
    }
    /**
   * Pauses the recording
   * Keeps microphone stream open but halts storage of audio
   * @returns {Promise<true>}
   */ async pause() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        else if (!this.recording) throw new Error("Already paused: please call .record() first");
        if (this._chunkProcessorBuffer.raw.byteLength) this._chunkProcessor(this._chunkProcessorBuffer);
        this.log("Pausing ...");
        await this._event("stop");
        this.recording = false;
        return true;
    }
    /**
   * Start recording stream and storing to memory from the connected audio source
   * @param {(data: { mono: Int16Array; raw: Int16Array }) => any} [chunkProcessor]
   * @param {number} [chunkSize] chunkProcessor will not be triggered until this size threshold met in mono audio
   * @returns {Promise<true>}
   */ async record(chunkProcessor = ()=>{}, chunkSize = 8192) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        else if (this.recording) throw new Error("Already recording: please call .pause() first");
        else if (typeof chunkProcessor !== "function") throw new Error(`chunkProcessor must be a function`);
        this._chunkProcessor = chunkProcessor;
        this._chunkProcessorSize = chunkSize;
        this._chunkProcessorBuffer = {
            raw: new ArrayBuffer(0),
            mono: new ArrayBuffer(0)
        };
        this.log("Recording ...");
        await this._event("start");
        this.recording = true;
        return true;
    }
    /**
   * Clears the audio buffer, empties stored recording
   * @returns {Promise<true>}
   */ async clear() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        await this._event("clear");
        return true;
    }
    /**
   * Reads the current audio stream data
   * @returns {Promise<{meanValues: Float32Array, channels: Array<Float32Array>}>}
   */ async read() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        this.log("Reading ...");
        const result = await this._event("read");
        return result;
    }
    /**
   * Saves the current audio stream to a file
   * @param {boolean} [force] Force saving while still recording
   * @returns {Promise<import('./wav_packer.js').WavPackerAudioType>}
   */ async save(force = false) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        if (!force && this.recording) throw new Error("Currently recording: please call .pause() first, or call .save(true) to force");
        this.log("Exporting ...");
        const exportData = await this._event("export");
        const packer = new (0, $a61750b8fbee4dae$export$13afda237b1c9846)();
        const result = packer.pack(this.sampleRate, exportData.audio);
        return result;
    }
    /**
   * Ends the current recording session and saves the result
   * @returns {Promise<import('./wav_packer.js').WavPackerAudioType>}
   */ async end() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        const _processor = this.processor;
        this.log("Stopping ...");
        await this._event("stop");
        this.recording = false;
        const tracks = this.stream.getTracks();
        tracks.forEach((track)=>track.stop());
        this.log("Exporting ...");
        const exportData = await this._event("export", {}, _processor);
        this.processor.disconnect();
        this.source.disconnect();
        this.node.disconnect();
        this.analyser.disconnect();
        this.stream = null;
        this.processor = null;
        this.source = null;
        this.node = null;
        const packer = new (0, $a61750b8fbee4dae$export$13afda237b1c9846)();
        const result = packer.pack(this.sampleRate, exportData.audio);
        return result;
    }
    /**
   * Performs a full cleanup of WavRecorder instance
   * Stops actively listening via microphone and removes existing listeners
   * @returns {Promise<true>}
   */ async quit() {
        this.listenForDeviceChange(null);
        // we do not reset this on end so that selections persist across starts
        this.deviceSelection = null;
        if (this.processor) await this.end();
        return true;
    }
}
globalThis.WavRecorder = $0d471c38435bdab6$export$439b217ca659a877;





/**
 * Resamples audio data from one sample rate to another using linear interpolation
 * @param {ArrayBuffer} inputBuffer - Input audio data as Int16 PCM
 * @param {number} inputSampleRate - Sample rate of input audio
 * @param {number} outputSampleRate - Desired output sample rate
 * @returns {ArrayBuffer} - Resampled audio data as Int16 PCM
 */ function $7cef7a69bdf8f84d$var$resampleAudioBuffer(inputBuffer, inputSampleRate, outputSampleRate) {
    if (inputSampleRate === outputSampleRate) return inputBuffer;
    const inputView = new Int16Array(inputBuffer);
    const ratio = inputSampleRate / outputSampleRate;
    const outputLength = Math.round(inputView.length / ratio);
    const outputBuffer = new ArrayBuffer(outputLength * 2);
    const outputView = new Int16Array(outputBuffer);
    for(let i = 0; i < outputLength; i++){
        const srcIndex = i * ratio;
        const srcIndexFloor = Math.floor(srcIndex);
        const srcIndexCeil = Math.min(srcIndexFloor + 1, inputView.length - 1);
        const t = srcIndex - srcIndexFloor;
        // Linear interpolation
        outputView[i] = Math.round(inputView[srcIndexFloor] * (1 - t) + inputView[srcIndexCeil] * t);
    }
    return outputBuffer;
}
class $7cef7a69bdf8f84d$export$2934cf2d25c67a48 {
    /**
   * Create a new MediaStreamRecorder instance
   * @param {{sampleRate?: number, outputToSpeakers?: boolean, debug?: boolean}} [options]
   * @returns {MediaStreamRecorder}
   */ constructor({ sampleRate: sampleRate = 44100, outputToSpeakers: outputToSpeakers = false, debug: debug = false } = {}){
        // Script source
        this.scriptSrc = (0, $2cf6a2d8a6d031bc$export$1f65f50a8cbff43c);
        // Config
        this.sampleRate = sampleRate;
        this.outputToSpeakers = outputToSpeakers;
        this.debug = !!debug;
        // State variables
        this.stream = null;
        this.processor = null;
        this.source = null;
        this.node = null;
        this.recording = false;
        // Event handling with AudioWorklet
        this._lastEventId = 0;
        this.eventReceipts = {};
        this.eventTimeout = 5000;
        // Process chunks of audio
        this._chunkProcessor = ()=>{};
        this._chunkProcessorSize = void 0;
        this._chunkProcessorBuffer = {
            raw: new ArrayBuffer(0),
            mono: new ArrayBuffer(0)
        };
    }
    /**
   * Logs data in debug mode
   * @param {...any} arguments
   * @returns {true}
   */ log() {
        if (this.debug) this.log(...arguments);
        return true;
    }
    /**
   * Retrieves the current sampleRate for the recorder
   * @returns {number}
   */ getSampleRate() {
        return this.sampleRate;
    }
    /**
   * Retrieves the current status of the recording
   * @returns {"ended"|"paused"|"recording"}
   */ getStatus() {
        if (!this.processor) return "ended";
        else if (!this.recording) return "paused";
        else return "recording";
    }
    /**
   * Sends an event to the AudioWorklet
   * @private
   * @param {string} name
   * @param {{[key: string]: any}} data
   * @param {AudioWorkletNode} [_processor]
   * @returns {Promise<{[key: string]: any}>}
   */ async _event(name, data = {}, _processor = null) {
        _processor = _processor || this.processor;
        if (!_processor) throw new Error("Can not send events without recording first");
        const message = {
            event: name,
            id: this._lastEventId++,
            data: data
        };
        _processor.port.postMessage(message);
        const t0 = new Date().valueOf();
        while(!this.eventReceipts[message.id]){
            if (new Date().valueOf() - t0 > this.eventTimeout) throw new Error(`Timeout waiting for "${name}" event`);
            await new Promise((res)=>setTimeout(()=>res(true), 1));
        }
        const payload = this.eventReceipts[message.id];
        delete this.eventReceipts[message.id];
        return payload;
    }
    /**
   * Begins a recording session for the given audioTrack
   * Microphone recording indicator will appear on browser tab but status will be "paused"
   * @param {MediaStreamTrack} [audioTrack] if no device provided, default device will be used
   * @returns {Promise<true>}
   */ async begin(audioTrack) {
        if (this.processor) throw new Error(`Already connected: please call .end() to start a new session`);
        if (!audioTrack || audioTrack.kind !== "audio") throw new Error("No audio track provided");
        this.stream = new MediaStream([
            audioTrack
        ]);
        // Firefox workaround: Firefox doesn't support connecting AudioNodes from
        // AudioContexts with different sample rates. So we create the AudioContext
        // at the native sample rate and resample manually.
        // This workaround may be temporary; Firefox is in the process of adding
        // support for this.
        // See: https://bugzilla.mozilla.org/show_bug.cgi?id=1674892
        const isFirefox = navigator.userAgent.toLowerCase().includes("firefox");
        let context;
        if (isFirefox) // Firefox: Use native sample rate and resample manually
        context = new AudioContext();
        else // Other browsers: Let the AudioContext handle resampling
        context = new AudioContext({
            sampleRate: this.sampleRate
        });
        const contextSampleRate = context.sampleRate;
        const source = context.createMediaStreamSource(this.stream);
        // Load and execute the module script.
        try {
            await context.audioWorklet.addModule(this.scriptSrc);
        } catch (e) {
            console.error(e);
            throw new Error(`Could not add audioWorklet module: ${this.scriptSrc}`);
        }
        const processor = new AudioWorkletNode(context, "audio_processor");
        processor.port.onmessage = (e)=>{
            const { event: event, id: id, data: data } = e.data;
            if (event === "receipt") this.eventReceipts[id] = data;
            else if (event === "chunk") {
                // Resample chunk data from native sample rate to target sample rate
                // (no-op if rates match, i.e. for non-Firefox browsers)
                const resampledData = {
                    raw: $7cef7a69bdf8f84d$var$resampleAudioBuffer(data.raw, contextSampleRate, this.sampleRate),
                    mono: $7cef7a69bdf8f84d$var$resampleAudioBuffer(data.mono, contextSampleRate, this.sampleRate)
                };
                if (this._chunkProcessorSize) {
                    const buffer = this._chunkProcessorBuffer;
                    this._chunkProcessorBuffer = {
                        raw: (0, $a61750b8fbee4dae$export$13afda237b1c9846).mergeBuffers(buffer.raw, resampledData.raw),
                        mono: (0, $a61750b8fbee4dae$export$13afda237b1c9846).mergeBuffers(buffer.mono, resampledData.mono)
                    };
                    if (this._chunkProcessorBuffer.mono.byteLength >= this._chunkProcessorSize) {
                        this._chunkProcessor(this._chunkProcessorBuffer);
                        this._chunkProcessorBuffer = {
                            raw: new ArrayBuffer(0),
                            mono: new ArrayBuffer(0)
                        };
                    }
                } else this._chunkProcessor(resampledData);
            }
        };
        const node = source.connect(processor);
        const analyser = context.createAnalyser();
        analyser.fftSize = 8192;
        analyser.smoothingTimeConstant = 0.1;
        node.connect(analyser);
        if (this.outputToSpeakers) {
            // eslint-disable-next-line no-console
            console.warn("Warning: Output to speakers may affect sound quality,\nespecially due to system audio feedback preventative measures.\nuse only for debugging");
            analyser.connect(context.destination);
        }
        this.source = source;
        this.node = node;
        this.analyser = analyser;
        this.processor = processor;
        return true;
    }
    /**
   * Gets the current frequency domain data from the recording track
   * @param {"frequency"|"music"|"voice"} [analysisType]
   * @param {number} [minDecibels] default -100
   * @param {number} [maxDecibels] default -30
   * @returns {import('./analysis/audio_analysis.js').AudioAnalysisOutputType}
   */ getFrequencies(analysisType = "frequency", minDecibels = -100, maxDecibels = -30) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        return (0, $5853841ab58516d4$export$2c3136da0bf130f9).getFrequencies(this.analyser, this.sampleRate, null, analysisType, minDecibels, maxDecibels);
    }
    /**
   * Pauses the recording
   * Keeps microphone stream open but halts storage of audio
   * @returns {Promise<true>}
   */ async pause() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        else if (!this.recording) throw new Error("Already paused: please call .record() first");
        if (this._chunkProcessorBuffer.raw.byteLength) this._chunkProcessor(this._chunkProcessorBuffer);
        this.log("Pausing ...");
        await this._event("stop");
        this.recording = false;
        return true;
    }
    /**
   * Start recording stream and storing to memory from the connected audio source
   * @param {(data: { mono: Int16Array; raw: Int16Array }) => any} [chunkProcessor]
   * @param {number} [chunkSize] chunkProcessor will not be triggered until this size threshold met in mono audio
   * @returns {Promise<true>}
   */ async record(chunkProcessor = ()=>{}, chunkSize = 8192) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        else if (this.recording) throw new Error("Already recording: HELLO please call .pause() first");
        else if (typeof chunkProcessor !== "function") throw new Error(`chunkProcessor must be a function`);
        this._chunkProcessor = chunkProcessor;
        this._chunkProcessorSize = chunkSize;
        this._chunkProcessorBuffer = {
            raw: new ArrayBuffer(0),
            mono: new ArrayBuffer(0)
        };
        this.log("Recording ...");
        await this._event("start");
        this.recording = true;
        return true;
    }
    /**
   * Clears the audio buffer, empties stored recording
   * @returns {Promise<true>}
   */ async clear() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        await this._event("clear");
        return true;
    }
    /**
   * Reads the current audio stream data
   * @returns {Promise<{meanValues: Float32Array, channels: Array<Float32Array>}>}
   */ async read() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        this.log("Reading ...");
        const result = await this._event("read");
        return result;
    }
    /**
   * Saves the current audio stream to a file
   * @param {boolean} [force] Force saving while still recording
   * @returns {Promise<import('./wav_packer.js').WavPackerAudioType>}
   */ async save(force = false) {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        if (!force && this.recording) throw new Error("Currently recording: please call .pause() first, or call .save(true) to force");
        this.log("Exporting ...");
        const exportData = await this._event("export");
        const packer = new (0, $a61750b8fbee4dae$export$13afda237b1c9846)();
        const result = packer.pack(this.sampleRate, exportData.audio);
        return result;
    }
    /**
   * Ends the current recording session and saves the result
   * @returns {Promise<import('./wav_packer.js').WavPackerAudioType>}
   */ async end() {
        if (!this.processor) throw new Error("Session ended: please call .begin() first");
        const _processor = this.processor;
        this.log("Stopping ...");
        await this._event("stop");
        this.recording = false;
        this.log("Exporting ...");
        const exportData = await this._event("export", {}, _processor);
        this.processor.disconnect();
        this.source.disconnect();
        this.node.disconnect();
        this.analyser.disconnect();
        this.stream = null;
        this.processor = null;
        this.source = null;
        this.node = null;
        const packer = new (0, $a61750b8fbee4dae$export$13afda237b1c9846)();
        const result = packer.pack(this.sampleRate, exportData.audio);
        return result;
    }
    /**
   * Performs a full cleanup of WavRecorder instance
   * Stops actively listening via microphone and removes existing listeners
   * @returns {Promise<true>}
   */ async quit() {
        this.listenForDeviceChange(null);
        if (this.processor) await this.end();
        return true;
    }
}
globalThis.WavRecorder = WavRecorder;





var $23859655abfc5f5c$var$__extends = undefined && undefined.__extends || function() {
    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf || ({
            __proto__: []
        }) instanceof Array && function(d, b) {
            d.__proto__ = b;
        } || function(d, b) {
            for(var p in b)if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
        };
        return extendStatics(d, b);
    };
    return function(d, b) {
        if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var $23859655abfc5f5c$var$__awaiter = undefined && undefined.__awaiter || function(thisArg, _arguments, P, generator) {
    function adopt(value) {
        return value instanceof P ? value : new P(function(resolve) {
            resolve(value);
        });
    }
    return new (P || (P = Promise))(function(resolve, reject) {
        function fulfilled(value) {
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }
        function rejected(value) {
            try {
                step(generator["throw"](value));
            } catch (e) {
                reject(e);
            }
        }
        function step(result) {
            result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
        }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var $23859655abfc5f5c$var$__generator = undefined && undefined.__generator || function(thisArg, body) {
    var _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(g && (g = 0, op[0] && (_ = 0)), _)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
};
var $23859655abfc5f5c$export$4a0c46dbbe2ddb67 = /** @class */ function() {
    function MediaManager() {
        this._callbacks = {};
        this._micEnabled = true;
        this._camEnabled = false;
        this._supportsScreenShare = false;
    }
    MediaManager.prototype.setUserAudioCallback = function(userAudioCallback) {
        this._userAudioCallback = userAudioCallback;
    };
    MediaManager.prototype.setClientOptions = function(options, override) {
        var _a, _b, _c;
        if (override === void 0) override = false;
        if (this._options && !override) return;
        this._options = options;
        this._callbacks = (_a = options.callbacks) !== null && _a !== void 0 ? _a : {};
        this._micEnabled = (_b = options.enableMic) !== null && _b !== void 0 ? _b : true;
        this._camEnabled = (_c = options.enableCam) !== null && _c !== void 0 ? _c : false;
    };
    Object.defineProperty(MediaManager.prototype, "supportsScreenShare", {
        get: function() {
            return this._supportsScreenShare;
        },
        enumerable: false,
        configurable: true
    });
    return MediaManager;
}();
var $23859655abfc5f5c$export$45c5b9bfba2f6304 = /** @class */ function(_super) {
    $23859655abfc5f5c$var$__extends(WavMediaManager, _super);
    function WavMediaManager(recorderChunkSize, recorderSampleRate) {
        if (recorderChunkSize === void 0) recorderChunkSize = undefined;
        if (recorderSampleRate === void 0) recorderSampleRate = 24000;
        var _this = _super.call(this) || this;
        _this._initialized = false;
        _this._recorderChunkSize = undefined;
        _this._recorderChunkSize = recorderChunkSize;
        _this._wavRecorder = new (0, $0d471c38435bdab6$export$439b217ca659a877)({
            sampleRate: recorderSampleRate
        });
        _this._wavStreamPlayer = new (0, $1e7ce9484a3c4077$export$9698d62c78b8f366)({
            sampleRate: 24000
        });
        return _this;
    }
    WavMediaManager.prototype.initialize = function() {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            var error_1;
            return $23859655abfc5f5c$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        _a.trys.push([
                            0,
                            2,
                            ,
                            3
                        ]);
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.begin()
                        ];
                    case 1:
                        _a.sent();
                        return [
                            3 /*break*/ ,
                            3
                        ];
                    case 2:
                        error_1 = _a.sent();
                        return [
                            3 /*break*/ ,
                            3
                        ];
                    case 3:
                        this._wavRecorder.listenForDeviceChange(null);
                        this._wavRecorder.listenForDeviceChange(this._handleAvailableDevicesUpdated.bind(this));
                        this._wavRecorder.listenForDeviceErrors(null);
                        this._wavRecorder.listenForDeviceErrors(this._handleDeviceError.bind(this));
                        return [
                            4 /*yield*/ ,
                            this._wavStreamPlayer.connect()
                        ];
                    case 4:
                        _a.sent();
                        this._initialized = true;
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype.connect = function() {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            var isAlreadyRecording;
            return $23859655abfc5f5c$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        if (!!this._initialized) return [
                            3 /*break*/ ,
                            2
                        ];
                        return [
                            4 /*yield*/ ,
                            this.initialize()
                        ];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        isAlreadyRecording = this._wavRecorder.getStatus() == "recording";
                        if (!(this._micEnabled && !isAlreadyRecording)) return [
                            3 /*break*/ ,
                            4
                        ];
                        return [
                            4 /*yield*/ ,
                            this._startRecording()
                        ];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        if (this._camEnabled) console.warn("WavMediaManager does not support video input.");
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype.disconnect = function() {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            return $23859655abfc5f5c$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        if (!this._initialized) return [
                            2 /*return*/ 
                        ];
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.end()
                        ];
                    case 1:
                        _a.sent();
                        return [
                            4 /*yield*/ ,
                            this._wavStreamPlayer.interrupt()
                        ];
                    case 2:
                        _a.sent();
                        this._initialized = false;
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype.userStartedSpeaking = function() {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            return $23859655abfc5f5c$var$__generator(this, function(_a) {
                return [
                    2 /*return*/ ,
                    this._wavStreamPlayer.interrupt()
                ];
            });
        });
    };
    WavMediaManager.prototype.bufferBotAudio = function(data, id) {
        return this._wavStreamPlayer.add16BitPCM(data, id);
    };
    WavMediaManager.prototype.getAllMics = function() {
        return this._wavRecorder.listDevices();
    };
    WavMediaManager.prototype.getAllCams = function() {
        // TODO: Video not supported yet
        return Promise.resolve([]);
    };
    WavMediaManager.prototype.getAllSpeakers = function() {
        // TODO: Implement speaker support
        return Promise.resolve([]);
    };
    WavMediaManager.prototype.updateMic = function(micId) {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            var prevMic, curMic, error_2;
            var _a, _b;
            return $23859655abfc5f5c$var$__generator(this, function(_c) {
                switch(_c.label){
                    case 0:
                        prevMic = this._wavRecorder.deviceSelection;
                        if (!(this._wavRecorder.getStatus() !== "ended")) return [
                            3 /*break*/ ,
                            2
                        ];
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.end()
                        ];
                    case 1:
                        _c.sent();
                        _c.label = 2;
                    case 2:
                        _c.trys.push([
                            2,
                            6,
                            ,
                            7
                        ]);
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.begin(micId)
                        ];
                    case 3:
                        _c.sent();
                        if (!this._micEnabled) return [
                            3 /*break*/ ,
                            5
                        ];
                        return [
                            4 /*yield*/ ,
                            this._startRecording()
                        ];
                    case 4:
                        _c.sent();
                        _c.label = 5;
                    case 5:
                        curMic = this._wavRecorder.deviceSelection;
                        if (curMic && prevMic && prevMic.label !== curMic.label) (_b = (_a = this._callbacks).onMicUpdated) === null || _b === void 0 || _b.call(_a, curMic);
                        return [
                            3 /*break*/ ,
                            7
                        ];
                    case 6:
                        error_2 = _c.sent();
                        return [
                            3 /*break*/ ,
                            7
                        ];
                    case 7:
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype.updateCam = function(camId) {
    // TODO: Video not supported yet
    };
    WavMediaManager.prototype.updateSpeaker = function(speakerId) {
    // TODO: Implement speaker support
    };
    Object.defineProperty(WavMediaManager.prototype, "selectedMic", {
        get: function() {
            var _a;
            return (_a = this._wavRecorder.deviceSelection) !== null && _a !== void 0 ? _a : {};
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WavMediaManager.prototype, "selectedCam", {
        get: function() {
            // TODO: Video not supported yet
            return {};
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WavMediaManager.prototype, "selectedSpeaker", {
        get: function() {
            // TODO: Implement speaker support
            return {};
        },
        enumerable: false,
        configurable: true
    });
    WavMediaManager.prototype.enableMic = function(enable) {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, Promise, function() {
            var _this = this;
            return $23859655abfc5f5c$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        this._micEnabled = enable;
                        if (!this._wavRecorder.stream) return [
                            2 /*return*/ 
                        ];
                        this._wavRecorder.stream.getAudioTracks().forEach(function(track) {
                            var _a, _b;
                            track.enabled = enable;
                            if (!enable) (_b = (_a = _this._callbacks).onTrackStopped) === null || _b === void 0 || _b.call(_a, track, $23859655abfc5f5c$var$localParticipant());
                        });
                        if (!enable) return [
                            3 /*break*/ ,
                            2
                        ];
                        return [
                            4 /*yield*/ ,
                            this._startRecording()
                        ];
                    case 1:
                        _a.sent();
                        return [
                            3 /*break*/ ,
                            4
                        ];
                    case 2:
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.pause()
                        ];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype.enableCam = function(enable) {
        // TODO: Video not supported yet
        console.warn("WavMediaManager does not support video input.");
    };
    WavMediaManager.prototype.enableScreenShare = function(enable) {
        // TODO: Screensharing not supported yet
        console.warn("WavMediaManager does not support screen sharing.");
    };
    Object.defineProperty(WavMediaManager.prototype, "isCamEnabled", {
        get: function() {
            // TODO: Video not supported yet
            return false;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WavMediaManager.prototype, "isMicEnabled", {
        get: function() {
            return this._micEnabled;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WavMediaManager.prototype, "isSharingScreen", {
        get: function() {
            // TODO: Screensharing not supported yet
            return false;
        },
        enumerable: false,
        configurable: true
    });
    WavMediaManager.prototype.tracks = function() {
        var _a;
        var tracks = (_a = this._wavRecorder.stream) === null || _a === void 0 ? void 0 : _a.getTracks()[0];
        return {
            local: tracks ? {
                audio: tracks
            } : {}
        };
    };
    WavMediaManager.prototype._startRecording = function() {
        return $23859655abfc5f5c$var$__awaiter(this, void 0, void 0, function() {
            var track;
            var _this = this;
            var _a, _b, _c;
            return $23859655abfc5f5c$var$__generator(this, function(_d) {
                switch(_d.label){
                    case 0:
                        return [
                            4 /*yield*/ ,
                            this._wavRecorder.record(function(data) {
                                var _a;
                                (_a = _this._userAudioCallback) === null || _a === void 0 || _a.call(_this, data.mono);
                            }, this._recorderChunkSize)
                        ];
                    case 1:
                        _d.sent();
                        track = (_a = this._wavRecorder.stream) === null || _a === void 0 ? void 0 : _a.getAudioTracks()[0];
                        if (track) (_c = (_b = this._callbacks).onTrackStarted) === null || _c === void 0 || _c.call(_b, track, $23859655abfc5f5c$var$localParticipant());
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    WavMediaManager.prototype._handleAvailableDevicesUpdated = function(devices) {
        var _a, _b, _c, _d;
        (_b = (_a = this._callbacks).onAvailableCamsUpdated) === null || _b === void 0 || _b.call(_a, devices.filter(function(d) {
            return d.kind === "videoinput";
        }));
        (_d = (_c = this._callbacks).onAvailableMicsUpdated) === null || _d === void 0 || _d.call(_c, devices.filter(function(d) {
            return d.kind === "audioinput";
        }));
        // if the current device went away or we're using the default and
        // the default changed, reset the mic.
        var defaultDevice = devices.find(function(d) {
            return d.deviceId === "default";
        });
        var currentDevice = this._wavRecorder.deviceSelection;
        if (currentDevice && (!devices.some(function(d) {
            return d.deviceId === currentDevice.deviceId;
        }) || currentDevice.deviceId === "default" && currentDevice.label !== (defaultDevice === null || defaultDevice === void 0 ? void 0 : defaultDevice.label))) this.updateMic("");
    };
    WavMediaManager.prototype._handleDeviceError = function(_a) {
        var _b, _c;
        var devices = _a.devices, type = _a.type, error = _a.error;
        var deviceError = new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, type, error === null || error === void 0 ? void 0 : error.message, error ? {
            sourceError: error
        } : undefined);
        (_c = (_b = this._callbacks).onDeviceError) === null || _c === void 0 || _c.call(_b, deviceError);
    };
    return WavMediaManager;
}($23859655abfc5f5c$export$4a0c46dbbe2ddb67);
var $23859655abfc5f5c$var$localParticipant = function() {
    return {
        id: "local",
        name: "",
        local: true
    };
};






var $1c088932741d88e6$var$__extends = undefined && undefined.__extends || function() {
    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf || ({
            __proto__: []
        }) instanceof Array && function(d, b) {
            d.__proto__ = b;
        } || function(d, b) {
            for(var p in b)if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
        };
        return extendStatics(d, b);
    };
    return function(d, b) {
        if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var $1c088932741d88e6$var$__awaiter = undefined && undefined.__awaiter || function(thisArg, _arguments, P, generator) {
    function adopt(value) {
        return value instanceof P ? value : new P(function(resolve) {
            resolve(value);
        });
    }
    return new (P || (P = Promise))(function(resolve, reject) {
        function fulfilled(value) {
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }
        function rejected(value) {
            try {
                step(generator["throw"](value));
            } catch (e) {
                reject(e);
            }
        }
        function step(result) {
            result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
        }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var $1c088932741d88e6$var$__generator = undefined && undefined.__generator || function(thisArg, body) {
    var _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(g && (g = 0, op[0] && (_ = 0)), _)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
};
var $1c088932741d88e6$export$c95c65abc5f47125 = /** @class */ function(_super) {
    $1c088932741d88e6$var$__extends(DailyMediaManager, _super);
    function DailyMediaManager(enablePlayer, enableRecording, onTrackStartedCallback, onTrackStoppedCallback, recorderChunkSize, recorderSampleRate, playerSampleRate) {
        if (enablePlayer === void 0) enablePlayer = true;
        if (enableRecording === void 0) enableRecording = true;
        if (recorderChunkSize === void 0) recorderChunkSize = undefined;
        if (recorderSampleRate === void 0) recorderSampleRate = 24000;
        if (playerSampleRate === void 0) playerSampleRate = 24000;
        var _a;
        var _this = _super.call(this) || this;
        _this._selectedCam = {};
        _this._selectedMic = {};
        _this._selectedSpeaker = {};
        _this._remoteAudioLevelInterval = null;
        _this._recorderChunkSize = undefined;
        _this._initialized = false;
        _this._connected = false;
        _this._currentAudioTrack = null;
        _this._connectResolve = null;
        _this.onTrackStartedCallback = onTrackStartedCallback;
        _this.onTrackStoppedCallback = onTrackStoppedCallback;
        _this._recorderChunkSize = recorderChunkSize;
        _this._supportsScreenShare = true;
        _this._daily = (_a = (0, ($parcel$interopDefault($7ytRB$dailycodailyjs))).getCallInstance()) !== null && _a !== void 0 ? _a : (0, ($parcel$interopDefault($7ytRB$dailycodailyjs))).createCallObject();
        if (enableRecording) _this._mediaStreamRecorder = new (0, $7cef7a69bdf8f84d$export$2934cf2d25c67a48)({
            sampleRate: recorderSampleRate
        });
        if (enablePlayer) _this._wavStreamPlayer = new (0, $1e7ce9484a3c4077$export$9698d62c78b8f366)({
            sampleRate: playerSampleRate
        });
        _this._daily.on("track-started", _this.handleTrackStarted.bind(_this));
        _this._daily.on("track-stopped", _this.handleTrackStopped.bind(_this));
        _this._daily.on("available-devices-updated", _this._handleAvailableDevicesUpdated.bind(_this));
        _this._daily.on("selected-devices-updated", _this._handleSelectedDevicesUpdated.bind(_this));
        _this._daily.on("camera-error", _this.handleDeviceError.bind(_this));
        _this._daily.on("local-audio-level", _this._handleLocalAudioLevel.bind(_this));
        return _this;
    }
    DailyMediaManager.prototype.initialize = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var infos, devices, cams, mics, speakers;
            var _this = this;
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            return $1c088932741d88e6$var$__generator(this, function(_o) {
                switch(_o.label){
                    case 0:
                        if (this._initialized) {
                            console.warn("DailyMediaManager already initialized");
                            return [
                                2 /*return*/ 
                            ];
                        }
                        return [
                            4 /*yield*/ ,
                            this._daily.startCamera({
                                startVideoOff: !this._camEnabled,
                                startAudioOff: !this._micEnabled,
                                dailyConfig: {
                                    useDevicePreferenceCookies: true
                                }
                            })
                        ];
                    case 1:
                        infos = _o.sent();
                        return [
                            4 /*yield*/ ,
                            this._daily.enumerateDevices()
                        ];
                    case 2:
                        devices = _o.sent().devices;
                        cams = devices.filter(function(d) {
                            return d.kind === "videoinput";
                        });
                        mics = devices.filter(function(d) {
                            return d.kind === "audioinput";
                        });
                        speakers = devices.filter(function(d) {
                            return d.kind === "audiooutput";
                        });
                        (_b = (_a = this._callbacks).onAvailableCamsUpdated) === null || _b === void 0 || _b.call(_a, cams);
                        (_d = (_c = this._callbacks).onAvailableMicsUpdated) === null || _d === void 0 || _d.call(_c, mics);
                        (_f = (_e = this._callbacks).onAvailableSpeakersUpdated) === null || _f === void 0 || _f.call(_e, speakers);
                        this._selectedCam = infos.camera;
                        (_h = (_g = this._callbacks).onCamUpdated) === null || _h === void 0 || _h.call(_g, infos.camera);
                        this._selectedMic = infos.mic;
                        (_k = (_j = this._callbacks).onMicUpdated) === null || _k === void 0 || _k.call(_j, infos.mic);
                        this._selectedSpeaker = infos.speaker;
                        (_m = (_l = this._callbacks).onSpeakerUpdated) === null || _m === void 0 || _m.call(_l, infos.speaker);
                        if (!!this._daily.isLocalAudioLevelObserverRunning()) return [
                            3 /*break*/ ,
                            4
                        ];
                        return [
                            4 /*yield*/ ,
                            this._daily.startLocalAudioLevelObserver(100)
                        ];
                    case 3:
                        _o.sent();
                        _o.label = 4;
                    case 4:
                        if (!this._wavStreamPlayer) return [
                            3 /*break*/ ,
                            6
                        ];
                        return [
                            4 /*yield*/ ,
                            this._wavStreamPlayer.connect()
                        ];
                    case 5:
                        _o.sent();
                        if (!this._remoteAudioLevelInterval) this._remoteAudioLevelInterval = setInterval(function() {
                            var _a;
                            var frequencies = _this._wavStreamPlayer.getFrequencies();
                            var aveVal = 0;
                            if ((_a = frequencies.values) === null || _a === void 0 ? void 0 : _a.length) aveVal = frequencies.values.reduce(function(a, c) {
                                return a + c;
                            }, 0) / frequencies.values.length;
                            _this._handleRemoteAudioLevel(aveVal);
                        }, 100);
                        _o.label = 6;
                    case 6:
                        this._initialized = true;
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.connect = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var _this = this;
            return $1c088932741d88e6$var$__generator(this, function(_a) {
                if (this._connected) {
                    console.warn("DailyMediaManager already connected");
                    return [
                        2 /*return*/ 
                    ];
                }
                this._connected = true;
                if (!this._initialized) return [
                    2 /*return*/ ,
                    new Promise(function(resolve) {
                        (function() {
                            return $1c088932741d88e6$var$__awaiter(_this, void 0, void 0, function() {
                                return $1c088932741d88e6$var$__generator(this, function(_a) {
                                    switch(_a.label){
                                        case 0:
                                            this._connectResolve = resolve;
                                            return [
                                                4 /*yield*/ ,
                                                this.initialize()
                                            ];
                                        case 1:
                                            _a.sent();
                                            return [
                                                2 /*return*/ 
                                            ];
                                    }
                                });
                            });
                        })();
                    })
                ];
                if (this._micEnabled) this._startRecording();
                return [
                    2 /*return*/ 
                ];
            });
        });
    };
    DailyMediaManager.prototype.disconnect = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var _a, _b;
            return $1c088932741d88e6$var$__generator(this, function(_c) {
                switch(_c.label){
                    case 0:
                        if (this._remoteAudioLevelInterval) clearInterval(this._remoteAudioLevelInterval);
                        this._remoteAudioLevelInterval = null;
                        this._daily.leave();
                        this._currentAudioTrack = null;
                        return [
                            4 /*yield*/ ,
                            (_a = this._mediaStreamRecorder) === null || _a === void 0 ? void 0 : _a.end()
                        ];
                    case 1:
                        _c.sent();
                        (_b = this._wavStreamPlayer) === null || _b === void 0 || _b.interrupt();
                        this._initialized = false;
                        this._connected = false;
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.userStartedSpeaking = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var _a;
            return $1c088932741d88e6$var$__generator(this, function(_b) {
                return [
                    2 /*return*/ ,
                    (_a = this._wavStreamPlayer) === null || _a === void 0 ? void 0 : _a.interrupt()
                ];
            });
        });
    };
    DailyMediaManager.prototype.bufferBotAudio = function(data, id) {
        var _a;
        return (_a = this._wavStreamPlayer) === null || _a === void 0 ? void 0 : _a.add16BitPCM(data, id);
    };
    DailyMediaManager.prototype.getAllMics = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var devices;
            return $1c088932741d88e6$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        return [
                            4 /*yield*/ ,
                            this._daily.enumerateDevices()
                        ];
                    case 1:
                        devices = _a.sent().devices;
                        return [
                            2 /*return*/ ,
                            devices.filter(function(device) {
                                return device.kind === "audioinput";
                            })
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.getAllCams = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var devices;
            return $1c088932741d88e6$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        return [
                            4 /*yield*/ ,
                            this._daily.enumerateDevices()
                        ];
                    case 1:
                        devices = _a.sent().devices;
                        return [
                            2 /*return*/ ,
                            devices.filter(function(device) {
                                return device.kind === "videoinput";
                            })
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.getAllSpeakers = function() {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var devices;
            return $1c088932741d88e6$var$__generator(this, function(_a) {
                switch(_a.label){
                    case 0:
                        return [
                            4 /*yield*/ ,
                            this._daily.enumerateDevices()
                        ];
                    case 1:
                        devices = _a.sent().devices;
                        return [
                            2 /*return*/ ,
                            devices.filter(function(device) {
                                return device.kind === "audiooutput";
                            })
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.updateMic = function(micId) {
        var _this = this;
        this._daily.setInputDevicesAsync({
            audioDeviceId: micId
        }).then(function(deviceInfo) {
            _this._selectedMic = deviceInfo.mic;
        });
    };
    DailyMediaManager.prototype.updateCam = function(camId) {
        var _this = this;
        this._daily.setInputDevicesAsync({
            videoDeviceId: camId
        }).then(function(deviceInfo) {
            _this._selectedCam = deviceInfo.camera;
        });
    };
    DailyMediaManager.prototype.updateSpeaker = function(speakerId) {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var dInfo, e_1, sID, speakers, defaultSpeaker_1, defaultSpeakerCp;
            var _this = this;
            var _a, _b, _c, _d;
            return $1c088932741d88e6$var$__generator(this, function(_e) {
                switch(_e.label){
                    case 0:
                        if (!!this._wavStreamPlayer) return [
                            3 /*break*/ ,
                            5
                        ];
                        _e.label = 1;
                    case 1:
                        _e.trys.push([
                            1,
                            3,
                            ,
                            4
                        ]);
                        return [
                            4 /*yield*/ ,
                            this._daily.setOutputDeviceAsync({
                                outputDeviceId: speakerId
                            })
                        ];
                    case 2:
                        dInfo = _e.sent();
                        this._selectedSpeaker = dInfo.speaker;
                        (_b = (_a = this._callbacks).onSpeakerUpdated) === null || _b === void 0 || _b.call(_a, this._selectedSpeaker);
                        return [
                            3 /*break*/ ,
                            4
                        ];
                    case 3:
                        e_1 = _e.sent();
                        console.error("Error setting output device", e_1);
                        return [
                            3 /*break*/ ,
                            4
                        ];
                    case 4:
                        return [
                            2 /*return*/ 
                        ];
                    case 5:
                        if (speakerId !== "default" && this._selectedSpeaker.deviceId === speakerId) return [
                            2 /*return*/ 
                        ];
                        sID = speakerId;
                        if (!(sID === "default")) return [
                            3 /*break*/ ,
                            7
                        ];
                        return [
                            4 /*yield*/ ,
                            this.getAllSpeakers()
                        ];
                    case 6:
                        speakers = _e.sent();
                        defaultSpeaker_1 = speakers.find(function(s) {
                            return s.deviceId === "default";
                        });
                        if (!defaultSpeaker_1) {
                            console.warn("No default speaker found");
                            return [
                                2 /*return*/ 
                            ];
                        }
                        speakers.splice(speakers.indexOf(defaultSpeaker_1), 1);
                        defaultSpeakerCp = speakers.find(function(s) {
                            return defaultSpeaker_1.label.includes(s.label);
                        });
                        sID = (_c = defaultSpeakerCp === null || defaultSpeakerCp === void 0 ? void 0 : defaultSpeakerCp.deviceId) !== null && _c !== void 0 ? _c : speakerId;
                        _e.label = 7;
                    case 7:
                        (_d = this._wavStreamPlayer) === null || _d === void 0 || _d.updateSpeaker(sID).then(function() {
                            var _a, _b;
                            _this._selectedSpeaker = {
                                deviceId: speakerId
                            };
                            (_b = (_a = _this._callbacks).onSpeakerUpdated) === null || _b === void 0 || _b.call(_a, _this._selectedSpeaker);
                        });
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    Object.defineProperty(DailyMediaManager.prototype, "selectedMic", {
        get: function() {
            return this._selectedMic;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DailyMediaManager.prototype, "selectedCam", {
        get: function() {
            return this._selectedCam;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DailyMediaManager.prototype, "selectedSpeaker", {
        get: function() {
            return this._selectedSpeaker;
        },
        enumerable: false,
        configurable: true
    });
    DailyMediaManager.prototype.enableMic = function(enable) {
        return $1c088932741d88e6$var$__awaiter(this, void 0, Promise, function() {
            var _a;
            return $1c088932741d88e6$var$__generator(this, function(_b) {
                this._micEnabled = enable;
                if (!((_a = this._daily.participants()) === null || _a === void 0 ? void 0 : _a.local)) return [
                    2 /*return*/ 
                ];
                this._daily.setLocalAudio(enable);
                if (this._mediaStreamRecorder) {
                    if (enable) {
                        if (this._mediaStreamRecorder.getStatus() === "paused") this._startRecording();
                         // else, we'll record on the track-started event
                    } else if (this._mediaStreamRecorder.getStatus() === "recording") this._mediaStreamRecorder.pause();
                }
                return [
                    2 /*return*/ 
                ];
            });
        });
    };
    DailyMediaManager.prototype.enableCam = function(enable) {
        this._camEnabled = enable;
        this._daily.setLocalVideo(enable);
    };
    DailyMediaManager.prototype.enableScreenShare = function(enable) {
        if (enable) this._daily.startScreenShare();
        else this._daily.stopScreenShare();
    };
    Object.defineProperty(DailyMediaManager.prototype, "isCamEnabled", {
        get: function() {
            return this._daily.localVideo();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DailyMediaManager.prototype, "isMicEnabled", {
        get: function() {
            return this._daily.localAudio();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DailyMediaManager.prototype, "isSharingScreen", {
        get: function() {
            return this._daily.localScreenAudio() || this._daily.localScreenVideo();
        },
        enumerable: false,
        configurable: true
    });
    DailyMediaManager.prototype.tracks = function() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        var participants = this._daily.participants();
        return {
            local: {
                audio: (_c = (_b = (_a = participants === null || participants === void 0 ? void 0 : participants.local) === null || _a === void 0 ? void 0 : _a.tracks) === null || _b === void 0 ? void 0 : _b.audio) === null || _c === void 0 ? void 0 : _c.persistentTrack,
                screenAudio: (_f = (_e = (_d = participants === null || participants === void 0 ? void 0 : participants.local) === null || _d === void 0 ? void 0 : _d.tracks) === null || _e === void 0 ? void 0 : _e.screenAudio) === null || _f === void 0 ? void 0 : _f.persistentTrack,
                screenVideo: (_j = (_h = (_g = participants === null || participants === void 0 ? void 0 : participants.local) === null || _g === void 0 ? void 0 : _g.tracks) === null || _h === void 0 ? void 0 : _h.screenVideo) === null || _j === void 0 ? void 0 : _j.persistentTrack,
                video: (_m = (_l = (_k = participants === null || participants === void 0 ? void 0 : participants.local) === null || _k === void 0 ? void 0 : _k.tracks) === null || _l === void 0 ? void 0 : _l.video) === null || _m === void 0 ? void 0 : _m.persistentTrack
            }
        };
    };
    DailyMediaManager.prototype._startRecording = function() {
        var _this = this;
        if (!this._connected || !this._mediaStreamRecorder) return;
        try {
            this._mediaStreamRecorder.record(function(data) {
                _this._userAudioCallback(data.mono);
            }, this._recorderChunkSize);
        } catch (e) {
            var err = e;
            if (!err.message.includes("Already recording")) console.error("Error starting recording", e);
        }
    };
    DailyMediaManager.prototype._handleAvailableDevicesUpdated = function(event) {
        var _a, _b, _c, _d, _e, _f;
        (_b = (_a = this._callbacks).onAvailableCamsUpdated) === null || _b === void 0 || _b.call(_a, event.availableDevices.filter(function(d) {
            return d.kind === "videoinput";
        }));
        (_d = (_c = this._callbacks).onAvailableMicsUpdated) === null || _d === void 0 || _d.call(_c, event.availableDevices.filter(function(d) {
            return d.kind === "audioinput";
        }));
        (_f = (_e = this._callbacks).onAvailableSpeakersUpdated) === null || _f === void 0 || _f.call(_e, event.availableDevices.filter(function(d) {
            return d.kind === "audiooutput";
        }));
        if (this._selectedSpeaker.deviceId === "default") this.updateSpeaker("default");
    };
    DailyMediaManager.prototype._handleSelectedDevicesUpdated = function(event) {
        var _a, _b, _c, _d, _e, _f;
        if (((_a = this._selectedCam) === null || _a === void 0 ? void 0 : _a.deviceId) !== event.devices.camera) {
            this._selectedCam = event.devices.camera;
            (_c = (_b = this._callbacks).onCamUpdated) === null || _c === void 0 || _c.call(_b, event.devices.camera);
        }
        if (((_d = this._selectedMic) === null || _d === void 0 ? void 0 : _d.deviceId) !== event.devices.mic) {
            this._selectedMic = event.devices.mic;
            (_f = (_e = this._callbacks).onMicUpdated) === null || _f === void 0 || _f.call(_e, event.devices.mic);
        }
    };
    DailyMediaManager.prototype.handleDeviceError = function(ev) {
        var _a, _b;
        var generateDeviceError = function(error) {
            var devices = [];
            switch(error.type){
                case "permissions":
                    error.blockedMedia.forEach(function(d) {
                        devices.push(d === "video" ? "cam" : "mic");
                    });
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, error.type, error.msg, {
                        blockedBy: error.blockedBy
                    });
                case "not-found":
                    error.missingMedia.forEach(function(d) {
                        devices.push(d === "video" ? "cam" : "mic");
                    });
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, error.type, error.msg);
                case "constraints":
                    error.failedMedia.forEach(function(d) {
                        devices.push(d === "video" ? "cam" : "mic");
                    });
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, error.type, error.msg, {
                        reason: error.reason
                    });
                case "cam-in-use":
                    devices.push("cam");
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, "in-use", error.msg);
                case "mic-in-use":
                    devices.push("mic");
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, "in-use", error.msg);
                case "cam-mic-in-use":
                    devices.push("cam");
                    devices.push("mic");
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, "in-use", error.msg);
                case "undefined-mediadevices":
                case "unknown":
                default:
                    devices.push("cam");
                    devices.push("mic");
                    return new (0, $7ytRB$pipecataiclientjs.DeviceError)(devices, error.type, error.msg);
            }
        };
        (_b = (_a = this._callbacks).onDeviceError) === null || _b === void 0 || _b.call(_a, generateDeviceError(ev.error));
    };
    DailyMediaManager.prototype._handleLocalAudioLevel = function(ev) {
        var _a, _b;
        (_b = (_a = this._callbacks).onLocalAudioLevel) === null || _b === void 0 || _b.call(_a, ev.audioLevel);
    };
    DailyMediaManager.prototype._handleRemoteAudioLevel = function(audioLevel) {
        var _a, _b;
        (_b = (_a = this._callbacks).onRemoteAudioLevel) === null || _b === void 0 || _b.call(_a, audioLevel, $1c088932741d88e6$var$botParticipant());
    };
    DailyMediaManager.prototype.handleTrackStarted = function(event) {
        return $1c088932741d88e6$var$__awaiter(this, void 0, void 0, function() {
            var status, _a, e_2, e_3;
            var _b, _c, _d, _e;
            return $1c088932741d88e6$var$__generator(this, function(_f) {
                switch(_f.label){
                    case 0:
                        if (!((_b = event.participant) === null || _b === void 0 ? void 0 : _b.local)) return [
                            2 /*return*/ 
                        ];
                        if (!(event.track.kind === "audio")) return [
                            3 /*break*/ ,
                            15
                        ];
                        if (!this._mediaStreamRecorder) return [
                            3 /*break*/ ,
                            14
                        ];
                        status = this._mediaStreamRecorder.getStatus();
                        _a = status;
                        switch(_a){
                            case "ended":
                                return [
                                    3 /*break*/ ,
                                    1
                                ];
                            case "paused":
                                return [
                                    3 /*break*/ ,
                                    5
                                ];
                            case "recording":
                                return [
                                    3 /*break*/ ,
                                    6
                                ];
                        }
                        return [
                            3 /*break*/ ,
                            6
                        ];
                    case 1:
                        _f.trys.push([
                            1,
                            3,
                            ,
                            4
                        ]);
                        return [
                            4 /*yield*/ ,
                            this._mediaStreamRecorder.begin(event.track)
                        ];
                    case 2:
                        _f.sent();
                        if (this._connected) {
                            this._startRecording();
                            if (this._connectResolve) {
                                this._connectResolve();
                                this._connectResolve = null;
                            }
                        }
                        return [
                            3 /*break*/ ,
                            4
                        ];
                    case 3:
                        e_2 = _f.sent();
                        return [
                            3 /*break*/ ,
                            4
                        ];
                    case 4:
                        return [
                            3 /*break*/ ,
                            14
                        ];
                    case 5:
                        this._startRecording();
                        return [
                            3 /*break*/ ,
                            14
                        ];
                    case 6:
                        if (!(this._currentAudioTrack !== event.track)) return [
                            3 /*break*/ ,
                            12
                        ];
                        return [
                            4 /*yield*/ ,
                            this._mediaStreamRecorder.end()
                        ];
                    case 7:
                        _f.sent();
                        _f.label = 8;
                    case 8:
                        _f.trys.push([
                            8,
                            10,
                            ,
                            11
                        ]);
                        return [
                            4 /*yield*/ ,
                            this._mediaStreamRecorder.begin(event.track)
                        ];
                    case 9:
                        _f.sent();
                        this._startRecording();
                        return [
                            3 /*break*/ ,
                            11
                        ];
                    case 10:
                        e_3 = _f.sent();
                        return [
                            3 /*break*/ ,
                            11
                        ];
                    case 11:
                        return [
                            3 /*break*/ ,
                            13
                        ];
                    case 12:
                        console.warn("track-started event received for current track and already recording");
                        _f.label = 13;
                    case 13:
                        return [
                            3 /*break*/ ,
                            14
                        ];
                    case 14:
                        this._currentAudioTrack = event.track;
                        _f.label = 15;
                    case 15:
                        (_d = (_c = this._callbacks).onTrackStarted) === null || _d === void 0 || _d.call(_c, event.track, event.participant ? $1c088932741d88e6$var$dailyParticipantToParticipant(event.participant) : undefined);
                        (_e = this.onTrackStartedCallback) === null || _e === void 0 || _e.call(this, event);
                        return [
                            2 /*return*/ 
                        ];
                }
            });
        });
    };
    DailyMediaManager.prototype.handleTrackStopped = function(event) {
        var _a, _b, _c, _d;
        if (!((_a = event.participant) === null || _a === void 0 ? void 0 : _a.local)) return;
        if (event.track.kind === "audio") {
            if (this._mediaStreamRecorder && this._mediaStreamRecorder.getStatus() === "recording") this._mediaStreamRecorder.pause();
        }
        (_c = (_b = this._callbacks).onTrackStopped) === null || _c === void 0 || _c.call(_b, event.track, event.participant ? $1c088932741d88e6$var$dailyParticipantToParticipant(event.participant) : undefined);
        (_d = this.onTrackStoppedCallback) === null || _d === void 0 || _d.call(this, event);
    };
    return DailyMediaManager;
}((0, $23859655abfc5f5c$export$4a0c46dbbe2ddb67));
var $1c088932741d88e6$var$dailyParticipantToParticipant = function(p) {
    return {
        id: p.user_id,
        local: p.local,
        name: p.user_name
    };
};
var $1c088932741d88e6$var$botParticipant = function() {
    return {
        id: "bot",
        local: false,
        name: "Bot"
    };
};


var $8381b73e1f76ead9$exports = {};

$parcel$export($8381b73e1f76ead9$exports, "SmallWebRTCTransport", () => $8381b73e1f76ead9$export$62043589d053a879);



class $8381b73e1f76ead9$var$TrackStatusMessage {
    constructor(receiver_index, enabled){
        this.type = "trackStatus";
        this.receiver_index = receiver_index;
        this.enabled = enabled;
    }
}
class $8381b73e1f76ead9$var$WebRTCTrack {
    constructor(track){
        this.track = track;
        this.status = "new";
    }
}
const $8381b73e1f76ead9$var$RENEGOTIATE_TYPE = "renegotiate";
class $8381b73e1f76ead9$var$RenegotiateMessage {
    constructor(){
        this.type = $8381b73e1f76ead9$var$RENEGOTIATE_TYPE;
    }
}
const $8381b73e1f76ead9$var$PEER_LEFT_TYPE = "peerLeft";
class $8381b73e1f76ead9$var$PeerLeftMessageMessage {
    constructor(){
        this.type = $8381b73e1f76ead9$var$PEER_LEFT_TYPE;
    }
}
// Interface for the structure of the signalling message
const $8381b73e1f76ead9$var$SIGNALLING_TYPE = "signalling";
class $8381b73e1f76ead9$var$SignallingMessageObject {
    constructor(message){
        this.type = $8381b73e1f76ead9$var$SIGNALLING_TYPE;
        this.message = message;
    }
}
const $8381b73e1f76ead9$var$AUDIO_TRANSCEIVER_INDEX = 0;
const $8381b73e1f76ead9$var$VIDEO_TRANSCEIVER_INDEX = 1;
const $8381b73e1f76ead9$var$SCREEN_VIDEO_TRANSCEIVER_INDEX = 2;
class $8381b73e1f76ead9$export$62043589d053a879 extends (0, $7ytRB$pipecataiclientjs.Transport) {
    constructor(opts = {}){
        super();
        this._webrtcRequest = null;
        // Trigger when the peer connection is finally ready or in case it has failed all the attempts to connect
        this._connectResolved = null;
        this._connectFailed = null;
        this.pc = null;
        this.dc = null;
        this.audioCodec = null;
        this.videoCodec = null;
        this.pc_id = null;
        this.offerUrlTemplate = null;
        this.reconnectionAttempts = 0;
        this.maxReconnectionAttempts = 3;
        this.isReconnecting = false;
        this.keepAliveInterval = null;
        this._iceServers = [];
        this._incomingTracks = new Map();
        this._canSendIceCandidates = false;
        this._candidateQueue = [];
        this.__flushTimeout = null;
        this._flushDelay = 200;
        this._iceServers = opts.iceServers ?? [];
        this._waitForICEGathering = opts.waitForICEGathering ?? false;
        this.audioCodec = opts.audioCodec ?? null;
        this.videoCodec = opts.videoCodec ?? null;
        this.offerUrlTemplate = opts.offerUrlTemplate ?? null;
        this._webrtcRequest = this._resolveRequestInfo(opts);
        this.mediaManager = opts.mediaManager || new (0, $1c088932741d88e6$export$c95c65abc5f47125)(false, false, async (event)=>{
            if (!this.pc) return;
            if (event.type == "audio") {
                (0, $7ytRB$pipecataiclientjs.logger).info("SmallWebRTCMediaManager replacing audio track");
                await this.getAudioTransceiver().sender.replaceTrack(event.track);
            } else if (event.type == "video") {
                (0, $7ytRB$pipecataiclientjs.logger).info("SmallWebRTCMediaManager replacing video track");
                await this.getVideoTransceiver().sender.replaceTrack(event.track);
            } else if (event.type == "screenVideo") {
                (0, $7ytRB$pipecataiclientjs.logger).info("SmallWebRTCMediaManager replacing screen video track");
                await this.getScreenVideoTransceiver().sender.replaceTrack(event.track);
            } else if (event.type == "screenAudio") (0, $7ytRB$pipecataiclientjs.logger).info("SmallWebRTCMediaManager does not yet support screen audio. Track is ignored.");
        }, (event)=>(0, $7ytRB$pipecataiclientjs.logger).debug("SmallWebRTCMediaManager Track stopped:", event));
    }
    initialize(options, messageHandler) {
        this._options = options;
        this._callbacks = options.callbacks ?? {};
        this._onMessage = messageHandler;
        this.mediaManager.setClientOptions(options);
        this.state = "disconnected";
        (0, $7ytRB$pipecataiclientjs.logger).debug("[RTVI Transport] Initialized");
    }
    async initDevices() {
        this.state = "initializing";
        await this.mediaManager.initialize();
        this.state = "initialized";
    }
    setAudioCodec(audioCodec) {
        this.audioCodec = audioCodec;
    }
    setVideoCodec(videoCodec) {
        this.videoCodec = videoCodec;
    }
    _resolveRequestInfo(params) {
        let requestInfo = null;
        const _webrtcUrl = params.webrtcUrl ?? params.connectionUrl ?? null;
        if (_webrtcUrl) {
            const key = params.webrtcUrl ? "webrtcUrl" : "connectionUrl";
            (0, $7ytRB$pipecataiclientjs.logger).warn(`${key} is deprecated. Use webrtcRequestParams instead.`);
            if (params.webrtcRequestParams) (0, $7ytRB$pipecataiclientjs.logger).warn(`Both ${key} and webrtcRequestParams provided. Using webrtcRequestParams.`);
            else if (typeof _webrtcUrl === "string") requestInfo = {
                endpoint: _webrtcUrl
            };
            else (0, $7ytRB$pipecataiclientjs.logger).error(`Invalid ${key} provided in params. Ignoring.`);
        }
        if (params.webrtcRequestParams) {
            if ((0, $7ytRB$pipecataiclientjs.isAPIRequest)(params.webrtcRequestParams)) // Override any previous request set in the constructor, do not try to merge
            requestInfo = params.webrtcRequestParams;
            else (0, $7ytRB$pipecataiclientjs.logger).error(`Invalid webrtcRequestParams provided in params. Ignoring.`);
        }
        return requestInfo ?? this._webrtcRequest;
    }
    _getStartEndpointAsString() {
        const startEndpoint = this.startBotParams?.endpoint;
        switch(typeof startEndpoint){
            case "string":
                return startEndpoint;
            case "object":
                if (startEndpoint instanceof URL) return startEndpoint.toString();
                if (typeof Request !== "undefined" && startEndpoint instanceof Request) return startEndpoint.url;
        }
        return;
    }
    _isValidObject(value) {
        if (value === null || value === undefined) return false;
        if (typeof value !== "object") throw new (0, $7ytRB$pipecataiclientjs.RTVIError)("Invalid connection parameters");
        return true;
    }
    _fixConnectionOptionsParams(params, supportedKeys) {
        const snakeToCamel = (snakeCaseString)=>{
            return snakeCaseString.replace(/_([a-z,A-Z])/g, (_, letter)=>letter.toUpperCase());
        };
        let result = {};
        let sessionId;
        for (const [key, val] of Object.entries(params)){
            const camelKey = snakeToCamel(key);
            if (camelKey === "sessionId") {
                sessionId = val;
                continue;
            }
            if (!supportedKeys.includes(camelKey)) {
                (0, $7ytRB$pipecataiclientjs.logger).warn(`Unrecognized connection parameter: ${key}. Ignored.`);
                continue;
            }
            result[camelKey] = val;
        }
        if (sessionId && this._shouldUseStartBotFallback(result)) result.webrtcRequestParams = this._buildRequestParamsBasedOnStartBotParams(sessionId);
        return result;
    }
    _shouldUseStartBotFallback(options) {
        const hasStartEndpoint = !!this._getStartEndpointAsString();
        const hasNoConnectionParams = !options.webrtcUrl && !options.connectionUrl && !options.webrtcRequestParams;
        return hasStartEndpoint && hasNoConnectionParams;
    }
    _buildRequestParamsBasedOnStartBotParams(sessionId) {
        const startEndpoint = this._getStartEndpointAsString();
        const offerUrl = this.offerUrlTemplate ? this.offerUrlTemplate.replace(":sessionId", sessionId) : startEndpoint.replace("/start", `/sessions/${sessionId}/api/offer`);
        if (typeof Request !== "undefined" && this.startBotParams.endpoint instanceof Request) return {
            endpoint: new Request(offerUrl, this.startBotParams?.endpoint)
        };
        return {
            endpoint: offerUrl,
            headers: this.startBotParams.headers
        };
    }
    _validateConnectionParams(connectParams) {
        if (!this._isValidObject(connectParams)) return undefined;
        const params = connectParams;
        const supportedKeys = [
            "webrtcUrl",
            "connectionUrl",
            "webrtcRequestParams",
            "iceConfig"
        ];
        const fixedParams = this._fixConnectionOptionsParams(params, supportedKeys);
        const webrtcRequestParams = this._resolveRequestInfo(fixedParams);
        if (webrtcRequestParams) fixedParams.webrtcRequestParams = webrtcRequestParams;
        delete fixedParams.connectionUrl;
        delete fixedParams.webrtcUrl;
        if (Object.keys(fixedParams).length === 0) return undefined;
        return fixedParams;
    }
    async _connect(connectParams) {
        if (this._abortController?.signal.aborted) return;
        this.state = "connecting";
        if (connectParams?.iceConfig?.iceServers) this._iceServers = connectParams?.iceConfig?.iceServers;
        // Note: There is no need to validate the params here, as they were already
        //       validated and fixed in the parent class's connect() method (which calls
        //       _validateConnectionParams() and passes the result to _connect()).
        this._webrtcRequest = connectParams?.webrtcRequestParams ?? this._webrtcRequest;
        if (!this._webrtcRequest) {
            (0, $7ytRB$pipecataiclientjs.logger).error("No request details provided for WebRTC connection");
            this.state = "error";
            throw new (0, $7ytRB$pipecataiclientjs.TransportStartError)();
        }
        await this.mediaManager.connect();
        await this.startNewPeerConnection();
        if (this._abortController?.signal.aborted) return;
        if (this.dc?.readyState !== "open") // Wait until we are actually connected and the data channel is ready
        await new Promise((resolve, reject)=>{
            this._connectResolved = resolve;
            this._connectFailed = reject;
        });
        this.state = "connected";
        this._callbacks.onConnected?.();
        this._callbacks.onBotConnected?.($8381b73e1f76ead9$var$botParticipant(this.pc_id));
    }
    syncTrackStatus() {
        // Sending the current status from the tracks to Pipecat
        this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$AUDIO_TRANSCEIVER_INDEX, this.mediaManager.isMicEnabled));
        this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$VIDEO_TRANSCEIVER_INDEX, this.mediaManager.isCamEnabled));
        if (this.mediaManager.supportsScreenShare) this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$SCREEN_VIDEO_TRANSCEIVER_INDEX, this.mediaManager.isSharingScreen && !!this.mediaManager.tracks().local.screenVideo));
    }
    sendReadyMessage() {
        this.state = "ready";
        // Sending message that the client is ready, just for testing
        //this.dc?.send(JSON.stringify({id: 'clientReady', label: 'rtvi-ai', type:'client-ready'}))
        this.sendMessage((0, $7ytRB$pipecataiclientjs.RTVIMessage).clientReady());
    }
    sendMessage(message) {
        if (!this.dc || this.dc.readyState !== "open") {
            (0, $7ytRB$pipecataiclientjs.logger).warn(`Datachannel is not ready. Message not sent: ${message}`);
            return;
        }
        // Note: This shouldn't happen since client-js should have already checked this
        if (!(0, $7ytRB$pipecataiclientjs.messageSizeWithinLimit)(message, this._maxMessageSize)) throw new (0, $7ytRB$pipecataiclientjs.MessageTooLargeError)("Message data too large. Max size is " + this._maxMessageSize);
        this.dc?.send(JSON.stringify(message));
    }
    sendSignallingMessage(message) {
        if (!this.dc || this.dc.readyState !== "open") {
            (0, $7ytRB$pipecataiclientjs.logger).warn(`Datachannel is not ready. Message not sent: ${message}`);
            return;
        }
        const signallingMessage = new $8381b73e1f76ead9$var$SignallingMessageObject(message);
        this.dc?.send(JSON.stringify(signallingMessage));
    }
    async _disconnect() {
        this.state = "disconnecting";
        await this.stop();
        this.state = "disconnected";
    }
    createPeerConnection() {
        const config = {
            iceServers: this._iceServers
        };
        let pc = new RTCPeerConnection(config);
        pc.onicecandidate = async (event)=>{
            if (event.candidate) {
                (0, $7ytRB$pipecataiclientjs.logger).debug("New ICE candidate:", event.candidate);
                await this.sendIceCandidate(event.candidate);
            } else (0, $7ytRB$pipecataiclientjs.logger).info("All ICE candidates have been sent.");
        };
        pc.addEventListener("icegatheringstatechange", ()=>{
            if (pc.iceGatheringState === "complete" && pc.iceConnectionState === "checking" && this._waitForICEGathering) {
                (0, $7ytRB$pipecataiclientjs.logger).info("Ice gathering completed and connection is still checking. Trying to reconnect.");
                // If ICE gathering has completed and the previous connection was still in the "checking" state,
                // we will reconnect to use all the new ICE candidates.
                this.attemptReconnection(false);
            }
        });
        pc.addEventListener("iceconnectionstatechange", ()=>this.handleICEConnectionStateChange());
        (0, $7ytRB$pipecataiclientjs.logger).debug(`iceConnectionState: ${pc.iceConnectionState}`);
        pc.addEventListener("signalingstatechange", ()=>{
            (0, $7ytRB$pipecataiclientjs.logger).debug(`signalingState: ${this.pc.signalingState}`);
            if (this.pc.signalingState == "stable") this.handleReconnectionCompleted();
        });
        (0, $7ytRB$pipecataiclientjs.logger).debug(`signalingState: ${pc.signalingState}`);
        pc.addEventListener("track", (evt)=>{
            const streamType = evt.transceiver ? evt.transceiver.mid === "0" ? "microphone" : evt.transceiver.mid === "1" ? "camera" : "screenVideo" : null;
            if (!streamType) {
                (0, $7ytRB$pipecataiclientjs.logger).warn("Received track without transceiver mid", evt);
                return;
            }
            (0, $7ytRB$pipecataiclientjs.logger).debug(`Received new remote track for ${streamType}`);
            this._incomingTracks.set(streamType, new $8381b73e1f76ead9$var$WebRTCTrack(evt.track));
            evt.track.addEventListener("unmute", ()=>{
                const t = this._incomingTracks.get(streamType);
                if (!t) return;
                (0, $7ytRB$pipecataiclientjs.logger).debug(`Remote track unmuted: ${streamType}`);
                t.status = "unmuted";
                this._callbacks.onTrackStarted?.(evt.track);
            });
            evt.track.addEventListener("mute", ()=>{
                const t = this._incomingTracks.get(streamType);
                if (!t || t.status !== "unmuted") return;
                (0, $7ytRB$pipecataiclientjs.logger).debug(`Remote track muted: ${streamType}`);
                t.status = "muted";
                this._callbacks.onTrackStopped?.(evt.track);
            });
            evt.track.addEventListener("ended", ()=>{
                (0, $7ytRB$pipecataiclientjs.logger).debug(`Remote track ended: ${streamType}`);
                this._callbacks.onTrackStopped?.(evt.track);
                this._incomingTracks.delete(streamType);
            });
        });
        return pc;
    }
    handleICEConnectionStateChange() {
        if (!this.pc) return;
        (0, $7ytRB$pipecataiclientjs.logger).debug(`ICE Connection State: ${this.pc.iceConnectionState}`);
        if (this.pc.iceConnectionState === "failed") {
            (0, $7ytRB$pipecataiclientjs.logger).debug("ICE connection failed, attempting restart.");
            this.attemptReconnection(true);
        } else if (this.pc.iceConnectionState === "disconnected") // Waiting before trying to reconnect to see if it handles it automatically
        setTimeout(()=>{
            if (this.pc?.iceConnectionState === "disconnected") {
                (0, $7ytRB$pipecataiclientjs.logger).debug("Still disconnected, attempting reconnection.");
                this.attemptReconnection(true);
            }
        }, 5000);
    }
    handleReconnectionCompleted() {
        this.reconnectionAttempts = 0;
        this.isReconnecting = false;
    }
    async attemptReconnection(recreatePeerConnection = false) {
        if (this.isReconnecting) {
            (0, $7ytRB$pipecataiclientjs.logger).debug("Reconnection already in progress, skipping.");
            return;
        }
        if (this.reconnectionAttempts >= this.maxReconnectionAttempts) {
            (0, $7ytRB$pipecataiclientjs.logger).debug("Max reconnection attempts reached. Stopping transport.");
            await this.stop();
            return;
        }
        this.isReconnecting = true;
        this.reconnectionAttempts++;
        (0, $7ytRB$pipecataiclientjs.logger).debug(`Reconnection attempt ${this.reconnectionAttempts}...`);
        // aiortc does not seem to work when just trying to restart the ice
        // so for this case we create a new peer connection on both sides
        if (recreatePeerConnection) {
            const oldPC = this.pc;
            await this.startNewPeerConnection(recreatePeerConnection);
            if (oldPC) {
                (0, $7ytRB$pipecataiclientjs.logger).debug("closing old peer connection");
                this.closePeerConnection(oldPC);
            }
        } else await this.negotiate();
    }
    async waitForIceGatheringComplete(timeoutMs = 2000) {
        const pc = this.pc;
        if (pc.iceGatheringState === "complete") return;
        (0, $7ytRB$pipecataiclientjs.logger).info("Waiting for ICE gathering to complete. Current state:", pc.iceGatheringState);
        return new Promise((resolve)=>{
            let timeoutId;
            const cleanup = ()=>{
                pc.removeEventListener("icegatheringstatechange", checkState);
                clearTimeout(timeoutId);
            };
            const checkState = ()=>{
                (0, $7ytRB$pipecataiclientjs.logger).debug("icegatheringstatechange:", pc.iceGatheringState);
                if (pc.iceGatheringState === "complete") {
                    cleanup();
                    resolve();
                }
            };
            const onTimeout = ()=>{
                (0, $7ytRB$pipecataiclientjs.logger).debug(`ICE gathering timed out after ${timeoutMs} ms.`);
                cleanup();
                resolve();
            };
            pc.addEventListener("icegatheringstatechange", checkState);
            timeoutId = setTimeout(onTimeout, timeoutMs);
            // Checking the state again to avoid race conditions
            checkState();
        });
    }
    async sendIceCandidate(candidate) {
        if (!this._webrtcRequest) {
            (0, $7ytRB$pipecataiclientjs.logger).error("No request details provided for WebRTC connection");
            return;
        }
        this._candidateQueue.push(candidate);
        // We are sending all the ice candidates each 200ms
        if (!this.__flushTimeout) this.__flushTimeout = setTimeout(()=>this.flushIceCandidates(), this._flushDelay);
    }
    async flushIceCandidates() {
        this.__flushTimeout = null;
        if (!this._webrtcRequest || this._candidateQueue.length === 0 || !this._canSendIceCandidates) return;
        // Drain queue
        const candidates = this._candidateQueue.splice(0, this._candidateQueue.length);
        let headers;
        try {
            if (typeof Request !== "undefined" && this._webrtcRequest.endpoint instanceof Request) {
                console.log("Using Request object headers");
                headers = this._webrtcRequest.endpoint.headers;
            } else headers = new Headers({
                "Content-Type": "application/json",
                ...Object.fromEntries((this._webrtcRequest.headers ?? new Headers()).entries())
            });
            const payload = {
                pc_id: this.pc_id,
                candidates: candidates.map((c)=>({
                        candidate: c.candidate,
                        sdp_mid: c.sdpMid,
                        sdp_mline_index: c.sdpMLineIndex
                    }))
            };
            await fetch(this._webrtcRequest.endpoint, {
                method: "PATCH",
                headers: headers,
                body: JSON.stringify(payload)
            });
        } catch (e) {
            (0, $7ytRB$pipecataiclientjs.logger).error(`Failed to send ICE candidate: ${e}`);
        }
    }
    async negotiate(recreatePeerConnection = false) {
        if (!this.pc) return Promise.reject("Peer connection is not initialized");
        if (!this._webrtcRequest) {
            (0, $7ytRB$pipecataiclientjs.logger).error("No request details provided for WebRTC connection");
            this.state = "error";
            throw new (0, $7ytRB$pipecataiclientjs.TransportStartError)();
        }
        try {
            // Create offer
            const offer = await this.pc.createOffer();
            await this.pc.setLocalDescription(offer);
            // Wait for ICE gathering to complete
            if (this._waitForICEGathering) await this.waitForIceGatheringComplete();
            let offerSdp = this.pc.localDescription;
            // Filter audio codec
            if (this.audioCodec && this.audioCodec !== "default") // @ts-ignore
            offerSdp.sdp = this.sdpFilterCodec("audio", this.audioCodec, offerSdp.sdp);
            // Filter video codec
            if (this.videoCodec && this.videoCodec !== "default") // @ts-ignore
            offerSdp.sdp = this.sdpFilterCodec("video", this.videoCodec, offerSdp.sdp);
            (0, $7ytRB$pipecataiclientjs.logger).debug(`Will create offer for peerId: ${this.pc_id}`);
            // Send offer to server
            const requestData = {
                sdp: offerSdp.sdp,
                type: offerSdp.type,
                pc_id: this.pc_id,
                restart_pc: recreatePeerConnection
            };
            let request;
            if (typeof Request !== "undefined" && this._webrtcRequest.endpoint instanceof Request) request = {
                endpoint: new Request(this._webrtcRequest.endpoint, {
                    body: JSON.stringify(requestData)
                })
            };
            else {
                request = (0, ($parcel$interopDefault($7ytRB$lodashcloneDeep)))(this._webrtcRequest);
                if (this._webrtcRequest.requestData) requestData.requestData = this._webrtcRequest.requestData;
                request.requestData = requestData;
            }
            const answer = await (0, $7ytRB$pipecataiclientjs.makeRequest)(request);
            // @ts-ignore
            this.pc_id = answer.pc_id;
            // @ts-ignore
            (0, $7ytRB$pipecataiclientjs.logger).debug(`Received answer for peer connection id ${answer.pc_id}`);
            await this.pc.setRemoteDescription(answer);
        } catch (e) {
            (0, $7ytRB$pipecataiclientjs.logger).debug(`Reconnection attempt ${this.reconnectionAttempts} failed: ${e}`);
            this.isReconnecting = false;
            setTimeout(()=>this.attemptReconnection(true), 2000);
        }
    }
    addInitialTransceivers() {
        // Transceivers always appear in creation-order for both peers
        // For now we support 3 transceivers meant to hold the following
        // tracks in the given order:
        // audio, video, screenVideo
        this.pc.addTransceiver("audio", {
            direction: "sendrecv"
        });
        this.pc.addTransceiver("video", {
            direction: "sendrecv"
        });
        if (this.mediaManager.supportsScreenShare) // For now, we only support receiving a single video track
        this.pc.addTransceiver("video", {
            direction: "sendonly"
        });
    }
    getAudioTransceiver() {
        // Transceivers always appear in creation-order for both peers
        // Look at addInitialTransceivers
        return this.pc.getTransceivers()[$8381b73e1f76ead9$var$AUDIO_TRANSCEIVER_INDEX];
    }
    getVideoTransceiver() {
        // Transceivers always appear in creation-order for both peers
        // Look at addInitialTransceivers
        return this.pc.getTransceivers()[$8381b73e1f76ead9$var$VIDEO_TRANSCEIVER_INDEX];
    }
    getScreenVideoTransceiver() {
        // Transceivers always appear in creation-order for both peers
        // Look at addInitialTransceivers
        return this.pc.getTransceivers()[$8381b73e1f76ead9$var$SCREEN_VIDEO_TRANSCEIVER_INDEX];
    }
    async startNewPeerConnection(recreatePeerConnection = false) {
        this.pc = this.createPeerConnection();
        this.addInitialTransceivers();
        this.dc = this.createDataChannel("chat", {
            ordered: true
        });
        await this.addUserMedia();
        await this.negotiate(recreatePeerConnection);
        // Sending the ice candidates
        this._canSendIceCandidates = true;
        await this.flushIceCandidates();
    }
    async addUserMedia() {
        (0, $7ytRB$pipecataiclientjs.logger).debug(`addUserMedia this.tracks(): ${this.tracks()}`);
        let audioTrack = this.tracks().local.audio;
        (0, $7ytRB$pipecataiclientjs.logger).debug(`addUserMedia audioTrack: ${audioTrack}`);
        if (audioTrack) await this.getAudioTransceiver().sender.replaceTrack(audioTrack);
        let videoTrack = this.tracks().local.video;
        (0, $7ytRB$pipecataiclientjs.logger).debug(`addUserMedia videoTrack: ${videoTrack}`);
        if (videoTrack) await this.getVideoTransceiver().sender.replaceTrack(videoTrack);
        if (this.mediaManager.supportsScreenShare) {
            videoTrack = this.tracks().local.screenVideo;
            (0, $7ytRB$pipecataiclientjs.logger).debug(`addUserMedia screenVideoTrack: ${videoTrack}`);
            if (videoTrack) await this.getScreenVideoTransceiver().sender.replaceTrack(videoTrack);
        }
    }
    // Method to handle a general message (this can be expanded for other types of messages)
    handleMessage(message) {
        try {
            const messageObj = JSON.parse(message); // Type is `any` initially
            (0, $7ytRB$pipecataiclientjs.logger).debug("received message:", messageObj);
            // Check if it's a signalling message
            if (messageObj.type === $8381b73e1f76ead9$var$SIGNALLING_TYPE) this.handleSignallingMessage(messageObj); // Delegate to handleSignallingMessage
            else // Bubble any messages with rtvi-ai label
            if (messageObj.label === "rtvi-ai") this._onMessage({
                id: messageObj.id,
                type: messageObj.type,
                data: messageObj.data
            });
        } catch (error) {
            (0, $7ytRB$pipecataiclientjs.logger).error("Failed to parse JSON message:", error);
        }
    }
    // Method to handle signalling messages specifically
    async handleSignallingMessage(messageObj) {
        // Cast the object to the correct type after verification
        const signallingMessage = messageObj;
        // Handle different signalling message types
        switch(signallingMessage.message.type){
            case $8381b73e1f76ead9$var$RENEGOTIATE_TYPE:
                this.attemptReconnection(false);
                break;
            case $8381b73e1f76ead9$var$PEER_LEFT_TYPE:
                this._callbacks.onBotDisconnected?.($8381b73e1f76ead9$var$botParticipant(this.pc_id));
                break;
            default:
                (0, $7ytRB$pipecataiclientjs.logger).warn("Unknown signalling message:", signallingMessage.message);
        }
    }
    createDataChannel(label, options) {
        const dc = this.pc.createDataChannel(label, options);
        dc.addEventListener("close", ()=>{
            (0, $7ytRB$pipecataiclientjs.logger).debug("datachannel closed");
            if (this.keepAliveInterval) {
                clearInterval(this.keepAliveInterval);
                this.keepAliveInterval = null;
            }
        });
        dc.addEventListener("open", ()=>{
            (0, $7ytRB$pipecataiclientjs.logger).debug("datachannel opened");
            this._maxMessageSize = this.pc?.sctp?.maxMessageSize ?? 65536;
            if (this._connectResolved) {
                this.syncTrackStatus();
                this._connectResolved();
                this._connectResolved = null;
                this._connectFailed = null;
            }
            // @ts-ignore
            this.keepAliveInterval = setInterval(()=>{
                const message = "ping: " + new Date().getTime();
                dc.send(message);
            }, 1000);
        });
        dc.addEventListener("message", (evt)=>{
            let message = evt.data;
            this.handleMessage(message);
        });
        return dc;
    }
    closePeerConnection(pc) {
        pc.getTransceivers().forEach((transceiver)=>{
            if (transceiver.stop) transceiver.stop();
        });
        pc.getSenders().forEach((sender)=>{
            sender.track?.stop();
        });
        pc.close();
    }
    async stop() {
        if (!this.pc) {
            (0, $7ytRB$pipecataiclientjs.logger).debug("Peer connection is already closed or null.");
            return;
        }
        if (this.dc) this.dc.close();
        this.closePeerConnection(this.pc);
        this.pc = null;
        await this.mediaManager.disconnect();
        // For some reason after we close the peer connection, it is not triggering the listeners
        this.pc_id = null;
        this.reconnectionAttempts = 0;
        this.isReconnecting = false;
        this._callbacks.onDisconnected?.();
        this._candidateQueue = [];
        this._canSendIceCandidates = false;
        if (this._connectFailed) this._connectFailed();
        this._connectFailed = null;
        this._connectResolved = null;
    }
    getAllMics() {
        return this.mediaManager.getAllMics();
    }
    getAllCams() {
        return this.mediaManager.getAllCams();
    }
    getAllSpeakers() {
        return this.mediaManager.getAllSpeakers();
    }
    async updateMic(micId) {
        return this.mediaManager.updateMic(micId);
    }
    updateCam(camId) {
        return this.mediaManager.updateCam(camId);
    }
    updateSpeaker(speakerId) {
        return this.mediaManager.updateSpeaker(speakerId);
    }
    get selectedMic() {
        return this.mediaManager.selectedMic;
    }
    get selectedCam() {
        return this.mediaManager.selectedCam;
    }
    get selectedSpeaker() {
        return this.mediaManager.selectedSpeaker;
    }
    set iceServers(iceServers) {
        this._iceServers = iceServers;
    }
    get iceServers() {
        return this._iceServers;
    }
    enableMic(enable) {
        this.mediaManager.enableMic(enable);
        this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$AUDIO_TRANSCEIVER_INDEX, enable));
    }
    enableCam(enable) {
        this.mediaManager.enableCam(enable);
        this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$VIDEO_TRANSCEIVER_INDEX, enable));
    }
    async enableScreenShare(enable) {
        if (!this.mediaManager.supportsScreenShare) throw new (0, $7ytRB$pipecataiclientjs.UnsupportedFeatureError)("enableScreenShare", "mediaManager", "Screen sharing is not supported by the current media manager");
        this.mediaManager.enableScreenShare(enable);
        this.sendSignallingMessage(new $8381b73e1f76ead9$var$TrackStatusMessage($8381b73e1f76ead9$var$SCREEN_VIDEO_TRANSCEIVER_INDEX, enable));
    }
    get isCamEnabled() {
        return this.mediaManager.isCamEnabled;
    }
    get isMicEnabled() {
        return this.mediaManager.isMicEnabled;
    }
    get isSharingScreen() {
        return this.mediaManager.isSharingScreen;
    }
    get state() {
        return this._state;
    }
    set state(state) {
        if (this._state === state) return;
        this._state = state;
        this._callbacks.onTransportStateChanged?.(state);
    }
    tracks() {
        return this.mediaManager.tracks();
    }
    sdpFilterCodec(kind, codec, realSdp) {
        const allowed = [];
        const rtxRegex = new RegExp("a=fmtp:(\\d+) apt=(\\d+)\\r$");
        const codecRegex = new RegExp("a=rtpmap:([0-9]+) " + this.escapeRegExp(codec));
        const videoRegex = new RegExp("(m=" + kind + " .*?)( ([0-9]+))*\\s*$");
        const lines = realSdp.split("\n");
        let isKind = false;
        for(let i = 0; i < lines.length; i++){
            if (lines[i].startsWith("m=" + kind + " ")) isKind = true;
            else if (lines[i].startsWith("m=")) isKind = false;
            if (isKind) {
                const match = lines[i].match(codecRegex);
                if (match) allowed.push(parseInt(match[1]));
                const matchRtx = lines[i].match(rtxRegex);
                if (matchRtx && allowed.includes(parseInt(matchRtx[2]))) allowed.push(parseInt(matchRtx[1]));
            }
        }
        const skipRegex = "a=(fmtp|rtcp-fb|rtpmap):([0-9]+)";
        let sdp = "";
        isKind = false;
        for(let i = 0; i < lines.length; i++){
            if (lines[i].startsWith("m=" + kind + " ")) isKind = true;
            else if (lines[i].startsWith("m=")) isKind = false;
            if (isKind) {
                const skipMatch = lines[i].match(skipRegex);
                if (skipMatch && !allowed.includes(parseInt(skipMatch[2]))) continue;
                else if (lines[i].match(videoRegex)) sdp += lines[i].replace(videoRegex, "$1 " + allowed.join(" ")) + "\n";
                else sdp += lines[i] + "\n";
            } else sdp += lines[i] + "\n";
        }
        return sdp;
    }
    escapeRegExp(string) {
        return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    }
}
$8381b73e1f76ead9$export$62043589d053a879.SERVICE_NAME = "small-webrtc-transport";
const $8381b73e1f76ead9$var$botParticipant = (pc_id)=>({
        id: pc_id || "bot",
        local: false,
        name: "bot"
    });


$parcel$exportWildcard(module.exports, $8381b73e1f76ead9$exports);
