// @pipecat-ai/client-js/dist/index.js
var $58KWk$events = require("events");
var $58KWk$uuid = require("uuid");
var $58KWk$bowser = require("bowser");


function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || Object.prototype.hasOwnProperty.call(dest, key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ var $c2758f3146a53d04$exports = {};
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ var $db0fd363e17a8e86$exports = {};

$parcel$export($db0fd363e17a8e86$exports, "A11ySnapshotStreamer", () => $db0fd363e17a8e86$export$d1b25383c8328718);
/**
 * Copyright (c) 2026, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ /**
 * Framework-agnostic helper that drives accessibility-snapshot
 * streaming. Wraps the walker, a `MutationObserver`, and the other
 * triggers (scrollend, resize, focus, visibilitychange) into a single
 * object with `start()` / `stop()`. React apps use
 * `useUISnapshot` which is a thin wrapper around this; vanilla JS
 * or non-React apps instantiate the class directly.
 */ /**
 * Copyright (c) 2026, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ // ---------------------------------------------------------------------------
// Tunables
// ---------------------------------------------------------------------------
const $5919cad7020843f0$var$MAX_DEPTH = 10;
const $5919cad7020843f0$var$MAX_NODES = 200;
const $5919cad7020843f0$var$MAX_CHILDREN_PER_NODE = 50;
const $5919cad7020843f0$var$NAME_MAX = 100;
// Tighter cap on emitted `<option>` children. Country / state pickers
// can have hundreds of entries; truncating at 20 keeps the snapshot
// useful without ballooning LLM context.
const $5919cad7020843f0$var$MAX_SELECT_OPTIONS = 20;
// Selections benefit from preserving paragraph structure, but the worker
// only needs enough text to disambiguate the referent. 2000 chars is
// roughly 500 tokens — meaningful context without dominating the
// `<ui_state>` injection.
const $5919cad7020843f0$var$SELECTION_TEXT_MAX = 2000;
// ---------------------------------------------------------------------------
// Ref registry: stable `e{N}` IDs per DOM node, persist as long as the
// node is mounted. WeakMap keeps us from leaking detached nodes.
// ---------------------------------------------------------------------------
const $5919cad7020843f0$var$refMap = new WeakMap();
const $5919cad7020843f0$var$WeakRefCtor = typeof WeakRef === "undefined" ? undefined : WeakRef;
const $5919cad7020843f0$var$refToElement = new Map();
let $5919cad7020843f0$var$refCounter = 0;
function $5919cad7020843f0$var$getRef(el) {
    const existing = $5919cad7020843f0$var$refMap.get(el);
    if (existing) return existing;
    const ref = `e${++$5919cad7020843f0$var$refCounter}`;
    $5919cad7020843f0$var$refMap.set(el, ref);
    $5919cad7020843f0$var$refToElement.set(ref, $5919cad7020843f0$var$WeakRefCtor ? new $5919cad7020843f0$var$WeakRefCtor(el) : el);
    return ref;
}
function $5919cad7020843f0$export$792dfd553abe8e9a(ref) {
    const entry = $5919cad7020843f0$var$refToElement.get(ref);
    if (!entry) return null;
    const el = "deref" in entry ? entry.deref() : entry;
    if (!el) {
        $5919cad7020843f0$var$refToElement.delete(ref);
        return null;
    }
    if (!el.isConnected) {
        $5919cad7020843f0$var$refToElement.delete(ref);
        return null;
    }
    return el;
}
function $5919cad7020843f0$export$5174ac1454ce16bc(el) {
    return $5919cad7020843f0$var$refMap.get(el) ?? null;
}
function $5919cad7020843f0$export$b09bfb5aa97446d5() {
    $5919cad7020843f0$var$refCounter = 0;
    $5919cad7020843f0$var$refToElement.clear();
}
// ---------------------------------------------------------------------------
// Inclusion / exclusion
// ---------------------------------------------------------------------------
const $5919cad7020843f0$var$EXCLUDED_TAGS = new Set([
    "script",
    "style",
    "link",
    "meta",
    "noscript",
    "template"
]);
function $5919cad7020843f0$var$isExcluded(el) {
    if ($5919cad7020843f0$var$EXCLUDED_TAGS.has(el.tagName.toLowerCase())) return true;
    if (el.getAttribute("aria-hidden") === "true") return true;
    // Explicit app-level opt-out for PII / sensitive subtrees. Skips the
    // element and its descendants entirely, without needing to also
    // hide it from screen readers (unlike aria-hidden).
    if (el.hasAttribute("data-a11y-exclude")) return true;
    if (el.hidden) return true;
    // offsetParent == null is a fast-ish check for display:none for most
    // elements (doesn't catch position:fixed hidden or visibility:hidden,
    // but catches the common case). Full getComputedStyle is ~10x slower
    // and we're running on every mutation.
    if (el instanceof HTMLElement && el.offsetParent === null && el.tagName !== "BODY") {
        // One false positive: position: fixed with display: block looks
        // like offsetParent=null but is still visible. Fall back to
        // getComputedStyle only for those.
        const style = window.getComputedStyle(el);
        if (style.display === "none" || style.visibility === "hidden") return true;
    // If fixed/sticky and rendered, let it through.
    }
    return false;
}
// ---------------------------------------------------------------------------
// Role derivation
// ---------------------------------------------------------------------------
const $5919cad7020843f0$var$INTERACTIVE_ROLES = new Set([
    "button",
    "link",
    "checkbox",
    "radio",
    "textbox",
    "combobox",
    "switch",
    "menuitem",
    "tab"
]);
const $5919cad7020843f0$var$LEAF_ROLES = new Set([
    ...$5919cad7020843f0$var$INTERACTIVE_ROLES,
    "heading",
    "img"
]);
function $5919cad7020843f0$var$hasAccessibleName(el) {
    if (el.hasAttribute("aria-label")) return true;
    if (el.hasAttribute("aria-labelledby")) return true;
    if ((el.textContent ?? "").trim().length > 0) return true;
    return false;
}
/**
 * Derive an ARIA-style role for an element. Returns `null` if the
 * element is a "pure wrapper" (no semantic value on its own); the
 * walker will flatten its children into the parent's list rather than
 * emitting a node.
 */ function $5919cad7020843f0$var$getRole(el) {
    const explicit = el.getAttribute("role");
    if (explicit) return explicit;
    const tag = el.tagName.toLowerCase();
    switch(tag){
        case "main":
        case "nav":
        case "header":
        case "aside":
        case "footer":
            return tag;
        case "article":
        case "section":
            return $5919cad7020843f0$var$hasAccessibleName(el) ? "region" : null;
        case "h1":
        case "h2":
        case "h3":
        case "h4":
        case "h5":
        case "h6":
            return "heading";
        case "a":
            return el.hasAttribute("href") ? "link" : null;
        case "button":
            return "button";
        case "input":
            {
                const type = el.type;
                if (type === "checkbox") return "checkbox";
                if (type === "radio") return "radio";
                if (type === "submit" || type === "button" || type === "reset") return "button";
                if (type === "hidden") return null;
                return "textbox";
            }
        case "select":
            return "combobox";
        case "textarea":
            return "textbox";
        case "label":
            return null; // consumed by its associated input
        case "img":
            return el.getAttribute("alt") !== null ? "img" : null;
        case "p":
            // Promote prose containers so paragraph-level deixis works:
            // selection anchors at the enclosing paragraph, and the worker
            // can address individual paragraphs by ref for select_text /
            // scroll_to / highlight. Pure-wrapper flattening would lose
            // both. Paragraphs are walked as non-leaf so inline links and
            // emphasis still nest underneath.
            return "paragraph";
        case "ul":
        case "ol":
            return "list";
        case "li":
            return "listitem";
        case "table":
            return "table";
        case "tr":
            return "row";
        case "th":
            return "columnheader";
        case "td":
            return "cell";
        default:
            // Promote anonymous containers that carry aria/tabindex/click.
            if (el.hasAttribute("aria-label") || el.hasAttribute("aria-labelledby") || el.hasAttribute("tabindex")) return "generic";
            return null;
    }
}
// ---------------------------------------------------------------------------
// Name / value / state / level
// ---------------------------------------------------------------------------
function $5919cad7020843f0$var$collapseWhitespace(s) {
    return s.replace(/\s+/g, " ").trim();
}
function $5919cad7020843f0$var$truncate(s, max = $5919cad7020843f0$var$NAME_MAX) {
    const collapsed = $5919cad7020843f0$var$collapseWhitespace(s);
    if (collapsed.length <= max) return collapsed;
    return collapsed.slice(0, max - 1) + "\u2026";
}
function $5919cad7020843f0$var$resolveLabelledBy(el) {
    const ids = el.getAttribute("aria-labelledby");
    if (!ids) return undefined;
    const parts = [];
    for (const id of ids.split(/\s+/)){
        if (!id) continue;
        const target = el.ownerDocument.getElementById(id);
        if (target) parts.push($5919cad7020843f0$var$collectAccessibleText(target));
    }
    const joined = parts.filter(Boolean).join(" ");
    return joined || undefined;
}
/**
 * Collect the accessible text content of an element by walking its
 * descendant text nodes and joining them with a single space across
 * element boundaries. This fixes the common case where a button
 * contains multiple `<span>` children whose text would otherwise
 * concatenate without a separator (`textContent` returns
 * `"FooBar"` for `<span>Foo</span><span>Bar</span>`).
 *
 * Skips aria-hidden and `EXCLUDED_TAGS` descendants.
 */ function $5919cad7020843f0$var$collectAccessibleText(el) {
    const parts = [];
    for(let i = 0; i < el.childNodes.length; i++){
        const child = el.childNodes[i];
        if (child.nodeType === 3 /* TEXT_NODE */ ) {
            const t = (child.textContent ?? "").trim();
            if (t) parts.push(t);
        } else if (child.nodeType === 1 /* ELEMENT_NODE */ ) {
            const childEl = child;
            if ($5919cad7020843f0$var$EXCLUDED_TAGS.has(childEl.tagName.toLowerCase())) continue;
            if (childEl.getAttribute("aria-hidden") === "true") continue;
            const sub = $5919cad7020843f0$var$collectAccessibleText(childEl);
            if (sub) parts.push(sub);
        }
    }
    return parts.join(" ").replace(/\s+/g, " ").trim();
}
function $5919cad7020843f0$var$getName(el, role) {
    const ariaLabel = el.getAttribute("aria-label");
    if (ariaLabel) return $5919cad7020843f0$var$truncate(ariaLabel);
    const labelled = $5919cad7020843f0$var$resolveLabelledBy(el);
    if (labelled) return $5919cad7020843f0$var$truncate(labelled);
    const tag = el.tagName.toLowerCase();
    // Form controls: prefer associated <label>.
    if (tag === "input" || tag === "textarea" || tag === "select") {
        const id = el.getAttribute("id");
        if (id) {
            // Iterate labels rather than using a CSS attribute selector so we
            // avoid needing CSS.escape (which isn't available in jsdom).
            const labels = el.ownerDocument.getElementsByTagName("label");
            for(let i = 0; i < labels.length; i++){
                if (labels[i].htmlFor === id && labels[i].textContent) return $5919cad7020843f0$var$truncate(labels[i].textContent ?? "");
            }
        }
        const wrappingLabel = el.closest("label");
        if (wrappingLabel?.textContent) return $5919cad7020843f0$var$truncate(wrappingLabel.textContent);
        // Fall back to placeholder for inputs.
        const placeholder = el.getAttribute("placeholder");
        if (placeholder) return $5919cad7020843f0$var$truncate(placeholder);
        return undefined;
    }
    // img: alt text.
    if (tag === "img") {
        const alt = el.getAttribute("alt");
        return alt ? $5919cad7020843f0$var$truncate(alt) : undefined;
    }
    // Leaf interactive / heading / link roles, plus paragraphs, use
    // text from descendants. Paragraphs aren't leaves (they may carry
    // nested links / emphasis) but we want the prose preview as the
    // node's name so the LLM has the gist without expanding text
    // children. `collectAccessibleText` joins sibling children with
    // spaces so `<span>Foo</span><span>Bar</span>` yields `"Foo Bar"`
    // rather than `"FooBar"`.
    if ($5919cad7020843f0$var$LEAF_ROLES.has(role) || role === "paragraph") {
        const text = $5919cad7020843f0$var$collectAccessibleText(el);
        return text ? $5919cad7020843f0$var$truncate(text) : undefined;
    }
    return undefined;
}
function $5919cad7020843f0$var$getValue(el) {
    if (el instanceof HTMLInputElement) {
        if (el.type === "password") return undefined;
        if (el.type === "checkbox" || el.type === "radio" || el.type === "hidden") return undefined;
        return el.value || undefined;
    }
    if (el instanceof HTMLTextAreaElement) return el.value || undefined;
    if (el instanceof HTMLSelectElement) {
        // Prefer the selected option's visible text; the raw `value` is
        // often a numeric id or sentinel that means nothing to a reader.
        const selected = el.selectedOptions[0];
        const text = selected?.text?.trim();
        if (text) return text;
        return el.value || undefined;
    }
    return undefined;
}
function $5919cad7020843f0$var$getState(el) {
    const state = [];
    if (el.ownerDocument.activeElement === el) state.push("focused");
    if (el.getAttribute("aria-expanded") === "true") state.push("expanded");
    if (el.getAttribute("aria-selected") === "true") state.push("selected");
    if (el.hasAttribute("disabled") || el.getAttribute("aria-disabled") === "true") state.push("disabled");
    const ariaChecked = el.getAttribute("aria-checked");
    if (ariaChecked === "true") state.push("checked");
    if (el instanceof HTMLInputElement) {
        if ((el.type === "checkbox" || el.type === "radio") && el.checked) {
            if (!state.includes("checked")) state.push("checked");
        }
    }
    return state;
}
/**
 * Return `true` when the element's bounding rect does not
 * intersect the current viewport at all. Any intersection, even
 * partial, counts as visible (matches user intuition). Elements
 * with zero-size rects (`width === 0 && height === 0`) are
 * treated as offscreen.
 */ function $5919cad7020843f0$var$isOffscreen(el) {
    const rect = el.getBoundingClientRect();
    if (rect.width === 0 && rect.height === 0) return true;
    const viewportWidth = window.innerWidth || document.documentElement.clientWidth;
    const viewportHeight = window.innerHeight || document.documentElement.clientHeight;
    if (rect.bottom <= 0) return true;
    if (rect.top >= viewportHeight) return true;
    if (rect.right <= 0) return true;
    if (rect.left >= viewportWidth) return true;
    return false;
}
function $5919cad7020843f0$var$getLevel(el, role) {
    if (role !== "heading") return undefined;
    const tag = el.tagName.toLowerCase();
    const m = tag.match(/^h([1-6])$/);
    if (m) return parseInt(m[1], 10);
    const aria = el.getAttribute("aria-level");
    if (aria) {
        const parsed = parseInt(aria, 10);
        if (!Number.isNaN(parsed)) return parsed;
    }
    return undefined;
}
function $5919cad7020843f0$var$getAriaIntAttr(el, name) {
    const raw = el.getAttribute(name);
    if (!raw) return undefined;
    const parsed = parseInt(raw, 10);
    if (Number.isNaN(parsed) || parsed < 0) return undefined;
    return parsed;
}
function $5919cad7020843f0$var$walk(el, depth, budget, opts, inheritSkipTextNodes = false) {
    if ($5919cad7020843f0$var$isExcluded(el)) return [];
    if (budget.count >= $5919cad7020843f0$var$MAX_NODES) return [];
    if (depth > $5919cad7020843f0$var$MAX_DEPTH) {
        budget.count++;
        return [
            {
                ref: $5919cad7020843f0$var$getRef(el),
                role: "ellipsis",
                name: "<truncated: max depth>"
            }
        ];
    }
    const role = $5919cad7020843f0$var$getRole(el);
    if (role === null) // Pure wrapper: inline children into the parent's child list.
    // Inherit the caller's text-skip flag so e.g. a <span> inside a
    // <p> doesn't leak its text as a duplicate text-role node.
    return $5919cad7020843f0$var$walkChildren(el, depth, budget, opts, {
        skipTextNodes: inheritSkipTextNodes
    });
    budget.count++;
    const node = {
        ref: $5919cad7020843f0$var$getRef(el),
        role: role
    };
    const name = $5919cad7020843f0$var$getName(el, role);
    if (name) node.name = name;
    const value = $5919cad7020843f0$var$getValue(el);
    if (value !== undefined) node.value = value;
    const state = $5919cad7020843f0$var$getState(el);
    if (opts.trackViewport && $5919cad7020843f0$var$isOffscreen(el)) state.push("offscreen");
    if (state.length) node.state = state;
    const level = $5919cad7020843f0$var$getLevel(el, role);
    if (level !== undefined) node.level = level;
    const colcount = $5919cad7020843f0$var$getAriaIntAttr(el, "aria-colcount");
    if (colcount !== undefined) node.colcount = colcount;
    const rowcount = $5919cad7020843f0$var$getAriaIntAttr(el, "aria-rowcount");
    if (rowcount !== undefined) node.rowcount = rowcount;
    if (!$5919cad7020843f0$var$LEAF_ROLES.has(role)) {
        // For paragraphs the text content is already in the node's
        // `name`. Skip raw text-node children so the same prose
        // doesn't appear twice in `<ui_state>`.
        const skipTextNodes = role === "paragraph";
        const children = $5919cad7020843f0$var$walkChildren(el, depth + 1, budget, opts, {
            skipTextNodes: skipTextNodes
        });
        if (children.length > 0) {
            if (children.length > $5919cad7020843f0$var$MAX_CHILDREN_PER_NODE) {
                const kept = children.slice(0, $5919cad7020843f0$var$MAX_CHILDREN_PER_NODE);
                kept.push({
                    ref: `${node.ref}.more`,
                    role: "ellipsis",
                    name: `${children.length - $5919cad7020843f0$var$MAX_CHILDREN_PER_NODE} more`
                });
                node.children = kept;
            } else node.children = children;
        }
    } else if (el instanceof HTMLSelectElement) {
        // Combobox is a leaf role, but for native `<select>` we synthesize
        // `option` children so the worker can see all available choices,
        // not just the one currently selected. Without this, an LLM has no
        // way to know what other values it could ask the user to pick.
        const options = $5919cad7020843f0$var$collectSelectOptions(el, budget, node.ref);
        if (options.length > 0) node.children = options;
    }
    return [
        node
    ];
}
function $5919cad7020843f0$var$collectSelectOptions(select, budget, parentRef) {
    const out = [];
    const all = select.options;
    let emitted = 0;
    for(let i = 0; i < all.length; i++){
        if (budget.count >= $5919cad7020843f0$var$MAX_NODES) break;
        const opt = all[i];
        if (opt.hidden) continue;
        if (opt.getAttribute("aria-hidden") === "true") continue;
        if (emitted >= $5919cad7020843f0$var$MAX_SELECT_OPTIONS) {
            budget.count++;
            out.push({
                ref: `${parentRef}.more`,
                role: "ellipsis",
                name: `${all.length - emitted} more`
            });
            break;
        }
        budget.count++;
        emitted++;
        const text = (opt.text || opt.value || "").trim();
        const optNode = {
            ref: $5919cad7020843f0$var$getRef(opt),
            role: "option",
            name: $5919cad7020843f0$var$truncate(text)
        };
        const state = [];
        if (opt.selected) state.push("selected");
        if (opt.disabled) state.push("disabled");
        if (state.length) optNode.state = state;
        out.push(optNode);
    }
    return out;
}
function $5919cad7020843f0$var$walkChildren(el, depth, budget, opts, childOpts = {}) {
    const out = [];
    // Iterate `childNodes` rather than `children` so we pick up
    // direct text nodes too (e.g. track titles sitting in a pure-wrapper
    // `<div>` next to a Play button). Without this, wrapper divs that
    // carry meaningful text lose it entirely when they get flattened.
    for(let i = 0; i < el.childNodes.length; i++){
        const child = el.childNodes[i];
        if (child.nodeType === 3 /* TEXT_NODE */ ) {
            if (childOpts.skipTextNodes) continue;
            const text = $5919cad7020843f0$var$collapseWhitespace(child.textContent ?? "");
            if (text) {
                if (budget.count >= $5919cad7020843f0$var$MAX_NODES) break;
                budget.count++;
                out.push({
                    ref: "",
                    role: "text",
                    name: $5919cad7020843f0$var$truncate(text)
                });
            }
        } else if (child.nodeType === 1 /* ELEMENT_NODE */ ) {
            const nodes = $5919cad7020843f0$var$walk(child, depth, budget, opts, childOpts.skipTextNodes);
            for (const n of nodes)out.push(n);
            if (budget.count >= $5919cad7020843f0$var$MAX_NODES) break;
        }
    }
    return out;
}
// ---------------------------------------------------------------------------
// Selection
// ---------------------------------------------------------------------------
/**
 * Look up a ref for `el` without assigning a new one.
 *
 * Selections frequently land on text nodes whose closest element
 * ancestor is a structural tag like `<p>` or `<span>` that the
 * walker treats as a pure wrapper and so doesn't ref. Climb the
 * ancestor chain until we find a walked element.
 *
 * Returns `null` when an `aria-hidden="true"` or
 * `data-a11y-exclude` ancestor is encountered before any
 * ref-bearing element, so selections inside subtrees the app has
 * explicitly hidden from accessibility don't leak into the snapshot.
 */ function $5919cad7020843f0$var$findRefBearingAncestor(start) {
    let el = start;
    while(el){
        if (el.getAttribute("aria-hidden") === "true" || el.hasAttribute("data-a11y-exclude")) return null;
        if ($5919cad7020843f0$var$refMap.has(el)) return el;
        el = el.parentElement;
    }
    return null;
}
function $5919cad7020843f0$var$clampSelectionText(text) {
    if (text.length <= $5919cad7020843f0$var$SELECTION_TEXT_MAX) return text;
    return text.slice(0, $5919cad7020843f0$var$SELECTION_TEXT_MAX - 1) + "\u2026";
}
function $5919cad7020843f0$export$e702c6c416f7c8de() {
    if (typeof document === "undefined") return null;
    // Input / textarea: own selection model, distinct from
    // `window.getSelection()`.
    const active = document.activeElement;
    if (active instanceof HTMLInputElement || active instanceof HTMLTextAreaElement) {
        const start = active.selectionStart;
        const end = active.selectionEnd;
        if (start !== null && end !== null && start !== end) {
            const ref = $5919cad7020843f0$var$refMap.get(active);
            if (!ref) return null;
            const text = active.value.slice(start, end);
            if (!text) return null;
            return {
                ref: ref,
                text: $5919cad7020843f0$var$clampSelectionText(text),
                start_offset: start,
                end_offset: end
            };
        }
    }
    // Document selection.
    const sel = document.getSelection();
    if (!sel || sel.rangeCount === 0 || sel.isCollapsed) return null;
    const text = sel.toString();
    if (!text) return null;
    const range = sel.getRangeAt(0);
    let common = range.commonAncestorContainer;
    // Climb to the nearest element node first.
    while(common && common.nodeType !== 1 /* ELEMENT_NODE */ )common = common.parentNode;
    const anchor = $5919cad7020843f0$var$findRefBearingAncestor(common);
    if (!anchor) return null;
    const ref = $5919cad7020843f0$var$refMap.get(anchor);
    if (!ref) return null;
    return {
        ref: ref,
        text: $5919cad7020843f0$var$clampSelectionText(text)
    };
}
function $5919cad7020843f0$export$326b5f56ac5b6149(root, options = {}) {
    const el = root ?? (typeof document !== "undefined" ? document.body : null);
    if (!el) return {
        root: {
            ref: "e0",
            role: "generic"
        },
        captured_at: Date.now()
    };
    const opts = {
        trackViewport: options.trackViewport ?? true
    };
    const budget = {
        count: 0
    };
    const children = $5919cad7020843f0$var$walkChildren(el, 0, budget, opts);
    // Capture selection after the walk so refs are populated for
    // anything visible on the page.
    const selection = $5919cad7020843f0$export$e702c6c416f7c8de();
    return {
        root: {
            ref: $5919cad7020843f0$var$getRef(el),
            role: "generic",
            ...children.length > 0 ? {
                children: children
            } : {}
        },
        captured_at: Date.now(),
        ...selection ? {
            selection: selection
        } : {}
    };
}


class $db0fd363e17a8e86$export$d1b25383c8328718 {
    constructor(emitSnapshot, options = {}){
        this.running = false;
        this.emitSnapshot = emitSnapshot;
        this.debounceMs = options.debounceMs ?? 300;
        this.trackViewport = options.trackViewport ?? true;
        this.logSnapshots = options.logSnapshots ?? false;
    }
    /**
     * Begin streaming. Safe to call multiple times; subsequent calls
     * are no-ops until `stop()` runs.
     */ start() {
        if (this.running) return;
        if (typeof document === "undefined") return;
        this.running = true;
        // Prime with an initial snapshot once the caller has had a chance
        // to mount the UI.
        this.schedule();
        this.observer = new MutationObserver(()=>this.schedule());
        this.observer.observe(document.body, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: [
                "role",
                "aria-label",
                "aria-labelledby",
                "aria-expanded",
                "aria-selected",
                "aria-checked",
                "aria-disabled",
                "aria-level",
                "aria-hidden",
                "aria-colcount",
                "aria-rowcount",
                "data-a11y-exclude",
                "disabled",
                "hidden",
                "tabindex",
                "href"
            ]
        });
        this.focusHandler = ()=>this.schedule();
        document.addEventListener("focusin", this.focusHandler);
        document.addEventListener("focusout", this.focusHandler);
        // Scrollend fires once when a scroll gesture settles - no
        // debounce needed, and firing at rest avoids fighting the
        // browser's animation frames. `capture: true` catches scroll
        // on any scrollable ancestor, not just the window.
        this.scrollEndHandler = ()=>this.schedule();
        window.addEventListener("scrollend", this.scrollEndHandler, {
            capture: true
        });
        // Resize changes the viewport rect, which shifts which nodes are
        // `[offscreen]`. Debounced via `schedule` so a drag-resize
        // doesn't thrash.
        this.resizeHandler = ()=>this.schedule();
        window.addEventListener("resize", this.resizeHandler);
        // Refresh state when the tab becomes visible again; ignore the
        // transition into `hidden`.
        this.visibilityHandler = ()=>{
            if (document.visibilityState === "visible") this.schedule();
        };
        document.addEventListener("visibilitychange", this.visibilityHandler);
        // `selectionchange` fires throughout a drag-select; the existing
        // debounce coalesces the burst into a single snapshot once the
        // user stops moving. Snapshots end up carrying the latest
        // selection alongside the rest of the screen state.
        this.selectionHandler = ()=>this.schedule();
        document.addEventListener("selectionchange", this.selectionHandler);
        // Native form controls (checkbox/radio toggles, text and select edits)
        // change via element PROPERTIES, which a MutationObserver can't see - the
        // attributeFilter above only catches them when an app mirrors state into
        // aria-*. Listen for `input`/`change` in the capture phase so a user edit
        // (or a programmatic change that dispatches these events) refreshes the
        // snapshot. Debounced via `schedule`, so a burst of typing coalesces into
        // a single snapshot at rest.
        this.formHandler = ()=>this.schedule();
        document.addEventListener("input", this.formHandler, {
            capture: true
        });
        document.addEventListener("change", this.formHandler, {
            capture: true
        });
    }
    /** Stop streaming. Safe to call before `start()` or multiple times. */ stop() {
        if (!this.running) return;
        this.running = false;
        if (this.timer !== undefined) {
            clearTimeout(this.timer);
            this.timer = undefined;
        }
        if (this.observer) {
            this.observer.disconnect();
            this.observer = undefined;
        }
        if (typeof document !== "undefined") {
            if (this.focusHandler) {
                document.removeEventListener("focusin", this.focusHandler);
                document.removeEventListener("focusout", this.focusHandler);
            }
            if (this.visibilityHandler) document.removeEventListener("visibilitychange", this.visibilityHandler);
            if (this.selectionHandler) document.removeEventListener("selectionchange", this.selectionHandler);
            if (this.formHandler) {
                document.removeEventListener("input", this.formHandler, {
                    capture: true
                });
                document.removeEventListener("change", this.formHandler, {
                    capture: true
                });
            }
        }
        if (typeof window !== "undefined") {
            if (this.scrollEndHandler) window.removeEventListener("scrollend", this.scrollEndHandler, {
                capture: true
            });
            if (this.resizeHandler) window.removeEventListener("resize", this.resizeHandler);
        }
        this.focusHandler = undefined;
        this.scrollEndHandler = undefined;
        this.resizeHandler = undefined;
        this.visibilityHandler = undefined;
        this.selectionHandler = undefined;
        this.formHandler = undefined;
    }
    schedule() {
        if (!this.running) return;
        if (this.timer !== undefined) clearTimeout(this.timer);
        this.timer = setTimeout(()=>this.emit(), this.debounceMs);
    }
    emit() {
        this.timer = undefined;
        if (!this.running) return;
        try {
            const snapshot = (0, $5919cad7020843f0$export$326b5f56ac5b6149)(undefined, {
                trackViewport: this.trackViewport
            });
            this.emitSnapshot(snapshot);
            if (this.logSnapshots) {
                const nodeCount = $db0fd363e17a8e86$var$countNodes(snapshot.root);
                const estTokens = Math.round(JSON.stringify(snapshot).length / 4);
                // Grouped so it collapses cleanly in DevTools. The raw tree
                // is included last so it can be expanded on demand.
                console.groupCollapsed(`[A11ySnapshotStreamer] emit: ${nodeCount} nodes, ~${estTokens} tokens`);
                console.log("snapshot:", snapshot);
                console.groupEnd();
            }
        } catch  {
        // Swallow walker errors so we don't crash the host app from a
        // background snapshot attempt. The browser's default error
        // reporting still surfaces them in DevTools.
        }
    }
}
function $db0fd363e17a8e86$var$countNodes(node) {
    let count = 1;
    const kids = node.children;
    if (kids) for (const child of kids)count += $db0fd363e17a8e86$var$countNodes(child);
    return count;
}


var $02f900e1428a591c$exports = {};

$parcel$export($02f900e1428a591c$exports, "PipecatClient", () => $02f900e1428a591c$export$8f7f86a77535f7a3);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
var $8cf63be2a2a16953$exports = {};
$8cf63be2a2a16953$exports = JSON.parse("{\"name\":\"@pipecat-ai/client-js\",\"version\":\"1.10.0\",\"license\":\"BSD-2-Clause\",\"main\":\"dist/index.js\",\"module\":\"dist/index.module.js\",\"types\":\"dist/index.d.ts\",\"source\":\"index.ts\",\"repository\":{\"type\":\"git\",\"url\":\"git+https://github.com/pipecat-ai/pipecat-client-web.git\"},\"exports\":{\".\":{\"types\":\"./dist/index.d.ts\",\"import\":\"./dist/index.module.js\",\"require\":\"./dist/index.js\"}},\"files\":[\"dist\",\"package.json\",\"README.md\"],\"scripts\":{\"build\":\"jest --silent --passWithNoTests && parcel build --no-cache\",\"dev\":\"parcel watch\",\"lint\":\"eslint . --report-unused-disable-directives --max-warnings 0\",\"test\":\"jest\"},\"jest\":{\"preset\":\"ts-jest\",\"testEnvironment\":\"jsdom\",\"setupFilesAfterEnv\":[\"<rootDir>/tests/jest.setup.ts\"]},\"devDependencies\":{\"@jest/globals\":\"^29.7.0\",\"@types/clone-deep\":\"^4.0.4\",\"@types/jest\":\"^29.5.12\",\"eslint\":\"^9.11.1\",\"eslint-config-prettier\":\"^9.1.0\",\"eslint-plugin-simple-import-sort\":\"^12.1.1\",\"jest\":\"^29.7.0\",\"jest-environment-jsdom\":\"^30.0.2\",\"ts-jest\":\"^29.2.5\",\"whatwg-fetch\":\"^3.6.20\"},\"dependencies\":{\"@types/events\":\"^3.0.3\",\"bowser\":\"^2.11.0\",\"clone-deep\":\"^4.0.1\",\"events\":\"^3.3.0\",\"typed-emitter\":\"^2.1.0\",\"uuid\":\"^11.1.1\"}}");


var $609db2a766ef81cc$exports = {};

$parcel$export($609db2a766ef81cc$exports, "findElementByRef", () => $5919cad7020843f0$export$792dfd553abe8e9a);
$parcel$export($609db2a766ef81cc$exports, "findRefForElement", () => $5919cad7020843f0$export$5174ac1454ce16bc);
$parcel$export($609db2a766ef81cc$exports, "serializeSelection", () => $5919cad7020843f0$export$e702c6c416f7c8de);
$parcel$export($609db2a766ef81cc$exports, "snapshotDocument", () => $5919cad7020843f0$export$326b5f56ac5b6149);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
var $920632f0518a5f1b$exports = {};

$parcel$export($920632f0518a5f1b$exports, "TransportStateEnum", () => $920632f0518a5f1b$export$8f2038d3679a1d9b);
var $920632f0518a5f1b$export$8f2038d3679a1d9b;
(function(TransportStateEnum) {
    TransportStateEnum["DISCONNECTED"] = "disconnected";
    TransportStateEnum["INITIALIZING"] = "initializing";
    TransportStateEnum["INITIALIZED"] = "initialized";
    TransportStateEnum["AUTHENTICATING"] = "authenticating";
    TransportStateEnum["AUTHENTICATED"] = "authenticated";
    TransportStateEnum["CONNECTING"] = "connecting";
    TransportStateEnum["CONNECTED"] = "connected";
    TransportStateEnum["READY"] = "ready";
    TransportStateEnum["DISCONNECTING"] = "disconnecting";
    TransportStateEnum["ERROR"] = "error";
})($920632f0518a5f1b$export$8f2038d3679a1d9b || ($920632f0518a5f1b$export$8f2038d3679a1d9b = {}));


var $b4e51e049033e221$exports = {};

$parcel$export($b4e51e049033e221$exports, "RTVIError", () => $b4e51e049033e221$export$59b4786f333aac02);
$parcel$export($b4e51e049033e221$exports, "ConnectionTimeoutError", () => $b4e51e049033e221$export$c67992fa684a81a6);
$parcel$export($b4e51e049033e221$exports, "StartBotError", () => $b4e51e049033e221$export$e7544ab812238a61);
$parcel$export($b4e51e049033e221$exports, "TransportStartError", () => $b4e51e049033e221$export$e0624a511a2c4e9);
$parcel$export($b4e51e049033e221$exports, "InvalidTransportParamsError", () => $b4e51e049033e221$export$b6ce555ea7f95fba);
$parcel$export($b4e51e049033e221$exports, "BotNotReadyError", () => $b4e51e049033e221$export$885fb96b850e8fbb);
$parcel$export($b4e51e049033e221$exports, "BotAlreadyStartedError", () => $b4e51e049033e221$export$cc240eab14fa4f50);
$parcel$export($b4e51e049033e221$exports, "UnsupportedFeatureError", () => $b4e51e049033e221$export$bd0820eb8444fcd9);
$parcel$export($b4e51e049033e221$exports, "MessageTooLargeError", () => $b4e51e049033e221$export$78e1011ee1942cf6);
$parcel$export($b4e51e049033e221$exports, "DeviceError", () => $b4e51e049033e221$export$64c9f614187c1e59);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ class $b4e51e049033e221$export$59b4786f333aac02 extends Error {
    constructor(message, status){
        super(message);
        this.status = status;
    }
}
class $b4e51e049033e221$export$c67992fa684a81a6 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Bot did not enter ready state within the specified timeout period.");
    }
}
class $b4e51e049033e221$export$e7544ab812238a61 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message, status){
        super(message ?? `Failed to connect / invalid auth bundle from base url`, status ?? 500);
        this.error = "invalid-request-error";
    }
}
class $b4e51e049033e221$export$e0624a511a2c4e9 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Unable to connect to transport");
    }
}
class $b4e51e049033e221$export$b6ce555ea7f95fba extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Invalid transport connection parameters");
    }
}
class $b4e51e049033e221$export$885fb96b850e8fbb extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Attempt to call action on transport when not in 'ready' state.");
    }
}
class $b4e51e049033e221$export$cc240eab14fa4f50 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Pipecat client has already been started. Please call disconnect() before starting again.");
    }
}
class $b4e51e049033e221$export$bd0820eb8444fcd9 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(feature, source, message){
        let msg = `${feature} not supported${message ? `: ${message}` : ""}`;
        if (source) msg = `${source} does not support ${feature}${message ? `: ${message}` : ""}`;
        super(msg);
        this.feature = feature;
    }
}
class $b4e51e049033e221$export$78e1011ee1942cf6 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(message){
        super(message ?? "Message size exceeds the maximum allowed limit for transport.");
    }
}
class $b4e51e049033e221$export$64c9f614187c1e59 extends $b4e51e049033e221$export$59b4786f333aac02 {
    constructor(devices, type, message, details){
        super(message ?? `Device error for ${devices.join(", ")}: ${type}`);
        this.devices = devices;
        this.type = type;
        this.details = details;
    }
}


var $82c0e475b1b90003$exports = {};

$parcel$export($82c0e475b1b90003$exports, "RTVIEvent", () => $82c0e475b1b90003$export$6b4624d233c61fcb);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ var $82c0e475b1b90003$export$6b4624d233c61fcb;
(function(RTVIEvent) {
    /** local connection state events */ RTVIEvent["Connected"] = "connected";
    RTVIEvent["Disconnected"] = "disconnected";
    RTVIEvent["TransportStateChanged"] = "transportStateChanged";
    /** remote connection state events */ RTVIEvent["BotStarted"] = "botStarted";
    RTVIEvent["BotConnected"] = "botConnected";
    RTVIEvent["BotReady"] = "botReady";
    RTVIEvent["BotDisconnected"] = "botDisconnected";
    RTVIEvent["Error"] = "error";
    /** server messaging */ RTVIEvent["ServerMessage"] = "serverMessage";
    RTVIEvent["ServerResponse"] = "serverResponse";
    RTVIEvent["MessageError"] = "messageError";
    /** UI Worker Protocol */ RTVIEvent["UICommand"] = "uiCommand";
    RTVIEvent["UIJobGroup"] = "uiJobGroup";
    /** service events */ RTVIEvent["Metrics"] = "metrics";
    // vad events
    RTVIEvent["BotStartedSpeaking"] = "botStartedSpeaking";
    RTVIEvent["BotStoppedSpeaking"] = "botStoppedSpeaking";
    RTVIEvent["UserStartedSpeaking"] = "userStartedSpeaking";
    RTVIEvent["UserStoppedSpeaking"] = "userStoppedSpeaking";
    // user mute strategy events
    RTVIEvent["UserMuteStarted"] = "userMuteStarted";
    RTVIEvent["UserMuteStopped"] = "userMuteStopped";
    // stt events
    RTVIEvent["UserTranscript"] = "userTranscript";
    RTVIEvent["BotOutput"] = "botOutput";
    // DEPRECATED
    RTVIEvent["BotTranscript"] = "botTranscript";
    // llm events
    RTVIEvent["BotLlmText"] = "botLlmText";
    RTVIEvent["BotLlmStarted"] = "botLlmStarted";
    RTVIEvent["BotLlmStopped"] = "botLlmStopped";
    // DEPRECATED
    RTVIEvent["LLMFunctionCall"] = "llmFunctionCall";
    RTVIEvent["LLMFunctionCallStarted"] = "llmFunctionCallStarted";
    RTVIEvent["LLMFunctionCallInProgress"] = "llmFunctionCallInProgress";
    RTVIEvent["LLMFunctionCallStopped"] = "llmFunctionCallStopped";
    RTVIEvent["BotLlmSearchResponse"] = "botLlmSearchResponse";
    // tts events
    RTVIEvent["BotTtsText"] = "botTtsText";
    RTVIEvent["BotTtsStarted"] = "botTtsStarted";
    RTVIEvent["BotTtsStopped"] = "botTtsStopped";
    /** participant events */ RTVIEvent["ParticipantConnected"] = "participantConnected";
    RTVIEvent["ParticipantLeft"] = "participantLeft";
    /** media events */ RTVIEvent["TrackStarted"] = "trackStarted";
    RTVIEvent["TrackStopped"] = "trackStopped";
    RTVIEvent["ScreenTrackStarted"] = "screenTrackStarted";
    RTVIEvent["ScreenTrackStopped"] = "screenTrackStopped";
    RTVIEvent["ScreenShareError"] = "screenShareError";
    RTVIEvent["LocalAudioLevel"] = "localAudioLevel";
    RTVIEvent["RemoteAudioLevel"] = "remoteAudioLevel";
    /** media device events */ RTVIEvent["AvailableCamsUpdated"] = "availableCamsUpdated";
    RTVIEvent["AvailableMicsUpdated"] = "availableMicsUpdated";
    RTVIEvent["AvailableSpeakersUpdated"] = "availableSpeakersUpdated";
    RTVIEvent["CamUpdated"] = "camUpdated";
    RTVIEvent["MicUpdated"] = "micUpdated";
    RTVIEvent["SpeakerUpdated"] = "speakerUpdated";
    RTVIEvent["DeviceError"] = "deviceError";
    RTVIEvent["MediaStateUpdated"] = "mediaStateUpdated";
})($82c0e475b1b90003$export$6b4624d233c61fcb || ($82c0e475b1b90003$export$6b4624d233c61fcb = {}));


var $b1a2fe36be912934$exports = {};

$parcel$export($b1a2fe36be912934$exports, "RTVI_PROTOCOL_VERSION", () => $b1a2fe36be912934$export$7bdaf0e0d661a8f5);
$parcel$export($b1a2fe36be912934$exports, "RTVI_MESSAGE_LABEL", () => $b1a2fe36be912934$export$882b13c7fda338f5);
$parcel$export($b1a2fe36be912934$exports, "RTVIMessageType", () => $b1a2fe36be912934$export$38b3db05cbf0e240);
$parcel$export($b1a2fe36be912934$exports, "AggregationType", () => $b1a2fe36be912934$export$fa4739a8a27f18c0);
$parcel$export($b1a2fe36be912934$exports, "setAboutClient", () => $b1a2fe36be912934$export$e4036f9b8ddb7379);
$parcel$export($b1a2fe36be912934$exports, "RTVIMessage", () => $b1a2fe36be912934$export$69aa9ab0334b212);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $b1a2fe36be912934$export$7bdaf0e0d661a8f5 = "1.4.0";
const $b1a2fe36be912934$export$882b13c7fda338f5 = "rtvi-ai";
var $b1a2fe36be912934$export$38b3db05cbf0e240;
(function(RTVIMessageType) {
    /** Outbound Messages */ RTVIMessageType["CLIENT_READY"] = "client-ready";
    RTVIMessageType["DISCONNECT_BOT"] = "disconnect-bot";
    // Client-to-server messages
    RTVIMessageType["CLIENT_MESSAGE"] = "client-message";
    RTVIMessageType["SEND_TEXT"] = "send-text";
    // UI Worker Protocol (client-to-server)
    RTVIMessageType["UI_EVENT"] = "ui-event";
    RTVIMessageType["UI_SNAPSHOT"] = "ui-snapshot";
    RTVIMessageType["UI_CANCEL_JOB_GROUP"] = "ui-cancel-job-group";
    // DEPRECATED
    RTVIMessageType["APPEND_TO_CONTEXT"] = "append-to-context";
    /**
     * Inbound Messages
     * Messages the server-side client sends to our client-side client regarding
     * its state or other non-service-specific messaging.
     */ RTVIMessageType["BOT_READY"] = "bot-ready";
    RTVIMessageType["ERROR"] = "error";
    RTVIMessageType["METRICS"] = "metrics";
    RTVIMessageType["SERVER_MESSAGE"] = "server-message";
    RTVIMessageType["SERVER_RESPONSE"] = "server-response";
    RTVIMessageType["ERROR_RESPONSE"] = "error-response";
    RTVIMessageType["APPEND_TO_CONTEXT_RESULT"] = "append-to-context-result";
    // UI Worker Protocol (server-to-client)
    RTVIMessageType["UI_COMMAND"] = "ui-command";
    RTVIMessageType["UI_JOB_GROUP"] = "ui-job-group";
    /** Speaking and Transcription Messages */ RTVIMessageType["USER_STARTED_SPEAKING"] = "user-started-speaking";
    RTVIMessageType["USER_STOPPED_SPEAKING"] = "user-stopped-speaking";
    RTVIMessageType["BOT_STARTED_SPEAKING"] = "bot-started-speaking";
    RTVIMessageType["BOT_STOPPED_SPEAKING"] = "bot-stopped-speaking";
    // User muted events. These events notify when the server is ignoring audio from the client.
    // The client should continue sending audio normally but may want to show some indication to the user.
    RTVIMessageType["USER_MUTE_STARTED"] = "user-mute-started";
    RTVIMessageType["USER_MUTE_STOPPED"] = "user-mute-stopped";
    RTVIMessageType["USER_TRANSCRIPTION"] = "user-transcription";
    RTVIMessageType["BOT_OUTPUT"] = "bot-output";
    // DEPRECATED
    RTVIMessageType["BOT_TRANSCRIPTION"] = "bot-transcription";
    /** LLM Messages */ RTVIMessageType["USER_LLM_TEXT"] = "user-llm-text";
    RTVIMessageType["BOT_LLM_TEXT"] = "bot-llm-text";
    RTVIMessageType["BOT_LLM_STARTED"] = "bot-llm-started";
    RTVIMessageType["BOT_LLM_STOPPED"] = "bot-llm-stopped";
    // Function calling
    // DECPRECATED
    RTVIMessageType["LLM_FUNCTION_CALL"] = "llm-function-call";
    RTVIMessageType["LLM_FUNCTION_CALL_STARTED"] = "llm-function-call-started";
    RTVIMessageType["LLM_FUNCTION_CALL_IN_PROGRESS"] = "llm-function-call-in-progress";
    RTVIMessageType["LLM_FUNCTION_CALL_STOPPED"] = "llm-function-call-stopped";
    RTVIMessageType["LLM_FUNCTION_CALL_RESULT"] = "llm-function-call-result";
    RTVIMessageType["BOT_LLM_SEARCH_RESPONSE"] = "bot-llm-search-response";
    /** TTS Messages */ RTVIMessageType["BOT_TTS_TEXT"] = "bot-tts-text";
    RTVIMessageType["BOT_TTS_STARTED"] = "bot-tts-started";
    RTVIMessageType["BOT_TTS_STOPPED"] = "bot-tts-stopped";
})($b1a2fe36be912934$export$38b3db05cbf0e240 || ($b1a2fe36be912934$export$38b3db05cbf0e240 = {}));
var $b1a2fe36be912934$export$fa4739a8a27f18c0;
(function(AggregationType) {
    AggregationType["WORD"] = "word";
    AggregationType["SENTENCE"] = "sentence";
})($b1a2fe36be912934$export$fa4739a8a27f18c0 || ($b1a2fe36be912934$export$fa4739a8a27f18c0 = {}));
// ----- Message Classes
let $b1a2fe36be912934$var$_aboutClient;
function $b1a2fe36be912934$export$e4036f9b8ddb7379(about) {
    // allow for partial updates to the about data
    // this allows the client to set the about data at any time
    // before sending the `client-ready` message and not worry about
    // overwriting existing data
    if ($b1a2fe36be912934$var$_aboutClient) $b1a2fe36be912934$var$_aboutClient = {
        ...$b1a2fe36be912934$var$_aboutClient,
        ...about
    };
    else // if no about data is set, set it to the provided value
    $b1a2fe36be912934$var$_aboutClient = about;
}
class $b1a2fe36be912934$export$69aa9ab0334b212 {
    constructor(type, data, id){
        this.label = $b1a2fe36be912934$export$882b13c7fda338f5;
        this.type = type;
        this.data = data;
        this.id = id || (0, $58KWk$uuid.v4)().slice(0, 8);
    }
    // Outbound message types
    static clientReady() {
        return new $b1a2fe36be912934$export$69aa9ab0334b212($b1a2fe36be912934$export$38b3db05cbf0e240.CLIENT_READY, {
            version: $b1a2fe36be912934$export$7bdaf0e0d661a8f5,
            about: $b1a2fe36be912934$var$_aboutClient || {
                library: (0, $8cf63be2a2a16953$exports.name),
                library_version: (0, $8cf63be2a2a16953$exports.version)
            }
        });
    }
    static disconnectBot() {
        return new $b1a2fe36be912934$export$69aa9ab0334b212($b1a2fe36be912934$export$38b3db05cbf0e240.DISCONNECT_BOT, {});
    }
    static error(message, fatal = false) {
        return new $b1a2fe36be912934$export$69aa9ab0334b212($b1a2fe36be912934$export$38b3db05cbf0e240.ERROR, {
            message: message,
            fatal: fatal
        });
    }
}


var $79eae4cbf0ee6583$exports = {};
/**
 * Copyright (c) 2026, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

$parcel$exportWildcard($609db2a766ef81cc$exports, $920632f0518a5f1b$exports);
$parcel$exportWildcard($609db2a766ef81cc$exports, $b4e51e049033e221$exports);
$parcel$exportWildcard($609db2a766ef81cc$exports, $82c0e475b1b90003$exports);
$parcel$exportWildcard($609db2a766ef81cc$exports, $b1a2fe36be912934$exports);
$parcel$exportWildcard($609db2a766ef81cc$exports, $79eae4cbf0ee6583$exports);




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
function $37a63b97d182bc86$export$f1586721024c4dab(_target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function(...args) {
        if (this.state === "ready") return originalMethod.apply(this, args);
        else throw new (0, $b4e51e049033e221$export$885fb96b850e8fbb)(`Attempt to call ${propertyKey.toString()} when transport not in ready state. Await connect() first.`);
    };
    return descriptor;
}
function $37a63b97d182bc86$export$ebc0d747cf8770bc(_target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;
    const states = [
        "authenticating",
        "connecting",
        "connected",
        "ready"
    ];
    descriptor.value = function(...args) {
        if (states.includes(this.state)) throw new (0, $b4e51e049033e221$export$cc240eab14fa4f50)(`Attempt to call ${propertyKey.toString()} when client already started. Please call disconnect() before starting again.`);
        else return originalMethod.apply(this, args);
    };
    return descriptor;
}
function $37a63b97d182bc86$export$808994d0d8c9acb3(states) {
    return function(_target, propertyKey, descriptor) {
        const originalMethod = descriptor.value;
        descriptor.get = function(...args) {
            if (states.includes(this.state)) return originalMethod.apply(this, args);
            else throw new (0, $b4e51e049033e221$export$885fb96b850e8fbb)(`Attempt to call ${propertyKey.toString()} when transport not in ${states}.`);
        };
        return descriptor;
    };
}
function $37a63b97d182bc86$export$5c35b4fe6fa8c9a6(...states) {
    states = [
        "ready",
        ...states
    ];
    return function(_target, propertyKey, descriptor) {
        const originalGetter = descriptor.get;
        descriptor.get = function() {
            if (states.includes(this.state)) return originalGetter?.apply(this);
            else throw new (0, $b4e51e049033e221$export$885fb96b850e8fbb)(`Attempt to call ${propertyKey.toString()} when transport not in ${states}. Await connect() first.`);
        };
        return descriptor;
    };
}


var $3b981481778defd6$exports = {};

$parcel$export($3b981481778defd6$exports, "MessageDispatcher", () => $3b981481778defd6$export$e9a960646cc432aa);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
var $42491a06a064f4c0$exports = {};

$parcel$export($42491a06a064f4c0$exports, "LogLevel", () => $42491a06a064f4c0$export$243e62d78d3b544d);
$parcel$export($42491a06a064f4c0$exports, "logger", () => $42491a06a064f4c0$export$af88d00dbe7f521);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ var $42491a06a064f4c0$export$243e62d78d3b544d;
(function(LogLevel) {
    LogLevel[LogLevel["NONE"] = 0] = "NONE";
    LogLevel[LogLevel["ERROR"] = 1] = "ERROR";
    LogLevel[LogLevel["WARN"] = 2] = "WARN";
    LogLevel[LogLevel["INFO"] = 3] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 4] = "DEBUG";
})($42491a06a064f4c0$export$243e62d78d3b544d || ($42491a06a064f4c0$export$243e62d78d3b544d = {}));
class $42491a06a064f4c0$var$Logger {
    constructor(){
        this.level = $42491a06a064f4c0$export$243e62d78d3b544d.DEBUG;
    }
    static getInstance() {
        if (!$42491a06a064f4c0$var$Logger.instance) $42491a06a064f4c0$var$Logger.instance = new $42491a06a064f4c0$var$Logger();
        return $42491a06a064f4c0$var$Logger.instance;
    }
    setLevel(level) {
        this.level = level;
    }
    debug(...args) {
        if (this.level >= $42491a06a064f4c0$export$243e62d78d3b544d.DEBUG) console.debug(...args);
    }
    info(...args) {
        if (this.level >= $42491a06a064f4c0$export$243e62d78d3b544d.INFO) console.info(...args);
    }
    warn(...args) {
        if (this.level >= $42491a06a064f4c0$export$243e62d78d3b544d.WARN) console.warn(...args);
    }
    error(...args) {
        if (this.level >= $42491a06a064f4c0$export$243e62d78d3b544d.ERROR) console.error(...args);
    }
}
const $42491a06a064f4c0$export$af88d00dbe7f521 = $42491a06a064f4c0$var$Logger.getInstance();


class $3b981481778defd6$export$e9a960646cc432aa {
    constructor(sendMethod){
        this._queue = new Array();
        this._gcInterval = undefined;
        this._queue = [];
        this._sendMethod = sendMethod;
    }
    disconnect() {
        this.clearQueue();
        clearInterval(this._gcInterval);
        this._gcInterval = undefined;
    }
    dispatch(message_data, type = (0, $b1a2fe36be912934$export$38b3db05cbf0e240).CLIENT_MESSAGE, timeout = 10000) {
        if (!this._gcInterval) // start garbage collection if not already running
        this._gcInterval = setInterval(()=>{
            this._gc();
        }, 2000); // Run garbage collection every 2 seconds
        const message = new (0, $b1a2fe36be912934$export$69aa9ab0334b212)(type, message_data);
        const promise = new Promise((resolve, reject)=>{
            this._queue.push({
                message: message,
                timestamp: Date.now(),
                timeout: timeout,
                resolve: resolve,
                reject: reject
            });
        });
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[MessageDispatcher] dispatch", message);
        try {
            this._sendMethod(message);
        } catch (e) {
            (0, $42491a06a064f4c0$export$af88d00dbe7f521).error("[MessageDispatcher] Error sending message", e);
            return Promise.reject(e);
        }
        this._gc();
        return promise;
    }
    clearQueue() {
        this._queue = [];
    }
    _resolveReject(message, resolve = true) {
        const queuedMessage = this._queue.find((msg)=>msg.message.id === message.id);
        if (queuedMessage) {
            if (resolve) {
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[MessageDispatcher] Resolve", message);
                queuedMessage.resolve(message);
            } else {
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[MessageDispatcher] Reject", message);
                queuedMessage.reject(message);
            }
            // Remove message from queue
            this._queue = this._queue.filter((msg)=>msg.message.id !== message.id);
            (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[MessageDispatcher] Queue", this._queue);
        }
        return message;
    }
    resolve(message) {
        return this._resolveReject(message, true);
    }
    reject(message) {
        return this._resolveReject(message, false);
    }
    _gc() {
        const expired = [];
        this._queue = this._queue.filter((msg)=>{
            const isValid = Date.now() - msg.timestamp < msg.timeout;
            if (!isValid) expired.push(msg);
            return isValid;
        });
        expired.forEach((msg)=>{
            if (msg.message.type === (0, $b1a2fe36be912934$export$38b3db05cbf0e240).CLIENT_MESSAGE) msg.reject(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).ERROR_RESPONSE, {
                error: "Timed out waiting for response",
                msgType: msg.message.data.t,
                data: msg.message.data.d,
                fatal: false
            }));
        });
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[MessageDispatcher] GC", this._queue);
    }
}



var $e93e1075413e9e2e$exports = {};

$parcel$export($e93e1075413e9e2e$exports, "isAPIRequest", () => $e93e1075413e9e2e$export$2dd7ca293b2783);
$parcel$export($e93e1075413e9e2e$exports, "makeRequest", () => $e93e1075413e9e2e$export$699251e5611cc6db);

function $e93e1075413e9e2e$export$2dd7ca293b2783(value) {
    if (typeof value === "object" && value !== null && Object.keys(value).includes("endpoint")) {
        const endpoint = value["endpoint"];
        return typeof endpoint === "string" || endpoint instanceof URL || typeof Request !== "undefined" && endpoint instanceof Request;
    }
    return false;
}
async function $e93e1075413e9e2e$export$699251e5611cc6db(cxnOpts, abortController) {
    if (!abortController) abortController = new AbortController();
    let handshakeTimeout;
    return new Promise((resolve, reject)=>{
        (async ()=>{
            if (cxnOpts.timeout) handshakeTimeout = setTimeout(async ()=>{
                abortController.abort();
                reject(new Error("Timed out"));
            }, cxnOpts.timeout);
            let request;
            if (typeof Request !== "undefined" && cxnOpts.endpoint instanceof Request) {
                request = new Request(cxnOpts.endpoint, {
                    signal: abortController.signal
                });
                if (cxnOpts.requestData) (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("[Pipecat Client] requestData in APIRequest is ignored when endpoint is a Request object");
                if (cxnOpts.headers) (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("[Pipecat Client] headers in APIRequest is ignored when endpoint is a Request object");
            } else request = new Request(cxnOpts.endpoint, {
                method: "POST",
                mode: "cors",
                headers: new Headers({
                    "Content-Type": "application/json",
                    ...Object.fromEntries((cxnOpts.headers ?? new Headers()).entries())
                }),
                body: JSON.stringify(cxnOpts.requestData),
                signal: abortController.signal
            });
            (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug(`[Pipecat Client] Fetching from ${request.url}`);
            fetch(request).then((res)=>{
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug(`[Pipecat Client] Received response from ${request.url}`, res);
                if (!res.ok) {
                    reject(res);
                    return;
                }
                return res.json();
            }).then((data)=>{
                resolve(data);
            }).catch((err)=>{
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).error(`[Pipecat Client] Error fetching: ${err}`);
                reject(err);
            }).finally(()=>{
                if (handshakeTimeout) clearTimeout(handshakeTimeout);
            });
        })();
    });
}


var $8e196b651aab6862$exports = {};

$parcel$export($8e196b651aab6862$exports, "Transport", () => $8e196b651aab6862$export$86495b081fef8e52);
$parcel$export($8e196b651aab6862$exports, "TransportWrapper", () => $8e196b651aab6862$export$82b6ede160a64a3c);
/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 
class $8e196b651aab6862$export$86495b081fef8e52 {
    constructor(){
        this._state = "disconnected";
        /**
         * Maximum allowed size in bytes for a single signaling/message payload.
         *
         * The default is 64 KiB (`64 * 1024`), which is chosen to stay well within
         * typical WebSocket, proxy, and intermediary limits and to discourage
         * transports from sending very large payloads in a single message.
         *
         * Transport implementations may override this value in subclasses or
         * constructors if their underlying transport has stricter or more relaxed
         * limits, as long as they continue to honor this field when enforcing
         * message size constraints.
         */ this._maxMessageSize = 65536; // 64 KiB
    }
    /**
     * Establishes a connection with the remote server. This is the main entry
     * point for the transport to start sending and receiving media and messages.
     * This is called from PipecatClient.connect() and should not be called directly.
     * @param connectParams - This type will ultimately be defned by the transport
     * implementation. It is used to pass connection parameters to the transport.
     */ connect(connectParams) {
        this._abortController = new AbortController();
        let validatedParams = connectParams;
        try {
            validatedParams = this._validateConnectionParams(connectParams);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (e) {
            throw new (0, $b4e51e049033e221$export$59b4786f333aac02)(`Invalid connection params: ${e.message}. Please check your connection params and try again.`);
        }
        return this._connect(validatedParams);
    }
    /**
     * Allow the transports to determine how the bot was started.
     */ get startBotParams() {
        return this._startBotParams;
    }
    /**
     * Set the parameters used to start the bot.
     * @param startBotParams
     */ set startBotParams(startBotParams) {
        if (typeof Request !== "undefined" && startBotParams.endpoint instanceof Request) {
            // If the endpoint is a Request object, we have to clone it.
            // Otherwise, we might run into issues with body being already used.
            this._startBotParams = {
                ...startBotParams,
                endpoint: startBotParams.endpoint.clone()
            };
            return;
        }
        this._startBotParams = startBotParams;
    }
    /**
     * Disconnects the transport from the remote server. This is called from
     * PipecatClient.disconnect() and should not be called directly.
     */ disconnect() {
        if (this._abortController) this._abortController.abort();
        return this._disconnect();
    }
    /**
     * Maximum size, in bytes, of a single message that this transport will attempt
     * to send. Callers should ensure that any outbound {@link RTVIMessage} payloads
     * do not exceed this limit to avoid transport or server errors.
     */ get maxMessageSize() {
        return this._maxMessageSize;
    }
}
class $8e196b651aab6862$export$82b6ede160a64a3c {
    constructor(transport){
        this._transport = transport;
        this._proxy = new Proxy(this._transport, {
            get: (target, prop, receiver)=>{
                if (typeof target[prop] === "function") {
                    let errMsg;
                    switch(String(prop)){
                        // Disable methods that modify the lifecycle of the call. These operations
                        // should be performed via the Pipecat client in order to keep state in sync.
                        case "initialize":
                            errMsg = `Direct calls to initialize() are disabled and used internally by the PipecatClient.`;
                            break;
                        case "initDevices":
                            errMsg = `Direct calls to initDevices() are disabled. Please use the PipecatClient.initDevices() wrapper or let PipecatClient.connect() call it for you.`;
                            break;
                        case "sendReadyMessage":
                            errMsg = `Direct calls to sendReadyMessage() are disabled and used internally by the PipecatClient.`;
                            break;
                        case "connect":
                            errMsg = `Direct calls to connect() are disabled. Please use the PipecatClient.connect() wrapper.`;
                            break;
                        case "disconnect":
                            errMsg = `Direct calls to disconnect() are disabled. Please use the PipecatClient.disconnect() wrapper.`;
                            break;
                    }
                    if (errMsg) return ()=>{
                        throw new Error(errMsg);
                    };
                    // Forward other method calls
                    return (...args)=>{
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
                        return target[prop](...args);
                    };
                }
                // Forward property access
                return Reflect.get(target, prop, receiver);
            }
        });
    }
    get proxy() {
        return this._proxy;
    }
}


var $109c6e1fa5564bb5$exports = {};

$parcel$export($109c6e1fa5564bb5$exports, "learnAboutClient", () => $109c6e1fa5564bb5$export$7eb7b0a641098f31);
$parcel$export($109c6e1fa5564bb5$exports, "messageSizeWithinLimit", () => $109c6e1fa5564bb5$export$48f8227f1e7323f5);


function $109c6e1fa5564bb5$export$7eb7b0a641098f31() {
    let about = {
        library: (0, $8cf63be2a2a16953$exports.name),
        library_version: (0, $8cf63be2a2a16953$exports.version),
        platform_details: {}
    };
    // This uses legacy browser user agent parsing, which we still fall
    // back to if the User Agent Hints API is not available
    let navAgentInfo = null;
    if (window?.navigator?.userAgent) try {
        navAgentInfo = (0, ($parcel$interopDefault($58KWk$bowser))).parse(window.navigator.userAgent);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (_) {
    // void
    }
    if (navAgentInfo?.browser?.name) about.platform_details.browser = navAgentInfo.browser.name;
    if (navAgentInfo?.browser?.name === "Safari" && !navAgentInfo.browser.version) about.platform_details.browser_version = "Web View";
    else if (navAgentInfo?.browser?.version) about.platform_details.browser_version = navAgentInfo.browser.version;
    if (navAgentInfo?.platform?.type) about.platform_details.platform_type = navAgentInfo.platform.type;
    if (navAgentInfo?.engine?.name) about.platform_details.engine = navAgentInfo.engine.name;
    if (navAgentInfo?.os) {
        about.platform = navAgentInfo.os.name;
        about.platform_version = navAgentInfo.os.version;
    }
    return about;
}
function $109c6e1fa5564bb5$export$48f8227f1e7323f5(message, maxSize) {
    const getSizeInBytes = (obj)=>{
        const jsonString = JSON.stringify(obj);
        const encoder = new TextEncoder();
        const bytes = encoder.encode(jsonString);
        return bytes.length;
    };
    const size = getSizeInBytes(message);
    return size <= maxSize;
}


var $02f900e1428a591c$var$__decorate = undefined && undefined.__decorate || function(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/**
 * Map a DeviceErrorType onto a DeviceErrorReason that captures the per-device
 * failure mode. Speaker-affecting errors are not represented here because
 * MediaState only tracks mic and cam.
 */ function $02f900e1428a591c$var$deviceErrorReasonFromType(type) {
    switch(type){
        case "in-use":
            return "already-in-use";
        case "permissions":
            return "blocked";
        case "not-found":
            return "not-found";
        case "undefined-mediadevices":
            return "not-supported";
        case "constraints":
            return "invalid-constraints";
        case "unknown":
        default:
            return "unknown";
    }
}
class $02f900e1428a591c$var$RTVIEventEmitter extends (0, ($parcel$interopDefault($58KWk$events))) {
}
class $02f900e1428a591c$export$8f7f86a77535f7a3 extends $02f900e1428a591c$var$RTVIEventEmitter {
    constructor(options){
        super();
        this._functionCallCallbacks = {};
        this._botTranscriptionWarned = false;
        this._llmFunctionCallWarned = false;
        // Per-device device state. Independent of TransportState — driven by
        // initDevices() and DeviceError events, never by transport connect/disconnect.
        // See common_types.ts for the rationale and the daily-react reference.
        this._mediaState = {
            mic: {
                state: "uninitialized"
            },
            cam: {
                state: "uninitialized"
            }
        };
        (0, $b1a2fe36be912934$export$e4036f9b8ddb7379)((0, $109c6e1fa5564bb5$export$7eb7b0a641098f31)());
        this._transport = options.transport;
        this._transportWrapper = new (0, $8e196b651aab6862$export$82b6ede160a64a3c)(this._transport);
        this._disconnectOnBotDisconnect = options.disconnectOnBotDisconnect ?? true;
        // Wrap transport callbacks with event triggers
        // This allows for either functional callbacks or .on / .off event listeners
        const wrappedCallbacks = {
            ...options.callbacks,
            onMessageError: (message)=>{
                options?.callbacks?.onMessageError?.(message);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).MessageError, message);
            },
            onError: (message)=>{
                options?.callbacks?.onError?.(message);
                try {
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).Error, message);
                } catch (e) {
                    if (e instanceof Error && e.message.includes("Unhandled error")) {
                        if (!options?.callbacks?.onError) (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("No onError callback registered to handle error", message);
                    } else (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("Could not emit error", message, e);
                }
                const data = message.data;
                if (data?.fatal) {
                    (0, $42491a06a064f4c0$export$af88d00dbe7f521).error("Fatal error reported. Disconnecting...");
                    this.disconnect();
                }
            },
            onConnected: ()=>{
                options?.callbacks?.onConnected?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).Connected);
            },
            onDisconnected: ()=>{
                options?.callbacks?.onDisconnected?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).Disconnected);
            },
            onTransportStateChanged: (state)=>{
                options?.callbacks?.onTransportStateChanged?.(state);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).TransportStateChanged, state);
                if (state === "ready") this._flushPendingUISnapshot();
            },
            onParticipantJoined: (p)=>{
                options?.callbacks?.onParticipantJoined?.(p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ParticipantConnected, p);
            },
            onParticipantLeft: (p)=>{
                options?.callbacks?.onParticipantLeft?.(p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ParticipantLeft, p);
            },
            onTrackStarted: (track, p)=>{
                options?.callbacks?.onTrackStarted?.(track, p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).TrackStarted, track, p);
            },
            onTrackStopped: (track, p)=>{
                options?.callbacks?.onTrackStopped?.(track, p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).TrackStopped, track, p);
            },
            onScreenTrackStarted: (track, p)=>{
                options?.callbacks?.onScreenTrackStarted?.(track, p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ScreenTrackStarted, track, p);
            },
            onScreenTrackStopped: (track, p)=>{
                options?.callbacks?.onScreenTrackStopped?.(track, p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ScreenTrackStopped, track, p);
            },
            onScreenShareError: (errorMessage)=>{
                options?.callbacks?.onScreenShareError?.(errorMessage);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ScreenShareError, errorMessage);
            },
            onAvailableCamsUpdated: (cams)=>{
                options?.callbacks?.onAvailableCamsUpdated?.(cams);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).AvailableCamsUpdated, cams);
            },
            onAvailableMicsUpdated: (mics)=>{
                options?.callbacks?.onAvailableMicsUpdated?.(mics);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).AvailableMicsUpdated, mics);
            },
            onAvailableSpeakersUpdated: (speakers)=>{
                options?.callbacks?.onAvailableSpeakersUpdated?.(speakers);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).AvailableSpeakersUpdated, speakers);
            },
            onCamUpdated: (cam)=>{
                // Real device selected → permission was granted. Upgrade MediaState
                // (no-op if already granted, sticky if blocked / in-use / etc).
                if (cam?.deviceId) this._markDeviceGranted("cam");
                options?.callbacks?.onCamUpdated?.(cam);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).CamUpdated, cam);
            },
            onMicUpdated: (mic)=>{
                if (mic?.deviceId) this._markDeviceGranted("mic");
                options?.callbacks?.onMicUpdated?.(mic);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).MicUpdated, mic);
            },
            onSpeakerUpdated: (speaker)=>{
                options?.callbacks?.onSpeakerUpdated?.(speaker);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).SpeakerUpdated, speaker);
            },
            onDeviceError: (error)=>{
                // Classify into MediaState in real time. Works the same whether the
                // error fires during an initDevices() call (mid-await) or out of band
                // (e.g. devicechange-driven). The post-transport Permissions API
                // re-query in initDevices() then has the final word for any 'denied'
                // override.
                this._classifyAndApplyDeviceError(error);
                options?.callbacks?.onDeviceError?.(error);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).DeviceError, error);
            },
            onBotStarted: (botResponse)=>{
                options?.callbacks?.onBotStarted?.(botResponse);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotStarted, botResponse);
            },
            onBotConnected: (p)=>{
                options?.callbacks?.onBotConnected?.(p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotConnected, p);
            },
            onBotReady: (botReadyData)=>{
                options?.callbacks?.onBotReady?.(botReadyData);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotReady, botReadyData);
            },
            onBotDisconnected: (p)=>{
                options?.callbacks?.onBotDisconnected?.(p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotDisconnected, p);
                if (this._disconnectOnBotDisconnect) {
                    (0, $42491a06a064f4c0$export$af88d00dbe7f521).info("Bot disconnected. Disconnecting client...");
                    this.disconnect();
                }
            },
            onUserStartedSpeaking: ()=>{
                options?.callbacks?.onUserStartedSpeaking?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UserStartedSpeaking);
            },
            onUserStoppedSpeaking: ()=>{
                options?.callbacks?.onUserStoppedSpeaking?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UserStoppedSpeaking);
            },
            onBotStartedSpeaking: ()=>{
                options?.callbacks?.onBotStartedSpeaking?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotStartedSpeaking);
            },
            onBotStoppedSpeaking: ()=>{
                options?.callbacks?.onBotStoppedSpeaking?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotStoppedSpeaking);
            },
            onRemoteAudioLevel: (level, p)=>{
                options?.callbacks?.onRemoteAudioLevel?.(level, p);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).RemoteAudioLevel, level, p);
            },
            onLocalAudioLevel: (level)=>{
                options?.callbacks?.onLocalAudioLevel?.(level);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).LocalAudioLevel, level);
            },
            onUserMuteStarted: ()=>{
                options?.callbacks?.onUserMuteStarted?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UserMuteStarted);
            },
            onUserMuteStopped: ()=>{
                options?.callbacks?.onUserMuteStopped?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UserMuteStopped);
            },
            onUserTranscript: (data)=>{
                options?.callbacks?.onUserTranscript?.(data);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UserTranscript, data);
            },
            onBotOutput: (data)=>{
                options?.callbacks?.onBotOutput?.(data);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotOutput, data);
            },
            onBotTranscript: (text)=>{
                const hasSubscriber = !!options?.callbacks?.onBotTranscript || this.listenerCount((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotTranscript) > 0;
                if (hasSubscriber && !this._botTranscriptionWarned) {
                    (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("[Pipecat Client] Bot transcription is deprecated. Please use the onBotOutput instead.");
                    this._botTranscriptionWarned = true;
                }
                options?.callbacks?.onBotTranscript?.(text);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotTranscript, text);
            },
            onBotLlmText: (text)=>{
                options?.callbacks?.onBotLlmText?.(text);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotLlmText, text);
            },
            onBotLlmStarted: ()=>{
                options?.callbacks?.onBotLlmStarted?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotLlmStarted);
            },
            onBotLlmStopped: ()=>{
                options?.callbacks?.onBotLlmStopped?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotLlmStopped);
            },
            onBotTtsText: (text)=>{
                options?.callbacks?.onBotTtsText?.(text);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotTtsText, text);
            },
            onBotTtsStarted: ()=>{
                options?.callbacks?.onBotTtsStarted?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotTtsStarted);
            },
            onBotTtsStopped: ()=>{
                options?.callbacks?.onBotTtsStopped?.();
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotTtsStopped);
            }
        };
        // Update options to reference wrapped callbacks and config defaults
        this._options = {
            ...options,
            callbacks: wrappedCallbacks,
            enableMic: options.enableMic ?? true,
            enableCam: options.enableCam ?? false,
            enableScreenShare: options.enableScreenShare ?? false
        };
        // Instantiate the transport class and bind message handler
        this._initialize();
        // Get package version number
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[Pipecat Client] Initialized", this.version);
    }
    setLogLevel(level) {
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).setLevel(level);
    }
    // ------ Transport methods
    /**
     * Initialize local media devices.
     *
     * Drives MediaState transitions: both mic and cam move to 'initializing' on
     * entry. On success, each device moves to 'granted' only if the transport
     * reports it as acquired (via onMicUpdated / onCamUpdated with a real
     * deviceId); otherwise that device falls back to 'uninitialized'. On
     * failure the in-flight DeviceError (if any) classifies the affected
     * device(s) per-device; anything still at 'initializing' falls back to
     * 'unknown'. The original error is always re-thrown.
     *
     * Calling this again after a failure is the recovery path — a second call
     * re-enters 'initializing' and reclassifies. There is no separate
     * retryDevices() method.
     */ async initDevices() {
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[Pipecat Client] Initializing devices...");
        // Both devices enter the lifecycle, regardless of enableMic / enableCam.
        // The actual transport behavior is asymmetric and not predictable from
        // options alone (e.g. daily-js's startCamera honors startVideoOff but not
        // startAudioOff — it acquires the mic even when the caller said
        // enableMic: false). MediaState mirrors what the transport actually did,
        // sourced from onMicUpdated / onCamUpdated events that fire with a real
        // deviceId only when permission was granted. Devices the transport never
        // speaks to fall back to 'uninitialized' below.
        this._setMediaState({
            mic: {
                state: "initializing"
            },
            cam: {
                state: "initializing"
            }
        });
        try {
            await this._transport.initDevices();
            // Transport resolved. Per-device transitions during the await:
            //   - onMicUpdated / onCamUpdated upgraded reported devices to 'granted'
            //   - onDeviceError applied per-device 'error' classifications
            // Anything still at 'initializing' wasn't reported either way — the
            // transport simply didn't speak to that device (e.g. daily-js skipping
            // cam under startVideoOff: true). Fall it back to 'uninitialized'.
            this._resolveLingeringInitializing({
                state: "uninitialized"
            });
        } catch (error) {
            // Transport rejected. Same as above but the fallback for unspoken-to
            // devices is 'unknown' — something failed and we can't tell whether
            // this device would have worked.
            this._resolveLingeringInitializing({
                state: "error",
                reason: "unknown"
            });
            throw error;
        } finally{
            // Re-query the Permissions API now that the prompt (if any) has been
            // dismissed. Authoritative source for 'denied' regardless of what the
            // transport said. See the helper for the under-reporting rationale.
            await this._enrichFromPermissionsAPI();
        }
    }
    /**
     * After the transport's initDevices() resolves or rejects, any device
     * still at 'initializing' didn't receive a 'granted' upgrade from
     * onMicUpdated / onCamUpdated and wasn't classified by a DeviceError
     * (which would have moved it to 'error'). Apply the supplied fallback so
     * it doesn't linger.
     *
     * On success, fallback is 'uninitialized' (the transport simply didn't
     * speak to that device — e.g. daily-js skipping cam under
     * startVideoOff: true). On failure, fallback is an 'unknown' error (we
     * know something went wrong but can't pin it on this device).
     */ _resolveLingeringInitializing(fallback) {
        const patch = {};
        for (const kind of [
            "mic",
            "cam"
        ])if (this._mediaState[kind].state === "initializing") patch[kind] = fallback;
        if (Object.keys(patch).length > 0) this._setMediaState(patch);
    }
    /**
     * Upgrade a device to 'granted'. Called from the wrapped onMicUpdated /
     * onCamUpdated when the transport reports an actual selected device.
     * Allowed from any state — a previously errored device (e.g. 'not-found'
     * because the cam was unplugged) can recover when the device reappears
     * on a subsequent initDevices() call.
     */ _markDeviceGranted(kind) {
        if (this._mediaState[kind].state !== "granted") this._setMediaState({
            [kind]: {
                state: "granted"
            }
        });
    }
    /**
     * startBot() is a method that initiates the bot by posting to a specified endpoint
     * that optionally returns connection parameters for establishing a transport session.
     * @param startBotParams
     * @returns Promise that resolves to TransportConnectionParams or unknown
     */ async startBot(startBotParams) {
        // Implicit init when devices haven't been initialized yet. Pre-Plan-A
        // this gate read transport.state === "disconnected", which fired both
        // pre-init AND post-session — leading to redundant initDevices() calls
        // on reconnect. needsInit(mediaState) is the unambiguous replacement.
        if (this.needsInit()) await this.initDevices();
        this._transport.state = "authenticating";
        this._transport.startBotParams = startBotParams;
        this._abortController = new AbortController();
        let response;
        try {
            response = await (0, $e93e1075413e9e2e$export$699251e5611cc6db)(startBotParams, this._abortController);
        } catch (e) {
            let errMsg = "An unknown error occurred while starting the bot.";
            let status;
            if (e instanceof Response) {
                const errResp = await e.json();
                errMsg = errResp.info ?? errResp.detail ?? e.statusText;
                status = e.status;
            } else if (e instanceof Error) errMsg = e.message;
            this._options.callbacks?.onError?.(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).ERROR_RESPONSE, {
                message: errMsg,
                fatal: true
            }));
            throw new $b4e51e049033e221$export$e7544ab812238a61(errMsg, status);
        }
        this._transport.state = "authenticated";
        this._options.callbacks?.onBotStarted?.(response);
        return response;
    }
    /**
     * The `connect` function establishes a transport session and awaits a
     * bot-ready signal, handling various connection states and errors.
     * @param {TransportConnectionParams} [connectParams] -
     * The `connectParams` parameter in the `connect` method should be of type
     * `TransportConnectionParams`. This parameter is passed to the transport
     * for establishing a transport session.
     * NOTE: `connectParams` as type `ConnectionEndpoint` IS NOW DEPRECATED. If you
     * want to authenticate and connect to a bot in one step, use
     * `startBotAndConnect()` instead.
     * @returns The `connect` method returns a Promise that resolves to an unknown value.
     */ async connect(connectParams) {
        if (connectParams && (0, $e93e1075413e9e2e$export$2dd7ca293b2783)(connectParams)) {
            (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("Calling connect with an API endpoint is deprecated. Use startBotAndConnect() instead.");
            return this.startBotAndConnect(connectParams);
        }
        // Establish transport session and await bot ready signal
        return new Promise((resolve, reject)=>{
            (async ()=>{
                this._connectResolve = resolve;
                if (this.needsInit()) await this.initDevices();
                try {
                    await this._transport.connect(connectParams);
                    await this._transport.sendReadyMessage();
                } catch (e) {
                    this.disconnect();
                    reject(e);
                    return;
                }
            })();
        });
    }
    async startBotAndConnect(startBotParams) {
        const connectionParams = await this.startBot(startBotParams);
        return this.connect(connectionParams);
    }
    /**
     * Disconnect the voice client from the transport
     * Reset / reinitialize transport and abort any pending requests
     */ async disconnect() {
        this.stopUISnapshotStream();
        await this._transport.disconnect();
        this._messageDispatcher.disconnect();
    }
    /**
     * The _initialize function performs internal set up of the transport and
     * message dispatcher.
     */ _initialize() {
        this._transport.initialize(this._options, this.handleMessage.bind(this));
        // Create a new message dispatch queue for async message handling
        this._messageDispatcher = new (0, $3b981481778defd6$export$e9a960646cc432aa)(this._sendMessage.bind(this));
    }
    /**
     * Apply a partial MediaState patch and emit MediaStateUpdated if the patch
     * actually changes anything. The callback always receives a fresh object.
     */ _setMediaState(patch) {
        const next = {
            ...this._mediaState,
            ...patch
        };
        if (this._statusEquals(next.mic, this._mediaState.mic) && this._statusEquals(next.cam, this._mediaState.cam)) return;
        this._mediaState = next;
        this._options.callbacks?.onMediaStateChanged?.(this.mediaState);
        this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).MediaStateUpdated, this.mediaState);
    }
    /**
     * Structural equality for two DeviceStatus values. Distinct error
     * statuses (different `reason` or `details`) are treated as distinct so a
     * status update fires when the underlying error changes, even if `state`
     * was already 'error'.
     */ _statusEquals(a, b) {
        if (a.state !== b.state) return false;
        if (a.state === "error" && b.state === "error") return a.reason === b.reason && a.details === b.details;
        return true;
    }
    /**
     * Map a DeviceError onto a partial MediaState patch and apply it. Mirrors
     * daily-react's camera-error classifier — affected devices flip to an
     * `'error'` status carrying the reason and the original error payload.
     */ _classifyAndApplyDeviceError(error) {
        const status = {
            state: "error",
            reason: $02f900e1428a591c$var$deviceErrorReasonFromType(error.type),
            details: error.details
        };
        const patch = {};
        if (error.devices.includes("mic")) patch.mic = status;
        if (error.devices.includes("cam")) patch.cam = status;
        if (Object.keys(patch).length === 0) return; // speaker-only or empty
        this._setMediaState(patch);
    }
    /**
     * Permissions API enrichment, run AFTER the transport's initDevices()
     * resolves. By that point the prompt (if any) has been dismissed, and the
     * Permissions API's `denied` answer is authoritative — it overrides any
     * under-reported DeviceError.
     *
     * Concrete case worth flagging: on a page where the user previously
     * blocked permissions, daily-js's `camera-error` only names whichever
     * device the transport tried first when re-initializing — even though
     * both are blocked. Re-querying here catches the missing one. Worth a
     * follow-up daily-js ticket.
     *
     * Silently no-ops where the API is unavailable (Safari, some mobile
     * browsers) or throws on an unsupported descriptor.
     */ async _enrichFromPermissionsAPI() {
        const permissions = globalThis.navigator?.permissions;
        if (!permissions?.query) return;
        const query = permissions.query.bind(permissions);
        const patch = {};
        await Promise.all([
            "mic",
            "cam"
        ].map(async (kind)=>{
            try {
                const result = await query({
                    name: kind === "mic" ? "microphone" : "camera"
                });
                if (result.state === "denied") patch[kind] = {
                    state: "error",
                    reason: "blocked"
                };
            } catch  {
            // Browsers may throw on unsupported descriptor names — swallow.
            }
        }));
        if (Object.keys(patch).length > 0) this._setMediaState(patch);
    }
    /**
     * Internal wrapper around the transport's sendMessage method
     */ _sendMessage(message) {
        if (!(0, $109c6e1fa5564bb5$export$48f8227f1e7323f5)(message, this._transport.maxMessageSize)) {
            const msg = `Message data too large. Max size is ${this._transport.maxMessageSize}`;
            this._options.callbacks?.onError?.((0, $b1a2fe36be912934$export$69aa9ab0334b212).error(msg, false));
            throw new $b4e51e049033e221$export$78e1011ee1942cf6(msg);
        }
        try {
            this._transport.sendMessage(message);
        } catch (error) {
            if (error instanceof Error) this._options.callbacks?.onError?.((0, $b1a2fe36be912934$export$69aa9ab0334b212).error(error.message, false));
            else this._options.callbacks?.onError?.((0, $b1a2fe36be912934$export$69aa9ab0334b212).error("Unknown error sending message", false));
            throw error;
        }
    }
    /**
     * Get the current state of the transport
     */ get connected() {
        return [
            "connected",
            "ready"
        ].includes(this._transport.state);
    }
    get transport() {
        return this._transportWrapper.proxy;
    }
    get state() {
        return this._transport.state;
    }
    /**
     * Per-device device state (mic, cam). Independent of transport state.
     *
     * Updated by initDevices() and DeviceError events. Returns a snapshot — to
     * track changes, subscribe to RTVIEvent.MediaStateUpdated or pass an
     * onMediaStateChanged callback in the client constructor.
     */ get mediaState() {
        // Deep snapshot — DeviceStatus is a nested object, so a shallow spread
        // would still hand the caller a reference to our per-device records.
        return {
            mic: {
                ...this._mediaState.mic
            },
            cam: {
                ...this._mediaState.cam
            }
        };
    }
    /**
     * Whether initDevices() still has work to do. Returns true if any device
     * the caller opted into (enableMic / enableCam) is still 'uninitialized'.
     * Devices the caller opted out of are not considered — they stay
     * 'uninitialized' by design and must not gate the implicit init.
     *
     * Used internally by connect() / startBot() to decide whether to drive an
     * implicit initDevices(); exposed publicly so consumers (e.g. step 3's
     * useMediaState hook) can branch on the same logic.
     */ needsInit() {
        if (this._options.enableMic !== false && this._mediaState.mic.state === "uninitialized") return true;
        if (this._options.enableCam !== false && this._mediaState.cam.state === "uninitialized") return true;
        return false;
    }
    get version() {
        return (0, (/*@__PURE__*/$parcel$interopDefault($8cf63be2a2a16953$exports))).version;
    }
    // ------ Device methods
    async getAllMics() {
        return await this._transport.getAllMics();
    }
    async getAllCams() {
        return await this._transport.getAllCams();
    }
    async getAllSpeakers() {
        return await this._transport.getAllSpeakers();
    }
    get selectedMic() {
        return this._transport.selectedMic;
    }
    get selectedCam() {
        return this._transport.selectedCam;
    }
    get selectedSpeaker() {
        return this._transport.selectedSpeaker;
    }
    updateMic(micId) {
        this._transport.updateMic(micId);
    }
    updateCam(camId) {
        this._transport.updateCam(camId);
    }
    updateSpeaker(speakerId) {
        this._transport.updateSpeaker(speakerId);
    }
    enableMic(enable) {
        this._transport.enableMic(enable);
    }
    get isMicEnabled() {
        return this._transport.isMicEnabled;
    }
    enableCam(enable) {
        this._transport.enableCam(enable);
    }
    get isCamEnabled() {
        return this._transport.isCamEnabled;
    }
    tracks() {
        return this._transport.tracks();
    }
    enableScreenShare(enable) {
        return this._transport.enableScreenShare(enable);
    }
    get isSharingScreen() {
        return this._transport.isSharingScreen;
    }
    // ------ Messages
    /**
     * Directly send a message to the bot via the transport.
     * Do not await a response.
     * @param msgType - a string representing the message type
     * @param data - a dictionary of data to send with the message
     */ sendClientMessage(msgType, data) {
        this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).CLIENT_MESSAGE, {
            t: msgType,
            d: data
        }));
    }
    /**
     * Send a named UI event to the server as a first-class RTVI
     * `ui-event` message.
     *
     * @param event - App-defined event.
     * @param payload - App-defined payload. Optional.
     */ sendUIEvent(event, payload) {
        const data = {
            event: event,
            payload: payload
        };
        this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).UI_EVENT, data));
    }
    /**
     * Start streaming UI snapshots to the server as
     * first-class `ui-snapshot` RTVI messages.
     *
     * Calling this again replaces any existing managed streamer with
     * the new options.
     */ startUISnapshotStream(options = {}) {
        this.stopUISnapshotStream();
        this._uiSnapshotStreamer = new (0, $db0fd363e17a8e86$export$d1b25383c8328718)((snapshot)=>{
            if (this.state !== "ready") {
                this._pendingUISnapshot = snapshot;
                return;
            }
            this._sendUISnapshot(snapshot);
        }, options);
        this._uiSnapshotStreamer.start();
    }
    /**
     * Stop the managed UI snapshot stream, if one is active.
     */ stopUISnapshotStream() {
        this._uiSnapshotStreamer?.stop();
        this._uiSnapshotStreamer = undefined;
        this._pendingUISnapshot = undefined;
    }
    _sendUISnapshot(snapshot) {
        const data = {
            tree: snapshot
        };
        this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).UI_SNAPSHOT, data));
    }
    _flushPendingUISnapshot() {
        if (!this._pendingUISnapshot || this.state !== "ready") return;
        const snapshot = this._pendingUISnapshot;
        this._pendingUISnapshot = undefined;
        this._sendUISnapshot(snapshot);
    }
    /**
     * Ask the server to cancel an in-flight UI job group.
     *
     * @param jobId - Shared job identifier of the group to cancel.
     * @param reason - Optional human-readable reason logged on the server.
     */ cancelUIJobGroup(jobId, reason) {
        const payload = {
            job_id: jobId
        };
        if (reason !== undefined) payload.reason = reason;
        this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).UI_CANCEL_JOB_GROUP, payload));
    }
    /**
     * Directly send a message to the bot via the transport.
     * Wait for and return the response.
     * @param msgType - a string representing the message type
     * @param data - a dictionary of data to send with the message
     * @param timeout - optional timeout in milliseconds for the response
     */ async sendClientRequest(msgType, data, timeout) {
        const msgData = {
            t: msgType,
            d: data
        };
        const response = await this._messageDispatcher.dispatch(msgData, (0, $b1a2fe36be912934$export$38b3db05cbf0e240).CLIENT_MESSAGE, timeout);
        const ret_data = response.data;
        return ret_data.d;
    }
    registerFunctionCallHandler(functionName, callback) {
        this._functionCallCallbacks[functionName] = callback;
    }
    unregisterFunctionCallHandler(functionName) {
        delete this._functionCallCallbacks[functionName];
    }
    unregisterAllFunctionCallHandlers() {
        this._functionCallCallbacks = {};
    }
    async appendToContext(context) {
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("appendToContext() is deprecated. Use sendText() instead.");
        await this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).APPEND_TO_CONTEXT, {
            role: context.role,
            content: context.content,
            run_immediately: context.run_immediately
        }));
        return true;
    }
    async sendText(content, options = {}) {
        await this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).SEND_TEXT, {
            content: content,
            options: options
        }));
    }
    /**
     * Disconnects the bot, but keeps the session alive
     */ disconnectBot() {
        this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).DISCONNECT_BOT, {}));
    }
    handleMessage(ev) {
        (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[RTVI Message]", ev);
        switch(ev.type){
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_READY:
                {
                    const data = ev.data;
                    const botVersion = data.version ? data.version.split(".").map(Number) : [
                        0,
                        0,
                        0
                    ];
                    (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug(`[Pipecat Client] Bot is ready. Version: ${data.version}`);
                    if (botVersion[0] < 1) (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("[Pipecat Client] Bot version is less than 1.0.0, which may not be compatible with this client.");
                    this._connectResolve?.(ev.data);
                    this._options.callbacks?.onBotReady?.(ev.data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).ERROR:
                this._options.callbacks?.onError?.(ev);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).SERVER_RESPONSE:
                this._messageDispatcher.resolve(ev);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).ERROR_RESPONSE:
                {
                    const resp = this._messageDispatcher.reject(ev);
                    this._options.callbacks?.onMessageError?.(resp);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).USER_STARTED_SPEAKING:
                this._options.callbacks?.onUserStartedSpeaking?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).USER_STOPPED_SPEAKING:
                this._options.callbacks?.onUserStoppedSpeaking?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_STARTED_SPEAKING:
                this._options.callbacks?.onBotStartedSpeaking?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_STOPPED_SPEAKING:
                this._options.callbacks?.onBotStoppedSpeaking?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).USER_MUTE_STARTED:
                this._options.callbacks?.onUserMuteStarted?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).USER_MUTE_STOPPED:
                this._options.callbacks?.onUserMuteStopped?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).USER_TRANSCRIPTION:
                {
                    const TranscriptData = ev.data;
                    this._options.callbacks?.onUserTranscript?.(TranscriptData);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_OUTPUT:
                this._options.callbacks?.onBotOutput?.(ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_TRANSCRIPTION:
                this._options.callbacks?.onBotTranscript?.(ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_LLM_TEXT:
                this._options.callbacks?.onBotLlmText?.(ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_LLM_STARTED:
                this._options.callbacks?.onBotLlmStarted?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_LLM_STOPPED:
                this._options.callbacks?.onBotLlmStopped?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_TTS_TEXT:
                this._options.callbacks?.onBotTtsText?.(ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_TTS_STARTED:
                this._options.callbacks?.onBotTtsStarted?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_TTS_STOPPED:
                this._options.callbacks?.onBotTtsStopped?.();
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).METRICS:
                this._options.callbacks?.onMetrics?.(ev.data);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).Metrics, ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).SERVER_MESSAGE:
                this._options.callbacks?.onServerMessage?.(ev.data);
                this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).ServerMessage, ev.data);
                break;
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).UI_COMMAND:
                {
                    const data = ev.data;
                    this._options.callbacks?.onUICommand?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UICommand, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).UI_JOB_GROUP:
                {
                    const data = ev.data;
                    this._options.callbacks?.onUIJobGroup?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).UIJobGroup, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).LLM_FUNCTION_CALL_STARTED:
                {
                    const data = ev.data;
                    this._options.callbacks?.onLLMFunctionCallStarted?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).LLMFunctionCallStarted, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).LLM_FUNCTION_CALL_IN_PROGRESS:
                {
                    const data = ev.data;
                    this._maybeTriggerFunctionCallCallback(data);
                    this._options.callbacks?.onLLMFunctionCallInProgress?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).LLMFunctionCallInProgress, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).LLM_FUNCTION_CALL_STOPPED:
                {
                    const data = ev.data;
                    this._options.callbacks?.onLLMFunctionCallStopped?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).LLMFunctionCallStopped, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).LLM_FUNCTION_CALL:
                {
                    const data = ev.data;
                    const inProgressData = {
                        function_name: data.function_name,
                        tool_call_id: data.tool_call_id,
                        arguments: data.args
                    };
                    this._maybeTriggerFunctionCallCallback(inProgressData);
                    if (this._options.callbacks?.onLLMFunctionCall) {
                        if (!this._llmFunctionCallWarned) {
                            (0, $42491a06a064f4c0$export$af88d00dbe7f521).warn("[Pipecat Client] onLLMFunctionCall is deprecated. Please use onLLMFunctionCallInProgress instead.");
                            this._llmFunctionCallWarned = true;
                        }
                    }
                    this._options.callbacks?.onLLMFunctionCall?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).LLMFunctionCall, data);
                    break;
                }
            case (0, $b1a2fe36be912934$export$38b3db05cbf0e240).BOT_LLM_SEARCH_RESPONSE:
                {
                    const data = ev.data;
                    this._options.callbacks?.onBotLlmSearchResponse?.(data);
                    this.emit((0, $82c0e475b1b90003$export$6b4624d233c61fcb).BotLlmSearchResponse, data);
                    break;
                }
            default:
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).debug("[Pipecat Client] Unrecognized message type", ev.type);
                break;
        }
    }
    _maybeTriggerFunctionCallCallback(data) {
        // Function call callbacks are meant only for function calls that need information
        // from the client to complete and generate a result. This process requires that
        // the event includes the function name. For client-side logic meant simply to
        // react to the fact that a function call is happening, you should use the
        // traditional onLLMFunctionCallStarted/InProgress/Stopped events instead.
        if (!data.function_name) return;
        const fc = this._functionCallCallbacks[data.function_name];
        if (fc) {
            const params = {
                functionName: data.function_name ?? "",
                arguments: data.arguments ?? {}
            };
            fc(params).then((result)=>{
                if (result == undefined) return;
                this._sendMessage(new (0, $b1a2fe36be912934$export$69aa9ab0334b212)((0, $b1a2fe36be912934$export$38b3db05cbf0e240).LLM_FUNCTION_CALL_RESULT, {
                    function_name: data.function_name,
                    tool_call_id: data.tool_call_id,
                    arguments: data.arguments ?? {},
                    result: result
                }));
            }).catch((error)=>{
                (0, $42491a06a064f4c0$export$af88d00dbe7f521).error("Error in function call callback", error);
            });
        }
    }
}
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$ebc0d747cf8770bc)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "startBot", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$ebc0d747cf8770bc)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "connect", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$ebc0d747cf8770bc)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "startBotAndConnect", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "sendClientMessage", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "sendUIEvent", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "cancelUIJobGroup", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "sendClientRequest", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "appendToContext", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "sendText", null);
$02f900e1428a591c$var$__decorate([
    (0, $37a63b97d182bc86$export$f1586721024c4dab)
], $02f900e1428a591c$export$8f7f86a77535f7a3.prototype, "disconnectBot", null);







$parcel$exportWildcard($c2758f3146a53d04$exports, $db0fd363e17a8e86$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $02f900e1428a591c$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $3b981481778defd6$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $42491a06a064f4c0$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $e93e1075413e9e2e$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $8e196b651aab6862$exports);
$parcel$exportWildcard($c2758f3146a53d04$exports, $109c6e1fa5564bb5$exports);



$parcel$exportWildcard(module.exports, $c2758f3146a53d04$exports);
$parcel$exportWildcard(module.exports, $609db2a766ef81cc$exports);
